// @blocksuite/data-view/dist/core/logical/composite-type.js
import Zod from 'zod';
export class FnTypeInstance {
    constructor(args, rt, vars) {
        this.args = args;
        this.rt = rt;
        this.vars = vars;
        this._validate = fnSchema;
        this._valueType = undefined;
        this.name = 'function';
    }
    subst(ctx) {
        const newCtx = { ...ctx };
        const args = [];
        for (const arg of this.args) {
            const newArg = arg.subst(newCtx);
            if (!newArg) {
                return;
            }
            args.push(newArg);
        }
        const rt = this.rt.subst(newCtx);
        if (!rt) {
            return;
        }
        return ct.fn.instance(args, rt);
    }
    unify(ctx, template, unify) {
        const newCtx = { ...ctx };
        for (let i = 0; i < template.args.length; i++) {
            const arg = template.args[i];
            const realArg = this.args[i];
            if (arg == null) {
                return false;
            }
            if (realArg != null) {
                if (!unify(newCtx, realArg, arg)) {
                    return false;
                }
            }
        }
        return unify(newCtx, template.rt, this.rt);
    }
    valueValidate(value) {
        return fnSchema.safeParse(value).success;
    }
}
const fnSchema = Zod.function();
export class ArrayTypeInstance {
    constructor(element) {
        this.element = element;
        this._valueType = undefined;
        this.name = 'array';
        this._validate = Zod.array(element._validate);
    }
    subst(ctx) {
        const ele = this.element.subst(ctx);
        if (!ele) {
            return;
        }
        return ct.array.instance(ele);
    }
    unify(ctx, type, unify) {
        return unify(ctx, this.element, type.element);
    }
    valueValidate(value) {
        return this._validate.safeParse(value).success;
    }
}
export const ct = {
    fn: {
        is: (type) => {
            return type.name === 'function';
        },
        instance: (args, rt, vars) => {
            return new FnTypeInstance(args, rt, vars ?? []);
        },
    },
    array: {
        is: (type) => {
            return type.name === 'array';
        },
        instance: (element) => {
            return new ArrayTypeInstance(element);
        },
    },
};
