// @blocksuite/data-view/dist/core/logical/type-variable.js
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import Zod from 'zod';
const unknownSchema = Zod.unknown();
export class TypeVarDefinitionInstance {
    constructor(varName, typeConstraint) {
        this.varName = varName;
        this.typeConstraint = typeConstraint;
        this.name = '__TypeVarDefine';
    }
}
export class TypeVarReferenceInstance {
    constructor(varName) {
        this.varName = varName;
        this._validate = unknownSchema;
        this._valueType = undefined;
        this.name = '__TypeVarReference';
    }
    subst(ctx) {
        return ctx[this.varName].type;
    }
    unify(_ctx, _type, _unify) {
        throw new BlockSuiteError(ErrorCode.DatabaseBlockError, 'unexpected type unify, type var reference');
    }
    valueValidate(_value) {
        return true;
    }
}
export const tv = {
    typeVarDefine: {
        create: (name, typeConstraint) => {
            return new TypeVarDefinitionInstance(name, typeConstraint);
        },
    },
    typeVarReference: {
        create: (name) => {
            return new TypeVarReferenceInstance(name);
        },
        is: (type) => {
            return type.name === '__TypeVarReference';
        },
    },
};
export const tRef = tv.typeVarReference.create;
export const tVar = tv.typeVarDefine.create;
