// @blocksuite/data-view/dist/core/detail/detail.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { ArrowDownBigIcon, ArrowUpBigIcon, PlusIcon, } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css, nothing, unsafeCSS } from 'lit';
import { property, query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { keyed } from 'lit/directives/keyed.js';
import { repeat } from 'lit/directives/repeat.js';
import { html } from 'lit/static-html.js';
import { dataViewCommonStyle } from '../common/css-variable.js';
import { renderUniLit, } from '../utils/uni-component/uni-component.js';
import { DetailSelection } from './selection.js';
const styles = css `
  ${unsafeCSS(dataViewCommonStyle('affine-data-view-record-detail'))}
  affine-data-view-record-detail {
    position: relative;
    display: flex;
    flex: 1;
    flex-direction: column;
    padding: 20px;
    gap: 12px;
    background-color: var(--affine-background-primary-color);
    border-radius: 8px;
    height: 100%;
    width: 100%;
    box-sizing: border-box;
  }

  .add-property {
    display: flex;
    align-items: center;
    gap: 4px;
    font-size: var(--data-view-cell-text-size);
    font-style: normal;
    font-weight: 400;
    line-height: var(--data-view-cell-text-line-height);
    color: var(--affine-text-disable-color);
    border-radius: 4px;
    padding: 6px 8px 6px 4px;
    cursor: pointer;
    margin-top: 8px;
    width: max-content;
  }

  .add-property:hover {
    background-color: var(--affine-hover-color);
  }

  .add-property .icon {
    display: flex;
    align-items: center;
  }

  .add-property .icon svg {
    fill: var(--affine-icon-color);
    width: 20px;
    height: 20px;
  }

  .switch-row {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 2px;
    border-radius: 4px;
    cursor: pointer;
    font-size: 22px;
    color: var(--affine-icon-color);
  }

  .switch-row:hover {
    background-color: var(--affine-hover-color);
  }

  .switch-row.disable {
    cursor: default;
    background: none;
    opacity: 0.5;
  }
`;
let RecordDetail = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _view_decorators;
    let _view_initializers = [];
    let _view_extraInitializers = [];
    let _addPropertyButton_decorators;
    let _addPropertyButton_initializers = [];
    let _addPropertyButton_extraInitializers = [];
    let _detailSlots_decorators;
    let _detailSlots_initializers = [];
    let _detailSlots_extraInitializers = [];
    let _openDoc_decorators;
    let _openDoc_initializers = [];
    let _openDoc_extraInitializers = [];
    let _rowId_decorators;
    let _rowId_initializers = [];
    let _rowId_extraInitializers = [];
    return class RecordDetail extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _view_decorators = [property({ attribute: false })];
            _addPropertyButton_decorators = [query('.add-property')];
            _detailSlots_decorators = [property({ attribute: false })];
            _openDoc_decorators = [property({ attribute: false })];
            _rowId_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _view_decorators, { kind: "accessor", name: "view", static: false, private: false, access: { has: obj => "view" in obj, get: obj => obj.view, set: (obj, value) => { obj.view = value; } }, metadata: _metadata }, _view_initializers, _view_extraInitializers);
            __esDecorate(this, null, _addPropertyButton_decorators, { kind: "accessor", name: "addPropertyButton", static: false, private: false, access: { has: obj => "addPropertyButton" in obj, get: obj => obj.addPropertyButton, set: (obj, value) => { obj.addPropertyButton = value; } }, metadata: _metadata }, _addPropertyButton_initializers, _addPropertyButton_extraInitializers);
            __esDecorate(this, null, _detailSlots_decorators, { kind: "accessor", name: "detailSlots", static: false, private: false, access: { has: obj => "detailSlots" in obj, get: obj => obj.detailSlots, set: (obj, value) => { obj.detailSlots = value; } }, metadata: _metadata }, _detailSlots_initializers, _detailSlots_extraInitializers);
            __esDecorate(this, null, _openDoc_decorators, { kind: "accessor", name: "openDoc", static: false, private: false, access: { has: obj => "openDoc" in obj, get: obj => obj.openDoc, set: (obj, value) => { obj.openDoc = value; } }, metadata: _metadata }, _openDoc_initializers, _openDoc_extraInitializers);
            __esDecorate(this, null, _rowId_decorators, { kind: "accessor", name: "rowId", static: false, private: false, access: { has: obj => "rowId" in obj, get: obj => obj.rowId, set: (obj, value) => { obj.rowId = value; } }, metadata: _metadata }, _rowId_initializers, _rowId_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = styles; }
        #view_accessor_storage;
        get view() { return this.#view_accessor_storage; }
        set view(value) { this.#view_accessor_storage = value; }
        get readonly() {
            return this.view.readonly$.value;
        }
        renderHeader() {
            const header = this.detailSlots?.header;
            if (header) {
                const props = {
                    view: this.view,
                    rowId: this.rowId,
                    openDoc: this.openDoc,
                };
                return renderUniLit(header, props);
            }
            return undefined;
        }
        renderNote() {
            const note = this.detailSlots?.note;
            if (note) {
                const props = {
                    view: this.view,
                    rowId: this.rowId,
                    openDoc: this.openDoc,
                };
                return renderUniLit(note, props);
            }
            return undefined;
        }
        connectedCallback() {
            super.connectedCallback();
            this.disposables.addFromEvent(this, 'click', e => {
                e.stopPropagation();
                this.selection.selection = undefined;
            });
            //FIXME: simulate as a widget
            this.dataset.widgetId = 'affine-detail-widget';
        }
        hasNext() {
            return this.view.rowNextGet(this.rowId) != null;
        }
        hasPrev() {
            return this.view.rowPrevGet(this.rowId) != null;
        }
        nextRow() {
            const rowId = this.view.rowNextGet(this.rowId);
            if (rowId == null) {
                return;
            }
            this.rowId = rowId;
            this.requestUpdate();
        }
        prevRow() {
            const rowId = this.view.rowPrevGet(this.rowId);
            if (rowId == null) {
                return;
            }
            this.rowId = rowId;
            this.requestUpdate();
        }
        render() {
            const properties = this.properties$.value;
            const upClass = classMap({
                'switch-row': true,
                disable: !this.hasPrev(),
            });
            const downClass = classMap({
                'switch-row': true,
                disable: !this.hasNext(),
            });
            return html `
      <div
        style="position: absolute;left: 20px;top:20px;display: flex;align-items:center;gap:4px;"
      >
        <div @click="${this.prevRow}" class="${upClass}">
          ${ArrowUpBigIcon()}
        </div>
        <div @click="${this.nextRow}" class="${downClass}">
          ${ArrowDownBigIcon()}
        </div>
      </div>
      <div
        style="width: 100%;max-width: var(--affine-editor-width);display: flex;flex-direction: column;margin: 0 auto;box-sizing: border-box;"
      >
        ${keyed(this.rowId, this.renderHeader())}
        ${repeat(properties, v => v.id, property => {
                return keyed(this.rowId, html ` <affine-data-view-record-field
                .view="${this.view}"
                .column="${property}"
                .rowId="${this.rowId}"
                data-column-id="${property.id}"
              ></affine-data-view-record-field>`);
            })}
        ${!this.readonly
                ? html ` <div class="add-property" @click="${this._clickAddProperty}">
              <div class="icon">${PlusIcon()}</div>
              Add Property
            </div>`
                : nothing}
      </div>
      ${keyed(this.rowId, this.renderNote())}
    `;
        }
        #addPropertyButton_accessor_storage;
        get addPropertyButton() { return this.#addPropertyButton_accessor_storage; }
        set addPropertyButton(value) { this.#addPropertyButton_accessor_storage = value; }
        #detailSlots_accessor_storage;
        get detailSlots() { return this.#detailSlots_accessor_storage; }
        set detailSlots(value) { this.#detailSlots_accessor_storage = value; }
        #openDoc_accessor_storage;
        get openDoc() { return this.#openDoc_accessor_storage; }
        set openDoc(value) { this.#openDoc_accessor_storage = value; }
        #rowId_accessor_storage;
        get rowId() { return this.#rowId_accessor_storage; }
        set rowId(value) { this.#rowId_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._clickAddProperty = () => {
                popMenu(popupTargetFromElement(this.addPropertyButton), {
                    options: {
                        title: {
                            text: 'Add property',
                        },
                        items: [
                            menu.group({
                                items: this.view.propertyMetas.map(meta => {
                                    return menu.action({
                                        name: meta.config.name,
                                        prefix: renderUniLit(this.view.propertyIconGet(meta.type)),
                                        select: () => {
                                            this.view.propertyAdd('end', meta.type);
                                        },
                                    });
                                }),
                            }),
                        ],
                    },
                });
            };
            this.#view_accessor_storage = __runInitializers(this, _view_initializers, void 0);
            this.properties$ = (__runInitializers(this, _view_extraInitializers), computed(() => {
                return this.view.detailProperties$.value.map(id => this.view.propertyGet(id));
            }));
            this.selection = new DetailSelection(this);
            this.#addPropertyButton_accessor_storage = __runInitializers(this, _addPropertyButton_initializers, void 0);
            this.#detailSlots_accessor_storage = (__runInitializers(this, _addPropertyButton_extraInitializers), __runInitializers(this, _detailSlots_initializers, void 0));
            this.#openDoc_accessor_storage = (__runInitializers(this, _detailSlots_extraInitializers), __runInitializers(this, _openDoc_initializers, void 0));
            this.#rowId_accessor_storage = (__runInitializers(this, _openDoc_extraInitializers), __runInitializers(this, _rowId_initializers, void 0));
            __runInitializers(this, _rowId_extraInitializers);
        }
    };
})();
export { RecordDetail };
export const createRecordDetail = (ops) => {
    return html ` <affine-data-view-record-detail
    .view=${ops.view}
    .rowId=${ops.rowId}
    .detailSlots=${ops.detail}
    .openDoc=${ops.openDoc}
    class="data-view-popup-container"
  ></affine-data-view-record-detail>`;
};
