// @blocksuite/data-view/dist/core/detail/field.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { DeleteIcon, DuplicateIcon, MoveLeftIcon, MoveRightIcon, } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css } from 'lit';
import { property, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { createRef } from 'lit/directives/ref.js';
import { html } from 'lit/static-html.js';
import { inputConfig, typeConfig } from '../common/property-menu.js';
import { renderUniLit } from '../utils/uni-component/uni-component.js';
let RecordField = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _column_decorators;
    let _column_initializers = [];
    let _column_extraInitializers = [];
    let _rowId_decorators;
    let _rowId_initializers = [];
    let _rowId_extraInitializers = [];
    let _editing_decorators;
    let _editing_initializers = [];
    let _editing_extraInitializers = [];
    let _isFocus_decorators;
    let _isFocus_initializers = [];
    let _isFocus_extraInitializers = [];
    let _view_decorators;
    let _view_initializers = [];
    let _view_extraInitializers = [];
    return class RecordField extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _column_decorators = [property({ attribute: false })];
            _rowId_decorators = [property({ attribute: false })];
            _editing_decorators = [state()];
            _isFocus_decorators = [state()];
            _view_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _column_decorators, { kind: "accessor", name: "column", static: false, private: false, access: { has: obj => "column" in obj, get: obj => obj.column, set: (obj, value) => { obj.column = value; } }, metadata: _metadata }, _column_initializers, _column_extraInitializers);
            __esDecorate(this, null, _rowId_decorators, { kind: "accessor", name: "rowId", static: false, private: false, access: { has: obj => "rowId" in obj, get: obj => obj.rowId, set: (obj, value) => { obj.rowId = value; } }, metadata: _metadata }, _rowId_initializers, _rowId_extraInitializers);
            __esDecorate(this, null, _editing_decorators, { kind: "accessor", name: "editing", static: false, private: false, access: { has: obj => "editing" in obj, get: obj => obj.editing, set: (obj, value) => { obj.editing = value; } }, metadata: _metadata }, _editing_initializers, _editing_extraInitializers);
            __esDecorate(this, null, _isFocus_decorators, { kind: "accessor", name: "isFocus", static: false, private: false, access: { has: obj => "isFocus" in obj, get: obj => obj.isFocus, set: (obj, value) => { obj.isFocus = value; } }, metadata: _metadata }, _isFocus_initializers, _isFocus_extraInitializers);
            __esDecorate(this, null, _view_decorators, { kind: "accessor", name: "view", static: false, private: false, access: { has: obj => "view" in obj, get: obj => obj.view, set: (obj, value) => { obj.view = value; } }, metadata: _metadata }, _view_initializers, _view_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    affine-data-view-record-field {
      display: flex;
      gap: 12px;
    }

    .field-left {
      padding: 6px;
      display: flex;
      height: max-content;
      align-items: center;
      gap: 6px;
      font-size: var(--data-view-cell-text-size);
      line-height: var(--data-view-cell-text-line-height);
      color: var(--affine-text-secondary-color);
      width: 160px;
      border-radius: 4px;
      cursor: pointer;
      user-select: none;
    }

    .field-left:hover {
      background-color: var(--affine-hover-color);
    }

    affine-data-view-record-field .icon {
      display: flex;
      align-items: center;
      width: 16px;
      height: 16px;
    }

    affine-data-view-record-field .icon svg {
      width: 16px;
      height: 16px;
      fill: var(--affine-icon-color);
    }

    .filed-name {
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .field-content {
      padding: 6px 8px;
      border-radius: 4px;
      flex: 1;
      cursor: pointer;
      display: flex;
      align-items: center;
      border: 1px solid transparent;
    }

    .field-content .affine-database-number {
      text-align: left;
      justify-content: start;
    }

    .field-content:hover {
      background-color: var(--affine-hover-color);
    }

    .field-content.is-editing {
      box-shadow: 0px 0px 0px 2px rgba(30, 150, 235, 0.3);
    }

    .field-content.is-focus {
      border: 1px solid var(--affine-primary-color);
    }

    .field-content.empty::before {
      content: 'Empty';
      color: var(--affine-text-disable-color);
      font-size: 14px;
      line-height: 22px;
    }
  `; }
        #column_accessor_storage;
        get column() { return this.#column_accessor_storage; }
        set column(value) { this.#column_accessor_storage = value; }
        #rowId_accessor_storage;
        get rowId() { return this.#rowId_accessor_storage; }
        set rowId(value) { this.#rowId_accessor_storage = value; }
        get cell() {
            return this._cell.value;
        }
        get readonly() {
            return this.view.readonly$.value;
        }
        render() {
            const column = this.column;
            const props = {
                cell: this.cell$.value,
                isEditing: this.editing,
                selectCurrentCell: this.changeEditing,
            };
            const renderer = this.column.renderer$.value;
            if (!renderer) {
                return;
            }
            const { view, edit } = renderer;
            const contentClass = classMap({
                'field-content': true,
                empty: !this.editing && this.cell$.value.isEmpty$.value,
                'is-editing': this.editing,
                'is-focus': this.isFocus,
            });
            return html `
      <div>
        <div class="field-left" @click="${this._clickLeft}">
          <div class="icon">
            <uni-lit .uni="${this.column.icon}"></uni-lit>
          </div>
          <div class="filed-name">${column.name$.value}</div>
        </div>
      </div>
      <div @click="${this._click}" class="${contentClass}">
        ${renderUniLit(this.editing && edit ? edit : view, props, {
                ref: this._cell,
                class: 'kanban-cell',
            })}
      </div>
    `;
        }
        #editing_accessor_storage;
        get editing() { return this.#editing_accessor_storage; }
        set editing(value) { this.#editing_accessor_storage = value; }
        #isFocus_accessor_storage;
        get isFocus() { return this.#isFocus_accessor_storage; }
        set isFocus(value) { this.#isFocus_accessor_storage = value; }
        #view_accessor_storage;
        get view() { return this.#view_accessor_storage; }
        set view(value) { this.#view_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._cell = createRef();
            this._click = (e) => {
                e.stopPropagation();
                if (this.readonly)
                    return;
                this.changeEditing(true);
            };
            this._clickLeft = (e) => {
                if (this.readonly)
                    return;
                const ele = e.currentTarget;
                const properties = this.view.detailProperties$.value;
                popMenu(popupTargetFromElement(ele), {
                    options: {
                        title: {
                            text: 'Property settings',
                        },
                        items: [
                            menu.group({
                                items: [inputConfig(this.column), typeConfig(this.column)],
                            }),
                            menu.group({
                                items: [
                                    menu.action({
                                        name: 'Move Up',
                                        prefix: html ` <div
                  style="transform: rotate(90deg);display:flex;align-items:center;"
                >
                  ${MoveLeftIcon()}
                </div>`,
                                        hide: () => properties.findIndex(v => v === this.column.id) === 0,
                                        select: () => {
                                            const index = properties.findIndex(v => v === this.column.id);
                                            const targetId = properties[index - 1];
                                            if (!targetId) {
                                                return;
                                            }
                                            this.view.propertyMove(this.column.id, {
                                                id: targetId,
                                                before: true,
                                            });
                                        },
                                    }),
                                    menu.action({
                                        name: 'Move Down',
                                        prefix: html ` <div
                  style="transform: rotate(90deg);display:flex;align-items:center;"
                >
                  ${MoveRightIcon()}
                </div>`,
                                        hide: () => properties.findIndex(v => v === this.column.id) ===
                                            properties.length - 1,
                                        select: () => {
                                            const index = properties.findIndex(v => v === this.column.id);
                                            const targetId = properties[index + 1];
                                            if (!targetId) {
                                                return;
                                            }
                                            this.view.propertyMove(this.column.id, {
                                                id: targetId,
                                                before: false,
                                            });
                                        },
                                    }),
                                ],
                            }),
                            menu.group({
                                name: 'operation',
                                items: [
                                    menu.action({
                                        name: 'Duplicate',
                                        prefix: DuplicateIcon(),
                                        hide: () => !this.column.duplicate || this.column.type$.value === 'title',
                                        select: () => {
                                            this.column.duplicate?.();
                                        },
                                    }),
                                    menu.action({
                                        name: 'Delete',
                                        prefix: DeleteIcon(),
                                        hide: () => !this.column.delete || this.column.type$.value === 'title',
                                        select: () => {
                                            this.column.delete?.();
                                        },
                                        class: { 'delete-item': true },
                                    }),
                                ],
                            }),
                        ],
                    },
                });
            };
            this.#column_accessor_storage = __runInitializers(this, _column_initializers, void 0);
            this.#rowId_accessor_storage = (__runInitializers(this, _column_extraInitializers), __runInitializers(this, _rowId_initializers, void 0));
            this.cell$ = (__runInitializers(this, _rowId_extraInitializers), computed(() => {
                return this.column.cellGet(this.rowId);
            }));
            this.changeEditing = (editing) => {
                const selection = this.closest('affine-data-view-record-detail')?.selection;
                if (selection) {
                    selection.selection = {
                        propertyId: this.column.id,
                        isEditing: editing,
                    };
                }
            };
            this.#editing_accessor_storage = __runInitializers(this, _editing_initializers, false);
            this.#isFocus_accessor_storage = (__runInitializers(this, _editing_extraInitializers), __runInitializers(this, _isFocus_initializers, false));
            this.#view_accessor_storage = (__runInitializers(this, _isFocus_extraInitializers), __runInitializers(this, _view_initializers, void 0));
            __runInitializers(this, _view_extraInitializers);
        }
    };
})();
export { RecordField };
