// @blocksuite/data-view/dist/core/utils/wc-dnd/sort/strategies/vertical-list-sorting.js
const defaultScale = {
    scaleX: 1,
    scaleY: 1,
};
export const verticalListSortingStrategy = ({ activeIndex, activeNodeRect: fallbackActiveRect, rects, overIndex, }) => {
    const activeNodeRect = rects[activeIndex] ?? fallbackActiveRect;
    const strategy = (index) => {
        if (index === activeIndex) {
            const overIndexRect = rects[overIndex];
            if (!overIndexRect) {
                return undefined;
            }
            return {
                x: 0,
                y: activeIndex < overIndex
                    ? overIndexRect.top +
                        overIndexRect.height -
                        (activeNodeRect.top + activeNodeRect.height)
                    : overIndexRect.top - activeNodeRect.top,
                ...defaultScale,
            };
        }
        const itemGap = getItemGap(rects, index, activeIndex);
        if (index > activeIndex && index <= overIndex) {
            return {
                x: 0,
                y: -activeNodeRect.height - itemGap,
                ...defaultScale,
            };
        }
        if (index < activeIndex && index >= overIndex) {
            return {
                x: 0,
                y: activeNodeRect.height + itemGap,
                ...defaultScale,
            };
        }
        return {
            x: 0,
            y: 0,
            ...defaultScale,
        };
    };
    return rects.map((_, index) => strategy(index));
};
function getItemGap(clientRects, index, activeIndex) {
    const currentRect = clientRects[index];
    const previousRect = clientRects[index - 1];
    const nextRect = clientRects[index + 1];
    if (!currentRect) {
        return 0;
    }
    if (activeIndex < index) {
        return previousRect
            ? currentRect.top - (previousRect.top + previousRect.height)
            : nextRect
                ? nextRect.top - (currentRect.top + currentRect.height)
                : 0;
    }
    return nextRect
        ? nextRect.top - (currentRect.top + currentRect.height)
        : previousRect
            ? currentRect.top - (previousRect.top + previousRect.height)
            : 0;
}
