// @prisma/fetch-engine/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  BinaryType: () => import_chunk_66KG5WKV.BinaryType,
  allEngineEnvVarsSet: () => import_chunk_XNR2KDTK.allEngineEnvVarsSet,
  deprecatedEnvVarMap: () => import_chunk_XNR2KDTK.deprecatedEnvVarMap,
  download: () => import_chunk_CCWZC3JD.download,
  engineEnvVarMap: () => import_chunk_XNR2KDTK.engineEnvVarMap,
  getBinaryEnvVarPath: () => import_chunk_XNR2KDTK.getBinaryEnvVarPath,
  getBinaryName: () => import_chunk_CCWZC3JD.getBinaryName,
  getCacheDir: () => import_chunk_4MDZEOKU.getCacheDir,
  getProxyAgent: () => import_chunk_CTLNI6CP.getProxyAgent,
  getVersion: () => import_chunk_CCWZC3JD.getVersion,
  maybeCopyToTmp: () => import_chunk_CCWZC3JD.maybeCopyToTmp,
  overwriteFile: () => import_chunk_4MDZEOKU.overwriteFile,
  plusX: () => import_chunk_CCWZC3JD.plusX,
  vercelPkgPathRegex: () => import_chunk_CCWZC3JD.vercelPkgPathRegex
});
module.exports = __toCommonJS(index_exports);
var import_chunk_CCWZC3JD = require("./chunk-CCWZC3JD.js");
var import_chunk_5564WBCW = require("./chunk-5564WBCW.js");
var import_chunk_7JLQJWOR = require("./chunk-7JLQJWOR.js");
var import_chunk_4GGGRRIN = require("./chunk-4GGGRRIN.js");
var import_chunk_FNYZ6ABB = require("./chunk-FNYZ6ABB.js");
var import_chunk_VLGS4WJI = require("./chunk-VLGS4WJI.js");
var import_chunk_7AOUSONO = require("./chunk-7AOUSONO.js");
var import_chunk_4MDZEOKU = require("./chunk-4MDZEOKU.js");
var import_chunk_XNR2KDTK = require("./chunk-XNR2KDTK.js");
var import_chunk_66KG5WKV = require("./chunk-66KG5WKV.js");
var import_chunk_FKGWOTGU = require("./chunk-FKGWOTGU.js");
var import_chunk_CTLNI6CP = require("./chunk-CTLNI6CP.js");
var import_chunk_2ESYSVXG = require("./chunk-2ESYSVXG.js");
