// @blocksuite/data-view/dist/core/data-source/base.js
import { computed } from '@preact/signals-core';
export class DataSourceBase {
    constructor() {
        this.context = new Map();
    }
    cellValueGet$(rowId, propertyId) {
        return computed(() => this.cellValueGet(rowId, propertyId));
    }
    contextGet(key) {
        return this.context.get(key.key) ?? key.defaultValue;
    }
    contextSet(key, value) {
        this.context.set(key.key, value);
    }
    propertyDataGet$(propertyId) {
        return computed(() => this.propertyDataGet(propertyId));
    }
    propertyDataTypeGet$(propertyId) {
        return computed(() => this.propertyDataTypeGet(propertyId));
    }
    propertyNameGet$(propertyId) {
        return computed(() => this.propertyNameGet(propertyId));
    }
    propertyReadonlyGet(_propertyId) {
        return false;
    }
    propertyReadonlyGet$(propertyId) {
        return computed(() => this.propertyReadonlyGet(propertyId));
    }
    propertyTypeGet$(propertyId) {
        return computed(() => this.propertyTypeGet(propertyId));
    }
    viewDataGet$(viewId) {
        return computed(() => this.viewDataGet(viewId));
    }
    viewMetaGet$(type) {
        return computed(() => this.viewMetaGet(type));
    }
    viewMetaGetById$(viewId) {
        return computed(() => this.viewMetaGetById(viewId));
    }
}
