// @blocksuite/data-view/dist/core/view-manager/single-view.js
import { computed, signal } from '@preact/signals-core';
import { CellBase } from './cell.js';
import { RowBase } from './row.js';
export class SingleViewBase {
    get dataSource() {
        return this.manager.dataSource;
    }
    get featureFlags$() {
        return this.dataSource.featureFlags$;
    }
    get isLocked() {
        return this.lockRows$.value;
    }
    get meta() {
        return this.dataSource.viewMetaGet(this.type);
    }
    get propertyMetas() {
        return this.dataSource.propertyMetas;
    }
    constructor(manager, id) {
        this.manager = manager;
        this.id = id;
        this.searchString = signal('');
        this.traitMap = new Map();
        this.data$ = computed(() => {
            return this.dataSource.viewDataGet(this.id);
        });
        this.lockRows$ = signal(false);
        this.isLocked$ = computed(() => {
            return this.lockRows$.value;
        });
        this.name$ = computed(() => {
            return this.data$.value?.name ?? '';
        });
        this.preRows = [];
        this.properties$ = computed(() => {
            return this.propertyIds$.value.map(id => this.propertyGet(id));
        });
        this.rows$ = computed(() => {
            if (this.lockRows$.value) {
                return this.preRows;
            }
            return (this.preRows = this.rowsMapping(this.dataSource.rows$.value));
        });
        this.vars$ = computed(() => {
            return this.propertiesWithoutFilter$.value.map(id => {
                const v = this.propertyGet(id);
                const propertyMeta = this.dataSource.propertyMetaGet(v.type$.value);
                return {
                    id: v.id,
                    name: v.name$.value,
                    type: propertyMeta.config.type({
                        data: v.data$.value,
                        dataSource: this.dataSource,
                    }),
                    icon: v.icon,
                    propertyType: v.type$.value,
                };
            });
        });
    }
    searchRowsMapping(rows, searchString) {
        return rows.filter(id => {
            if (searchString) {
                const containsSearchString = this.propertyIds$.value.some(propertyId => {
                    return this.cellStringValueGet(id, propertyId)
                        ?.toLowerCase()
                        .includes(searchString?.toLowerCase());
                });
                if (!containsSearchString) {
                    return false;
                }
            }
            return this.isShow(id);
        });
    }
    cellGet(rowId, propertyId) {
        return new CellBase(this, propertyId, rowId);
    }
    cellJsonValueGet(rowId, propertyId) {
        const type = this.propertyTypeGet(propertyId);
        if (!type) {
            return;
        }
        return this.dataSource.propertyMetaGet(type).config.cellToJson({
            value: this.dataSource.cellValueGet(rowId, propertyId),
            data: this.propertyDataGet(propertyId),
            dataSource: this.dataSource,
        });
    }
    cellStringValueGet(rowId, propertyId) {
        const type = this.propertyTypeGet(propertyId);
        if (!type) {
            return;
        }
        return (this.dataSource.propertyMetaGet(type).config.cellToString({
            value: this.dataSource.cellValueGet(rowId, propertyId),
            data: this.propertyDataGet(propertyId),
            dataSource: this.dataSource,
        }) ?? '');
    }
    cellValueGet(rowId, propertyId) {
        const type = this.propertyTypeGet(propertyId);
        if (!type) {
            return;
        }
        const cellValue = this.dataSource.cellValueGet(rowId, propertyId);
        return (this.dataSource.propertyMetaGet(type).config.formatValue?.({
            value: cellValue,
            data: this.propertyDataGet(propertyId),
            dataSource: this.dataSource,
        }) ?? cellValue);
    }
    cellValueSet(rowId, propertyId, value) {
        this.dataSource.cellValueChange(rowId, propertyId, value);
    }
    contextGet(key) {
        return this.dataSource.contextGet(key);
    }
    dataUpdate(updater) {
        this.dataSource.viewDataUpdate(this.id, updater);
    }
    delete() {
        this.manager.viewDelete(this.id);
    }
    duplicate() {
        this.manager.viewDuplicate(this.id);
    }
    lockRows(lock) {
        this.lockRows$.value = lock;
    }
    nameSet(name) {
        this.dataUpdate(() => {
            return {
                name,
            };
        });
    }
    propertyAdd(position, type) {
        const id = this.dataSource.propertyAdd(position, type);
        this.propertyMove(id, position);
        return id;
    }
    propertyDataGet(propertyId) {
        return this.dataSource.propertyDataGet(propertyId);
    }
    propertyDataSet(propertyId, data) {
        this.dataSource.propertyDataSet(propertyId, data);
    }
    propertyDataTypeGet(propertyId) {
        const type = this.propertyTypeGet(propertyId);
        if (!type) {
            return;
        }
        return this.dataSource.propertyMetaGet(type).config.type({
            data: this.propertyDataGet(propertyId),
            dataSource: this.dataSource,
        });
    }
    propertyDelete(propertyId) {
        this.dataSource.propertyDelete(propertyId);
    }
    propertyDuplicate(propertyId) {
        const id = this.dataSource.propertyDuplicate(propertyId);
        this.propertyMove(id, {
            before: false,
            id: propertyId,
        });
    }
    propertyIconGet(type) {
        return this.dataSource.propertyMetaGet(type).renderer.icon;
    }
    propertyIdGetByIndex(index) {
        return this.propertyIds$.value[index];
    }
    propertyIndexGet(propertyId) {
        return this.propertyIds$.value.indexOf(propertyId);
    }
    propertyMetaGet(type) {
        return this.dataSource.propertyMetaGet(type);
    }
    propertyNameGet(propertyId) {
        return this.dataSource.propertyNameGet(propertyId);
    }
    propertyNameSet(propertyId, name) {
        this.dataSource.propertyNameSet(propertyId, name);
    }
    propertyNextGet(propertyId) {
        return this.propertyGet(this.propertyIdGetByIndex(this.propertyIndexGet(propertyId) + 1));
    }
    propertyParseValueFromString(propertyId, cellData) {
        const type = this.propertyTypeGet(propertyId);
        if (!type) {
            return;
        }
        return (this.dataSource.propertyMetaGet(type).config.cellFromString({
            value: cellData,
            data: this.propertyDataGet(propertyId),
            dataSource: this.dataSource,
        }) ?? '');
    }
    propertyPreGet(propertyId) {
        return this.propertyGet(this.propertyIdGetByIndex(this.propertyIndexGet(propertyId) - 1));
    }
    propertyReadonlyGet(propertyId) {
        return this.dataSource.propertyReadonlyGet(propertyId);
    }
    propertyTypeGet(propertyId) {
        return this.dataSource.propertyTypeGet(propertyId);
    }
    propertyTypeSet(propertyId, type) {
        this.dataSource.propertyTypeSet(propertyId, type);
    }
    rowAdd(insertPosition) {
        return this.dataSource.rowAdd(insertPosition);
    }
    rowDelete(ids) {
        this.dataSource.rowDelete(ids);
    }
    rowGet(rowId) {
        return new RowBase(this, rowId);
    }
    rowMove(rowId, position) {
        this.dataSource.rowMove(rowId, position);
    }
    rowsMapping(rows) {
        return this.searchRowsMapping(rows, this.searchString.value);
    }
    setSearch(str) {
        this.searchString.value = str;
    }
    traitGet(key) {
        return this.traitMap.get(key.key);
    }
    traitSet(key, value) {
        this.traitMap.set(key.key, value);
        return value;
    }
}
