// @blocksuite/affine-shared/dist/adapters/plain-text/delta-converter.js
import { createIdentifier } from '@blocksuite/global/di';
import { DeltaASTConverter, } from '../types/adapter.js';
export const InlineDeltaToPlainTextAdapterMatcherIdentifier = createIdentifier('InlineDeltaToPlainTextAdapterMatcher');
export class PlainTextDeltaConverter extends DeltaASTConverter {
    constructor(configs, inlineDeltaMatchers, plainTextASTToDeltaMatchers) {
        super();
        this.configs = configs;
        this.inlineDeltaMatchers = inlineDeltaMatchers;
        this.plainTextASTToDeltaMatchers = plainTextASTToDeltaMatchers;
    }
    astToDelta(ast) {
        const context = {
            configs: this.configs,
            options: Object.create(null),
            toDelta: (ast) => this.astToDelta(ast),
        };
        for (const matcher of this.plainTextASTToDeltaMatchers) {
            if (matcher.match(ast)) {
                return matcher.toDelta(ast, context);
            }
        }
        return [];
    }
    deltaToAST(deltas) {
        return deltas.map(delta => {
            const context = {
                configs: this.configs,
                current: { content: delta.insert },
            };
            for (const matcher of this.inlineDeltaMatchers) {
                if (matcher.match(delta)) {
                    context.current = matcher.toAST(delta, context);
                }
            }
            return context.current.content;
        });
    }
}
