// @blocksuite/affine-block-embed/dist/common/embed-block-element.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { CaptionedBlockComponent, SelectedStyle, } from '@blocksuite/affine-components/caption';
import { EMBED_CARD_HEIGHT, EMBED_CARD_MIN_WIDTH, EMBED_CARD_WIDTH, } from '@blocksuite/affine-shared/consts';
import { DocModeProvider, DragHandleConfigExtension, } from '@blocksuite/affine-shared/services';
import { captureEventTarget, convertDragPreviewDocToEdgeless, convertDragPreviewEdgelessToDoc, } from '@blocksuite/affine-shared/utils';
import { isGfxBlockComponent } from '@blocksuite/block-std';
import { html } from 'lit';
import { query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { styleMap } from 'lit/directives/style-map.js';
export const EmbedDragHandleOption = DragHandleConfigExtension({
    flavour: /affine:embed-*/,
    edgeless: true,
    onDragEnd: props => {
        const { state, draggingElements } = props;
        if (draggingElements.length !== 1 ||
            draggingElements[0].model.flavour.match(/affine:embed-*/) === null)
            return false;
        const blockComponent = draggingElements[0];
        const isInSurface = isGfxBlockComponent(blockComponent);
        const target = captureEventTarget(state.raw.target);
        const isTargetEdgelessContainer = target?.classList.contains('edgeless-container');
        if (isInSurface) {
            const style = blockComponent._cardStyle;
            const targetStyle = style === 'vertical' || style === 'cube' ? 'horizontal' : style;
            return convertDragPreviewEdgelessToDoc({
                blockComponent,
                style: targetStyle,
                ...props,
            });
        }
        else if (isTargetEdgelessContainer) {
            const style = blockComponent._cardStyle;
            return convertDragPreviewDocToEdgeless({
                blockComponent,
                cssSelector: '.embed-block-container',
                width: EMBED_CARD_WIDTH[style],
                height: EMBED_CARD_HEIGHT[style],
                ...props,
            });
        }
        return false;
    },
});
let EmbedBlockComponent = (() => {
    let _classSuper = CaptionedBlockComponent;
    let _embedBlock_decorators;
    let _embedBlock_initializers = [];
    let _embedBlock_extraInitializers = [];
    return class EmbedBlockComponent extends _classSuper {
        constructor() {
            super(...arguments);
            this._fetchAbortController = new AbortController();
            this._cardStyle = 'horizontal';
            /**
             * The actual rendered scale of the embed card.
             * By default, it is set to 1.
             */
            this._scale = 1;
            this.blockDraggable = true;
            /**
             * The style of the embed card.
             * You can use this to change the height and width of the card.
             * By default, the height and width are set to `_cardHeight` and `_cardWidth` respectively.
             */
            this.embedContainerStyle = {};
            this.renderEmbed = (content) => {
                if (this._cardStyle === 'horizontal' ||
                    this._cardStyle === 'horizontalThin' ||
                    this._cardStyle === 'list') {
                    this.style.display = 'block';
                    const mode = this.std.get(DocModeProvider).getEditorMode();
                    if (mode === 'edgeless') {
                        this.style.minWidth = `${EMBED_CARD_MIN_WIDTH}px`;
                    }
                }
                const selected = !!this.selected?.is('block');
                return html `
      <div
        draggable="${this.blockDraggable ? 'true' : 'false'}"
        class=${classMap({
                    'embed-block-container': true,
                    'selected-style': selected,
                })}
        style=${styleMap({
                    height: `${this._cardHeight}px`,
                    width: '100%',
                    ...this.embedContainerStyle,
                })}
      >
        ${content()}
      </div>
    `;
            };
            this.#blockContainerStyles_accessor_storage = {
                margin: '18px 0',
            };
            this.#embedBlock_accessor_storage = __runInitializers(this, _embedBlock_initializers, void 0);
            this.#selectedStyle_accessor_storage = (__runInitializers(this, _embedBlock_extraInitializers), SelectedStyle.Border);
            this.#useCaptionEditor_accessor_storage = true;
            this.#useZeroWidth_accessor_storage = true;
        }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _embedBlock_decorators = [query('.embed-block-container')];
            __esDecorate(this, null, _embedBlock_decorators, { kind: "accessor", name: "embedBlock", static: false, private: false, access: { has: obj => "embedBlock" in obj, get: obj => obj.embedBlock, set: (obj, value) => { obj.embedBlock = value; } }, metadata: _metadata }, _embedBlock_initializers, _embedBlock_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        /**
         * The height of the current embed card. Changes based on the card style.
         */
        get _cardHeight() {
            return EMBED_CARD_HEIGHT[this._cardStyle];
        }
        /**
         * The width of the current embed card. Changes based on the card style.
         */
        get _cardWidth() {
            return EMBED_CARD_WIDTH[this._cardStyle];
        }
        get fetchAbortController() {
            return this._fetchAbortController;
        }
        connectedCallback() {
            super.connectedCallback();
            if (this._fetchAbortController.signal.aborted)
                this._fetchAbortController = new AbortController();
            this.contentEditable = 'false';
        }
        disconnectedCallback() {
            super.disconnectedCallback();
            this._fetchAbortController.abort();
        }
        #blockContainerStyles_accessor_storage;
        get blockContainerStyles() { return this.#blockContainerStyles_accessor_storage; }
        set blockContainerStyles(value) { this.#blockContainerStyles_accessor_storage = value; }
        #embedBlock_accessor_storage;
        get embedBlock() { return this.#embedBlock_accessor_storage; }
        set embedBlock(value) { this.#embedBlock_accessor_storage = value; }
        #selectedStyle_accessor_storage;
        get selectedStyle() { return this.#selectedStyle_accessor_storage; }
        set selectedStyle(value) { this.#selectedStyle_accessor_storage = value; }
        #useCaptionEditor_accessor_storage;
        get useCaptionEditor() { return this.#useCaptionEditor_accessor_storage; }
        set useCaptionEditor(value) { this.#useCaptionEditor_accessor_storage = value; }
        #useZeroWidth_accessor_storage;
        get useZeroWidth() { return this.#useZeroWidth_accessor_storage; }
        set useZeroWidth(value) { this.#useZeroWidth_accessor_storage = value; }
    };
})();
export { EmbedBlockComponent };
