// @blocksuite/affine-block-embed/dist/common/to-edgeless-embed-block.js
import { blockComponentSymbol, GfxElementSymbol, toGfxBlockComponent, } from '@blocksuite/block-std';
import { Bound } from '@blocksuite/global/utils';
export function toEdgelessEmbedBlock(block) {
    var _a, _b;
    return class extends toGfxBlockComponent(block) {
        constructor() {
            super(...arguments);
            this._isDragging = false;
            this._isResizing = false;
            this._isSelected = false;
            this._showOverlay = false;
            this[_a] = true;
            this.blockDraggable = false;
            this.embedContainerStyle = {};
            this[_b] = true;
            this.#blockContainerStyles_accessor_storage = undefined;
        }
        static { _a = blockComponentSymbol, _b = GfxElementSymbol; }
        get bound() {
            return Bound.deserialize(this.model.xywh);
        }
        get rootService() {
            return this.std.getService('affine:page');
        }
        _handleClick(_) {
            return;
        }
        connectedCallback() {
            super.connectedCallback();
            const rootService = this.rootService;
            this._disposables.add(
            // @ts-expect-error TODO: fix after edgeless slots are migrated to extension
            rootService.slots.elementResizeStart.on(() => {
                this._isResizing = true;
                this._showOverlay = true;
            }));
            this._disposables.add(
            // @ts-expect-error TODO: fix after edgeless slots are migrated to extension
            rootService.slots.elementResizeEnd.on(() => {
                this._isResizing = false;
                this._showOverlay =
                    this._isResizing || this._isDragging || !this._isSelected;
            }));
        }
        renderGfxBlock() {
            const bound = Bound.deserialize(this.model.xywh);
            this.embedContainerStyle.width = `${bound.w}px`;
            this.embedContainerStyle.height = `${bound.h}px`;
            this.blockContainerStyles = {
                width: `${bound.w}px`,
            };
            this._scale = bound.w / this._cardWidth;
            return this.renderPageContent();
        }
        #blockContainerStyles_accessor_storage;
        get blockContainerStyles() { return this.#blockContainerStyles_accessor_storage; }
        set blockContainerStyles(value) { this.#blockContainerStyles_accessor_storage = value; }
    };
}
