// @blocksuite/data-view/dist/widget-presets/quick-setting-bar/filter/condition-view.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popFilterableSimpleMenu, popMenu, popupTargetFromElement, subMenuMiddleware, } from '@blocksuite/affine-components/context-menu';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher } from '@blocksuite/global/utils';
import { ArrowDownSmallIcon, ArrowRightSmallIcon, DeleteIcon, } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css, html } from 'lit';
import { property } from 'lit/decorators.js';
import { getRefType } from '../../../core/expression/ref/ref.js';
import { filterMatcher } from '../../../core/filter/filter-fn/matcher.js';
import { literalItemsMatcher } from '../../../core/filter/literal/index.js';
import { renderUniLit, t, typeSystem, } from '../../../core/index.js';
let FilterConditionView = (() => {
    let _classSuper = SignalWatcher(ShadowlessElement);
    let _value_decorators;
    let _value_initializers = [];
    let _value_extraInitializers = [];
    let _vars_decorators;
    let _vars_initializers = [];
    let _vars_extraInitializers = [];
    let _index_decorators;
    let _index_initializers = [];
    let _index_extraInitializers = [];
    let _onChange_decorators;
    let _onChange_initializers = [];
    let _onChange_extraInitializers = [];
    return class FilterConditionView extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _value_decorators = [property({ attribute: false })];
            _vars_decorators = [property({ attribute: false })];
            _index_decorators = [property({ attribute: false })];
            _onChange_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _value_decorators, { kind: "accessor", name: "value", static: false, private: false, access: { has: obj => "value" in obj, get: obj => obj.value, set: (obj, value) => { obj.value = value; } }, metadata: _metadata }, _value_initializers, _value_extraInitializers);
            __esDecorate(this, null, _vars_decorators, { kind: "accessor", name: "vars", static: false, private: false, access: { has: obj => "vars" in obj, get: obj => obj.vars, set: (obj, value) => { obj.vars = value; } }, metadata: _metadata }, _vars_initializers, _vars_extraInitializers);
            __esDecorate(this, null, _index_decorators, { kind: "accessor", name: "index", static: false, private: false, access: { has: obj => "index" in obj, get: obj => obj.index, set: (obj, value) => { obj.index = value; } }, metadata: _metadata }, _index_initializers, _index_extraInitializers);
            __esDecorate(this, null, _onChange_decorators, { kind: "accessor", name: "onChange", static: false, private: false, access: { has: obj => "onChange" in obj, get: obj => obj.onChange, set: (obj, value) => { obj.onChange = value; } }, metadata: _metadata }, _onChange_initializers, _onChange_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    filter-condition-view {
    }

    .filter-condition-expression {
      display: flex;
      align-items: center;
      gap: 4px;
    }

    .filter-condition-delete {
      border-radius: 4px;
      display: flex;
      align-items: center;
      justify-content: center;
      height: max-content;
      cursor: pointer;
    }

    .filter-condition-delete:hover {
      background-color: var(--affine-hover-color);
    }

    .filter-condition-delete svg {
      width: 16px;
      height: 16px;
    }

    .filter-condition-function-name {
      font-size: 12px;
      line-height: 20px;
      color: var(--affine-text-secondary-color);
      padding: 2px 8px;
      border-radius: 4px;
      cursor: pointer;
    }

    .filter-condition-function-name:hover {
      background-color: var(--affine-hover-color);
    }

    .filter-condition-arg {
      font-size: 12px;
      font-style: normal;
      font-weight: 600;
      padding: 0 4px;
      height: 100%;
      display: flex;
      align-items: center;
    }
  `; }
        #value_accessor_storage;
        get value() { return this.#value_accessor_storage; }
        set value(value) { this.#value_accessor_storage = value; }
        #vars_accessor_storage;
        get vars() { return this.#vars_accessor_storage; }
        set vars(value) { this.#vars_accessor_storage = value; }
        getArgItems(argType, index) {
            return literalItemsMatcher.getItems(argType, computed(() => {
                return this.filter$.value?.args[index]?.value;
            }), value => {
                const filter = this.filter$.value;
                if (!filter) {
                    return;
                }
                const args = filter.args.slice();
                args[index] = { type: 'literal', value };
                this.setFilter({
                    ...filter,
                    args: args,
                });
            });
        }
        getArgsItems() {
            return (this.fnType$.value?.args
                .slice(1)
                .flatMap((arg, i) => this.getArgItems(arg, i)) ?? []);
        }
        render() {
            const leftVar = this.leftVar$.value;
            if (!leftVar) {
                return html ` <data-view-component-button
        hoverType="border"
        .text="${html `Invalid filter rule`}"
      ></data-view-component-button>`;
            }
            return html `
      <data-view-component-button
        hoverType="border"
        .icon="${renderUniLit(leftVar.icon)}"
        @click="${this.onClickButton}"
        .text="${html `<span
          style="overflow: hidden;max-width: 230px;text-overflow: ellipsis"
          >${this.text$.value}</span
        >`}"
        .postfix="${ArrowDownSmallIcon()}"
      ></data-view-component-button>
    `;
        }
        #index_accessor_storage;
        get index() { return this.#index_accessor_storage; }
        set index(value) { this.#index_accessor_storage = value; }
        #onChange_accessor_storage;
        get onChange() { return this.#onChange_accessor_storage; }
        set onChange(value) { this.#onChange_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.onClickButton = (evt) => {
                this.popConditionEdit(popupTargetFromElement(evt.currentTarget));
            };
            this.popConditionEdit = (target) => {
                const type = this.leftVar$.value?.type;
                if (!type) {
                    return;
                }
                const fn = this.fnConfig$.value;
                if (!fn) {
                    popFilterableSimpleMenu(target, this.getFunctionItems(target));
                    return;
                }
                const handler = popMenu(target, {
                    options: {
                        items: [
                            menu.group({
                                items: [
                                    menu.action({
                                        name: fn.label,
                                        postfix: ArrowRightSmallIcon(),
                                        select: ele => {
                                            popMenu(popupTargetFromElement(ele), {
                                                options: {
                                                    items: [
                                                        menu.group({
                                                            items: this.getFunctionItems(target, () => {
                                                                handler.close();
                                                            }),
                                                        }),
                                                    ],
                                                },
                                                middleware: subMenuMiddleware,
                                            });
                                            return false;
                                        },
                                    }),
                                ],
                            }),
                            menu.dynamic(() => this.getArgsItems()),
                            menu.group({
                                items: [
                                    menu.action({
                                        name: 'Delete',
                                        class: { 'delete-item': true },
                                        prefix: DeleteIcon(),
                                        select: () => {
                                            const list = this.value.value.slice();
                                            list.splice(this.index, 1);
                                            this.onChange(list);
                                        },
                                    }),
                                ],
                            }),
                        ],
                    },
                });
            };
            this.#value_accessor_storage = __runInitializers(this, _value_initializers, void 0);
            this.filter$ = (__runInitializers(this, _value_extraInitializers), computed(() => {
                const filter = this.value.value[this.index];
                if (!filter || filter.type !== 'filter') {
                    return;
                }
                return filter;
            }));
            this.args$ = computed(() => {
                return this.filter$.value?.args.map(v => v.value);
            });
            this.fnConfig$ = computed(() => {
                return filterMatcher.getFilterByName(this.filter$.value?.function);
            });
            this.#vars_accessor_storage = __runInitializers(this, _vars_initializers, void 0);
            this.fnType$ = (__runInitializers(this, _vars_extraInitializers), computed(() => {
                const fnConfig = this.fnConfig$.value;
                const filter = this.filter$.value;
                if (!fnConfig || !filter) {
                    return;
                }
                const refType = getRefType(this.vars.value, filter.left);
                if (!refType) {
                    return;
                }
                const fnTemplate = t.fn.instance([fnConfig.self, ...fnConfig.args], t.boolean.instance(), fnConfig.vars);
                return typeSystem.instanceFn(fnTemplate, [refType], t.boolean.instance(), {});
            }));
            this.getFunctionItems = (target, onSelect) => {
                const filter = this.filter$.value;
                if (!filter) {
                    return [];
                }
                const type = getRefType(this.vars.value, filter?.left);
                if (!type) {
                    return [];
                }
                return filterMatcher.filterListBySelfType(type).map(v => {
                    const selected = v.name === filter.function;
                    return menu.action({
                        name: v.label,
                        isSelected: selected,
                        select: () => {
                            this.setFilter({
                                ...filter,
                                function: v.name,
                            });
                            onSelect?.();
                            this.popConditionEdit(target);
                        },
                    });
                });
            };
            this.leftVar$ = computed(() => {
                return this.vars.value.find(v => v.id === this.filter$.value?.left.name);
            });
            this.setFilter = (filter) => {
                const list = this.value.value.slice();
                list[this.index] = filter;
                this.onChange(list);
            };
            this.text$ = computed(() => {
                const name = this.leftVar$.value?.name ?? '';
                const data = this.fnConfig$.value;
                const type = this.fnType$.value;
                const argValues = this.args$.value;
                if (!type || !argValues || !data) {
                    return;
                }
                const argDataList = argValues.map((v, i) => v ? { value: v, type: type.args[i + 1] } : undefined);
                const valueString = data.shortString?.(...argDataList) ?? '';
                if (valueString) {
                    return `${name}${valueString}`;
                }
                return name;
            });
            this.#index_accessor_storage = __runInitializers(this, _index_initializers, void 0);
            this.#onChange_accessor_storage = (__runInitializers(this, _index_extraInitializers), __runInitializers(this, _onChange_initializers, void 0));
            __runInitializers(this, _onChange_extraInitializers);
        }
    };
})();
export { FilterConditionView };
