// @blocksuite/data-view/dist/widget-presets/quick-setting-bar/filter/list-view.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher } from '@blocksuite/global/utils';
import { ArrowDownSmallIcon, FilterIcon, PlusIcon, } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css, html } from 'lit';
import { property } from 'lit/decorators.js';
import { popCreateFilter } from '../../../core/index.js';
import { popFilterGroup } from './group-panel-view.js';
let FilterBar = (() => {
    let _classSuper = SignalWatcher(ShadowlessElement);
    let _filterGroup_decorators;
    let _filterGroup_initializers = [];
    let _filterGroup_extraInitializers = [];
    let _onChange_decorators;
    let _onChange_initializers = [];
    let _onChange_extraInitializers = [];
    let _vars_decorators;
    let _vars_initializers = [];
    let _vars_extraInitializers = [];
    return class FilterBar extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _filterGroup_decorators = [property({ attribute: false })];
            _onChange_decorators = [property({ attribute: false })];
            _vars_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _filterGroup_decorators, { kind: "accessor", name: "filterGroup", static: false, private: false, access: { has: obj => "filterGroup" in obj, get: obj => obj.filterGroup, set: (obj, value) => { obj.filterGroup = value; } }, metadata: _metadata }, _filterGroup_initializers, _filterGroup_extraInitializers);
            __esDecorate(this, null, _onChange_decorators, { kind: "accessor", name: "onChange", static: false, private: false, access: { has: obj => "onChange" in obj, get: obj => obj.onChange, set: (obj, value) => { obj.onChange = value; } }, metadata: _metadata }, _onChange_initializers, _onChange_extraInitializers);
            __esDecorate(this, null, _vars_decorators, { kind: "accessor", name: "vars", static: false, private: false, access: { has: obj => "vars" in obj, get: obj => obj.vars, set: (obj, value) => { obj.vars = value; } }, metadata: _metadata }, _vars_initializers, _vars_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    filter-bar {
      display: flex;
      gap: 8px;
      overflow-x: scroll;
      margin-bottom: -10px;
      padding-bottom: 2px;
      align-items: center;
    }

    .filter-group-tag {
      font-size: 12px;
      font-style: normal;
      font-weight: 600;
      line-height: 20px;
      display: flex;
      align-items: center;
      padding: 4px;
      background-color: var(--affine-white);
    }

    .filter-bar-add-filter {
      white-space: nowrap;
      color: var(--affine-text-secondary-color);
      padding: 4px 8px;
      display: flex;
      align-items: center;
      gap: 6px;
      font-size: 14px;
      font-style: normal;
      font-weight: 400;
      line-height: 22px;
    }

    filter-bar::-webkit-scrollbar {
      -webkit-appearance: none;
      display: block;
    }

    filter-bar::-webkit-scrollbar-thumb {
      border-radius: 2px;
      background-color: transparent;
    }

    filter-bar::-webkit-scrollbar:horizontal {
      height: 8px;
    }

    filter-bar:hover::-webkit-scrollbar-thumb {
      border-radius: 16px;
      background-color: var(--affine-black-30);
    }

    filter-bar:hover::-webkit-scrollbar-track {
      //background-color: var(--affine-hover-color);
    }
  `; }
        #filterGroup_accessor_storage;
        get filterGroup() { return this.#filterGroup_accessor_storage; }
        set filterGroup(value) { this.#filterGroup_accessor_storage = value; }
        deleteFilter(i) {
            this.onChange({
                ...this.filterGroup.value,
                conditions: this.filterGroup.value.conditions.filter((_, index) => index !== i),
            });
        }
        render() {
            return html ` ${this.renderFilters()} ${this.renderAddFilter()} `;
        }
        renderCondition(i) {
            const condition = this.conditions$.value[i];
            if (!condition) {
                return;
            }
            if (condition.type === 'filter') {
                return html ` <filter-condition-view
        .vars="${this.vars}"
        .index="${i}"
        .value="${this.conditions$}"
        .onChange="${this.setConditions}"
      ></filter-condition-view>`;
            }
            const expandGroup = (e) => {
                this.expandGroup(popupTargetFromElement(e.currentTarget), i);
            };
            const length = condition.conditions.length;
            const text = length > 1 ? `${length} rules` : `${length} rule`;
            return html ` <data-view-component-button
      hoverType="border"
      .icon="${FilterIcon()}"
      @click="${expandGroup}"
      .text="${html `${text}`}"
      .postfix="${ArrowDownSmallIcon()}"
    ></data-view-component-button>`;
        }
        renderFilters() {
            return this.filterGroup.value.conditions.map((_, i) => this.renderCondition(i));
        }
        updated() {
            this.updateMoreFilterPanel?.();
        }
        #onChange_accessor_storage;
        get onChange() { return this.#onChange_accessor_storage; }
        set onChange(value) { this.#onChange_accessor_storage = value; }
        #vars_accessor_storage;
        get vars() { return this.#vars_accessor_storage; }
        set vars(value) { this.#vars_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._setFilter = (index, filter) => {
                this.onChange({
                    ...this.filterGroup.value,
                    conditions: this.filterGroup.value.conditions.map((v, i) => index === i ? filter : v),
                });
            };
            this.addFilter = (e) => {
                const element = popupTargetFromElement(e.target);
                popCreateFilter(element, {
                    vars: this.vars,
                    onSelect: filter => {
                        const index = this.filterGroup.value.conditions.length;
                        this.onChange({
                            ...this.filterGroup.value,
                            conditions: [...this.filterGroup.value.conditions, filter],
                        });
                        requestAnimationFrame(() => {
                            this.expandGroup(element, index);
                        });
                    },
                });
            };
            this.expandGroup = (position, i) => {
                if (this.filterGroup.value.conditions[i]?.type !== 'group') {
                    return;
                }
                popFilterGroup(position, {
                    vars: this.vars,
                    value$: computed(() => {
                        return this.filterGroup.value.conditions[i];
                    }),
                    onChange: filter => {
                        if (filter) {
                            this._setFilter(i, filter);
                        }
                        else {
                            this.deleteFilter(i);
                        }
                    },
                });
            };
            this.#filterGroup_accessor_storage = __runInitializers(this, _filterGroup_initializers, void 0);
            this.conditions$ = (__runInitializers(this, _filterGroup_extraInitializers), computed(() => {
                return this.filterGroup.value.conditions;
            }));
            this.renderAddFilter = () => {
                return html ` <div
      style="height: 100%;"
      class="filter-bar-add-filter dv-icon-16 dv-round-4 dv-hover"
      @click="${this.addFilter}"
    >
      ${PlusIcon()} Add filter
    </div>`;
            };
            this.setConditions = (conditions) => {
                this.onChange({
                    ...this.filterGroup.value,
                    conditions: conditions,
                });
            };
            this.#onChange_accessor_storage = __runInitializers(this, _onChange_initializers, void 0);
            this.#vars_accessor_storage = (__runInitializers(this, _onChange_extraInitializers), __runInitializers(this, _vars_initializers, void 0));
            __runInitializers(this, _vars_extraInitializers);
        }
    };
})();
export { FilterBar };
