// @blocksuite/inline/dist/components/v-element.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { DisposableGroup, SignalWatcher } from '@blocksuite/global/utils';
import { effect, signal } from '@preact/signals-core';
import { html, LitElement } from 'lit';
import { property } from 'lit/decorators.js';
import { styleMap } from 'lit/directives/style-map.js';
import { ZERO_WIDTH_SPACE } from '../consts.js';
import { isInlineRangeIntersect } from '../utils/inline-range.js';
let VElement = (() => {
    let _classSuper = SignalWatcher(LitElement);
    let _delta_decorators;
    let _delta_initializers = [];
    let _delta_extraInitializers = [];
    let _endOffset_decorators;
    let _endOffset_initializers = [];
    let _endOffset_extraInitializers = [];
    let _inlineEditor_decorators;
    let _inlineEditor_initializers = [];
    let _inlineEditor_extraInitializers = [];
    let _lineIndex_decorators;
    let _lineIndex_initializers = [];
    let _lineIndex_extraInitializers = [];
    let _startOffset_decorators;
    let _startOffset_initializers = [];
    let _startOffset_extraInitializers = [];
    return class VElement extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _delta_decorators = [property({ type: Object })];
            _endOffset_decorators = [property({ attribute: false })];
            _inlineEditor_decorators = [property({ attribute: false })];
            _lineIndex_decorators = [property({ attribute: false })];
            _startOffset_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _delta_decorators, { kind: "accessor", name: "delta", static: false, private: false, access: { has: obj => "delta" in obj, get: obj => obj.delta, set: (obj, value) => { obj.delta = value; } }, metadata: _metadata }, _delta_initializers, _delta_extraInitializers);
            __esDecorate(this, null, _endOffset_decorators, { kind: "accessor", name: "endOffset", static: false, private: false, access: { has: obj => "endOffset" in obj, get: obj => obj.endOffset, set: (obj, value) => { obj.endOffset = value; } }, metadata: _metadata }, _endOffset_initializers, _endOffset_extraInitializers);
            __esDecorate(this, null, _inlineEditor_decorators, { kind: "accessor", name: "inlineEditor", static: false, private: false, access: { has: obj => "inlineEditor" in obj, get: obj => obj.inlineEditor, set: (obj, value) => { obj.inlineEditor = value; } }, metadata: _metadata }, _inlineEditor_initializers, _inlineEditor_extraInitializers);
            __esDecorate(this, null, _lineIndex_decorators, { kind: "accessor", name: "lineIndex", static: false, private: false, access: { has: obj => "lineIndex" in obj, get: obj => obj.lineIndex, set: (obj, value) => { obj.lineIndex = value; } }, metadata: _metadata }, _lineIndex_initializers, _lineIndex_extraInitializers);
            __esDecorate(this, null, _startOffset_decorators, { kind: "accessor", name: "startOffset", static: false, private: false, access: { has: obj => "startOffset" in obj, get: obj => obj.startOffset, set: (obj, value) => { obj.startOffset = value; } }, metadata: _metadata }, _startOffset_initializers, _startOffset_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        connectedCallback() {
            super.connectedCallback();
            this.disposables.add(effect(() => {
                const inlineRange = this.inlineEditor.inlineRange$.value;
                this.selected.value =
                    !!inlineRange &&
                        isInlineRangeIntersect(inlineRange, {
                            index: this.startOffset,
                            length: this.endOffset - this.startOffset,
                        });
            }));
        }
        createRenderRoot() {
            return this;
        }
        async getUpdateComplete() {
            const result = await super.getUpdateComplete();
            const span = this.querySelector('[data-v-element="true"]');
            const el = span.firstElementChild;
            await el.updateComplete;
            const vTexts = Array.from(this.querySelectorAll('v-text'));
            await Promise.all(vTexts.map(vText => vText.updateComplete));
            return result;
        }
        render() {
            const inlineEditor = this.inlineEditor;
            const attributeRenderer = inlineEditor.attributeService.attributeRenderer;
            const renderProps = {
                delta: this.delta,
                selected: this.selected.value,
                startOffset: this.startOffset,
                endOffset: this.endOffset,
                lineIndex: this.lineIndex,
                editor: inlineEditor,
            };
            const isEmbed = inlineEditor.isEmbed(this.delta);
            if (isEmbed) {
                if (this.delta.insert.length !== 1) {
                    throw new BlockSuiteError(ErrorCode.InlineEditorError, `The length of embed node should only be 1.
          This seems to be an internal issue with inline editor.
          Please go to https://github.com/toeverything/blocksuite/issues
          to report it.`);
                }
                return html `<span
        data-v-embed="true"
        data-v-element="true"
        contenteditable="false"
        style=${styleMap({ userSelect: 'none' })}
        >${attributeRenderer(renderProps)}</span
      >`;
            }
            // we need to avoid \n appearing before and after the span element, which will
            // cause the unexpected space
            return html `<span data-v-element="true"
      >${attributeRenderer(renderProps)}</span
    >`;
        }
        #delta_accessor_storage;
        get delta() { return this.#delta_accessor_storage; }
        set delta(value) { this.#delta_accessor_storage = value; }
        #endOffset_accessor_storage;
        get endOffset() { return this.#endOffset_accessor_storage; }
        set endOffset(value) { this.#endOffset_accessor_storage = value; }
        #inlineEditor_accessor_storage;
        get inlineEditor() { return this.#inlineEditor_accessor_storage; }
        set inlineEditor(value) { this.#inlineEditor_accessor_storage = value; }
        #lineIndex_accessor_storage;
        get lineIndex() { return this.#lineIndex_accessor_storage; }
        set lineIndex(value) { this.#lineIndex_accessor_storage = value; }
        #startOffset_accessor_storage;
        get startOffset() { return this.#startOffset_accessor_storage; }
        set startOffset(value) { this.#startOffset_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.disposables = new DisposableGroup();
            this.selected = signal(false);
            this.#delta_accessor_storage = __runInitializers(this, _delta_initializers, {
                insert: ZERO_WIDTH_SPACE,
            });
            this.#endOffset_accessor_storage = (__runInitializers(this, _delta_extraInitializers), __runInitializers(this, _endOffset_initializers, void 0));
            this.#inlineEditor_accessor_storage = (__runInitializers(this, _endOffset_extraInitializers), __runInitializers(this, _inlineEditor_initializers, void 0));
            this.#lineIndex_accessor_storage = (__runInitializers(this, _inlineEditor_extraInitializers), __runInitializers(this, _lineIndex_initializers, void 0));
            this.#startOffset_accessor_storage = (__runInitializers(this, _lineIndex_extraInitializers), __runInitializers(this, _startOffset_initializers, void 0));
            __runInitializers(this, _startOffset_extraInitializers);
        }
    };
})();
export { VElement };
