// @blocksuite/inline/dist/components/v-line.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { assertExists } from '@blocksuite/global/utils';
import { html, LitElement } from 'lit';
import { property } from 'lit/decorators.js';
import { styleMap } from 'lit/directives/style-map.js';
import { INLINE_ROOT_ATTR, ZERO_WIDTH_SPACE } from '../consts.js';
import { EmbedGap } from './embed-gap.js';
let VLine = (() => {
    let _classSuper = LitElement;
    let _elements_decorators;
    let _elements_initializers = [];
    let _elements_extraInitializers = [];
    let _endOffset_decorators;
    let _endOffset_initializers = [];
    let _endOffset_extraInitializers = [];
    let _index_decorators;
    let _index_initializers = [];
    let _index_extraInitializers = [];
    let _startOffset_decorators;
    let _startOffset_initializers = [];
    let _startOffset_extraInitializers = [];
    return class VLine extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _elements_decorators = [property({ attribute: false })];
            _endOffset_decorators = [property({ attribute: false })];
            _index_decorators = [property({ attribute: false })];
            _startOffset_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _elements_decorators, { kind: "accessor", name: "elements", static: false, private: false, access: { has: obj => "elements" in obj, get: obj => obj.elements, set: (obj, value) => { obj.elements = value; } }, metadata: _metadata }, _elements_initializers, _elements_extraInitializers);
            __esDecorate(this, null, _endOffset_decorators, { kind: "accessor", name: "endOffset", static: false, private: false, access: { has: obj => "endOffset" in obj, get: obj => obj.endOffset, set: (obj, value) => { obj.endOffset = value; } }, metadata: _metadata }, _endOffset_initializers, _endOffset_extraInitializers);
            __esDecorate(this, null, _index_decorators, { kind: "accessor", name: "index", static: false, private: false, access: { has: obj => "index" in obj, get: obj => obj.index, set: (obj, value) => { obj.index = value; } }, metadata: _metadata }, _index_initializers, _index_extraInitializers);
            __esDecorate(this, null, _startOffset_decorators, { kind: "accessor", name: "startOffset", static: false, private: false, access: { has: obj => "startOffset" in obj, get: obj => obj.startOffset, set: (obj, value) => { obj.startOffset = value; } }, metadata: _metadata }, _startOffset_initializers, _startOffset_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        get inlineEditor() {
            const rootElement = this.closest(`[${INLINE_ROOT_ATTR}]`);
            assertExists(rootElement, 'v-line must be inside a v-root');
            const inlineEditor = rootElement.inlineEditor;
            assertExists(inlineEditor, 'v-line must be inside a v-root with inline-editor');
            return inlineEditor;
        }
        get vElements() {
            return Array.from(this.querySelectorAll('v-element'));
        }
        get vTextContent() {
            return this.vElements.reduce((acc, el) => acc + el.delta.insert, '');
        }
        get vTextLength() {
            return this.vElements.reduce((acc, el) => acc + el.delta.insert.length, 0);
        }
        // you should use vElements.length or vTextLength because v-element corresponds to the actual delta
        get vTexts() {
            return Array.from(this.querySelectorAll('v-text'));
        }
        createRenderRoot() {
            return this;
        }
        firstUpdated() {
            this.style.display = 'block';
            this.addEventListener('mousedown', e => {
                if (e.detail >= 2 && this.startOffset === this.endOffset) {
                    e.preventDefault();
                    return;
                }
                if (e.detail >= 3) {
                    e.preventDefault();
                    this.inlineEditor.setInlineRange({
                        index: this.startOffset,
                        length: this.endOffset - this.startOffset,
                    });
                }
            });
        }
        // vTexts.length > 0 does not mean the line is not empty,
        async getUpdateComplete() {
            const result = await super.getUpdateComplete();
            await Promise.all(this.vElements.map(el => el.updateComplete));
            return result;
        }
        render() {
            if (!this.isConnected)
                return;
            if (this.inlineEditor.vLineRenderer) {
                return this.inlineEditor.vLineRenderer(this);
            }
            return this.renderVElements();
        }
        renderVElements() {
            if (this.elements.length === 0) {
                // don't use v-element because it not correspond to the actual delta
                return html `<div><v-text .str=${ZERO_WIDTH_SPACE}></v-text></div>`;
            }
            const inlineEditor = this.inlineEditor;
            const renderElements = this.elements.flatMap(([template, delta], index) => {
                if (inlineEditor.isEmbed(delta)) {
                    if (delta.insert.length !== 1) {
                        throw new BlockSuiteError(ErrorCode.InlineEditorError, `The length of embed node should only be 1.
            This seems to be an internal issue with inline editor.
            Please go to https://github.com/toeverything/blocksuite/issues
            to report it.`);
                    }
                    // we add `EmbedGap` to make cursor can be placed between embed elements
                    if (index === 0) {
                        const nextDelta = this.elements[index + 1]?.[1];
                        if (!nextDelta || inlineEditor.isEmbed(nextDelta)) {
                            return [EmbedGap, template, EmbedGap];
                        }
                        else {
                            return [EmbedGap, template];
                        }
                    }
                    else {
                        const nextDelta = this.elements[index + 1]?.[1];
                        if (!nextDelta || inlineEditor.isEmbed(nextDelta)) {
                            return [template, EmbedGap];
                        }
                        else {
                            return [template];
                        }
                    }
                }
                return template;
            });
            // prettier will generate \n and cause unexpected space and line break
            // prettier-ignore
            return html `<div style=${styleMap({
                // this padding is used to make cursor can be placed at the
                // start and end of the line when the first and last element is embed element
                padding: '0 0.5px',
                display: 'inline-block',
            })}>${renderElements}</div>`;
        }
        #elements_accessor_storage = __runInitializers(this, _elements_initializers, []);
        get elements() { return this.#elements_accessor_storage; }
        set elements(value) { this.#elements_accessor_storage = value; }
        #endOffset_accessor_storage = (__runInitializers(this, _elements_extraInitializers), __runInitializers(this, _endOffset_initializers, void 0));
        get endOffset() { return this.#endOffset_accessor_storage; }
        set endOffset(value) { this.#endOffset_accessor_storage = value; }
        #index_accessor_storage = (__runInitializers(this, _endOffset_extraInitializers), __runInitializers(this, _index_initializers, void 0));
        get index() { return this.#index_accessor_storage; }
        set index(value) { this.#index_accessor_storage = value; }
        #startOffset_accessor_storage = (__runInitializers(this, _index_extraInitializers), __runInitializers(this, _startOffset_initializers, void 0));
        get startOffset() { return this.#startOffset_accessor_storage; }
        set startOffset(value) { this.#startOffset_accessor_storage = value; }
        constructor() {
            super(...arguments);
            __runInitializers(this, _startOffset_extraInitializers);
        }
    };
})();
export { VLine };
