// @blocksuite/block-std/dist/gfx/tool/tool.js
import { createIdentifier } from '@blocksuite/global/di';
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { DisposableGroup } from '@blocksuite/global/utils';
import { Extension } from '../../extension/extension.js';
import { GfxControllerIdentifier } from '../identifiers.js';
export class BaseTool extends Extension {
    static { this.toolName = ''; }
    get active() {
        return this.gfx.tool.currentTool$.peek() === this;
    }
    get allowDragWithRightButton() {
        return false;
    }
    get controller() {
        return this.gfx.tool;
    }
    get doc() {
        return this.gfx.doc;
    }
    get std() {
        return this.gfx.std;
    }
    get toolName() {
        return this.constructor.toolName;
    }
    constructor(gfx) {
        super();
        this.gfx = gfx;
        this.activatedOption = {};
        this.addHook = (evtName, handler) => {
            this.eventTarget.addHook(evtName, handler);
        };
        /**
         * The `disposable` will be disposed when the tool is unloaded.
         */
        this.disposable = new DisposableGroup();
    }
    static setup(di) {
        if (!this.toolName) {
            throw new BlockSuiteError(ErrorCode.ValueNotExists, `The tool constructor '${this.name}' should have a static 'toolName' property.`);
        }
        di.addImpl(ToolIdentifier(this.toolName), this, [GfxControllerIdentifier]);
    }
    /**
     * Called when the tool is activated.
     * @param _ - The data passed as second argument when calling `ToolController.use`.
     */
    activate(_) { }
    click(_) { }
    contextMenu(_) { }
    /**
     * Called when the tool is deactivated.
     */
    deactivate() { }
    doubleClick(_) { }
    dragEnd(_) { }
    dragMove(_) { }
    dragStart(_) { }
    /**
     * Called when the tool is registered.
     */
    mounted() { }
    pointerDown(_) { }
    pointerMove(_) { }
    pointerOut(_) { }
    pointerUp(_) { }
    tripleClick(_) { }
    /**
     * Called when the tool is unloaded, usually when the whole `ToolController` is destroyed.
     */
    unmounted() { }
}
export const ToolIdentifier = createIdentifier('GfxTool');
