// @blocksuite/presets/dist/fragments/frame-panel/header/frame-panel-header.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { createButtonPopper, DocModeProvider, EdgelessRootService, EditPropsStore, } from '@blocksuite/blocks';
import { DisposableGroup, WithDisposable } from '@blocksuite/global/utils';
import { css, html, LitElement } from 'lit';
import { property, query, state } from 'lit/decorators.js';
import { SettingsIcon, SmallFrameNavigatorIcon } from '../../_common/icons.js';
const styles = css `
  :host {
    display: flex;
    width: 100%;
    justify-content: start;
  }

  .frame-panel-header {
    display: flex;
    width: 100%;
    height: 36px;
    align-items: center;
    justify-content: space-between;
    box-sizing: border-box;
    padding: 0 8px;
  }

  .all-frames-setting {
    display: flex;
    align-items: center;
    gap: 8px;
    width: 100px;
    height: 24px;
    margin: 8px 4px;
  }

  .all-frames-setting-button svg {
    color: var(--affine-icon-secondary);
  }

  .all-frames-setting-button:hover svg,
  .all-frames-setting-button.active svg {
    color: var(--affine-icon-color);
  }

  .all-frames-setting-label {
    width: 68px;
    height: 22px;
    font-size: var(--affine-font-sm);
    font-weight: 500;
    line-height: 22px;
    color: var(--light-text-color-text-secondary-color, #8e8d91);
  }

  .frames-setting-container {
    display: none;
    justify-content: center;
    align-items: center;
    background: var(--affine-background-overlay-panel-color);
    box-shadow: var(--affine-shadow-2);
    border-radius: 8px;
  }

  .frames-setting-container[data-show] {
    display: flex;
  }

  .presentation-button {
    display: flex;
    align-items: center;
    gap: 4px;
    box-sizing: border-box;
    width: 117px;
    height: 28px;
    padding: 4px 8px;
    border-radius: 8px;
    margin: 4px 0;
    border: 1px solid var(--affine-border-color);
    background: var(--affine-white);
  }

  .presentation-button:hover {
    background: var(--affine-hover-color);
    cursor: pointer;
  }

  .presentation-button svg {
    fill: var(--affine-icon-color);
    margin-right: 4px;
  }

  .presentation-button-label {
    font-size: 12px;
    font-weight: 500;
    line-height: 20px;
  }
`;
export const AFFINE_FRAME_PANEL_HEADER = 'affine-frame-panel-header';
let FramePanelHeader = (() => {
    let _classSuper = WithDisposable(LitElement);
    let __frameSettingButton_decorators;
    let __frameSettingButton_initializers = [];
    let __frameSettingButton_extraInitializers = [];
    let __frameSettingMenu_decorators;
    let __frameSettingMenu_initializers = [];
    let __frameSettingMenu_extraInitializers = [];
    let __settingPopperShow_decorators;
    let __settingPopperShow_initializers = [];
    let __settingPopperShow_extraInitializers = [];
    let _editorHost_decorators;
    let _editorHost_initializers = [];
    let _editorHost_extraInitializers = [];
    return class FramePanelHeader extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            __frameSettingButton_decorators = [query('.all-frames-setting-button')];
            __frameSettingMenu_decorators = [query('.frames-setting-container')];
            __settingPopperShow_decorators = [state()];
            _editorHost_decorators = [property({ attribute: false })];
            __esDecorate(this, null, __frameSettingButton_decorators, { kind: "accessor", name: "_frameSettingButton", static: false, private: false, access: { has: obj => "_frameSettingButton" in obj, get: obj => obj._frameSettingButton, set: (obj, value) => { obj._frameSettingButton = value; } }, metadata: _metadata }, __frameSettingButton_initializers, __frameSettingButton_extraInitializers);
            __esDecorate(this, null, __frameSettingMenu_decorators, { kind: "accessor", name: "_frameSettingMenu", static: false, private: false, access: { has: obj => "_frameSettingMenu" in obj, get: obj => obj._frameSettingMenu, set: (obj, value) => { obj._frameSettingMenu = value; } }, metadata: _metadata }, __frameSettingMenu_initializers, __frameSettingMenu_extraInitializers);
            __esDecorate(this, null, __settingPopperShow_decorators, { kind: "accessor", name: "_settingPopperShow", static: false, private: false, access: { has: obj => "_settingPopperShow" in obj, get: obj => obj._settingPopperShow, set: (obj, value) => { obj._settingPopperShow = value; } }, metadata: _metadata }, __settingPopperShow_initializers, __settingPopperShow_extraInitializers);
            __esDecorate(this, null, _editorHost_decorators, { kind: "accessor", name: "editorHost", static: false, private: false, access: { has: obj => "editorHost" in obj, get: obj => obj.editorHost, set: (obj, value) => { obj.editorHost = value; } }, metadata: _metadata }, _editorHost_initializers, _editorHost_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = styles; }
        get _edgelessRootService() {
            return this.editorHost.std.getOptional(EdgelessRootService);
        }
        _tryLoadNavigatorStateLocalRecord() {
            this._navigatorMode = this.editorHost.std
                .get(EditPropsStore)
                .getStorage('presentFillScreen')
                ? 'fill'
                : 'fit';
        }
        connectedCallback() {
            super.connectedCallback();
            this._tryLoadNavigatorStateLocalRecord();
        }
        disconnectedCallback() {
            super.disconnectedCallback();
            if (this._edgelessDisposables) {
                this._clearEdgelessDisposables();
            }
        }
        firstUpdated() {
            const disposables = this.disposables;
            this._framesSettingMenuPopper = createButtonPopper(this._frameSettingButton, this._frameSettingMenu, ({ display }) => {
                this._settingPopperShow = display === 'show';
            }, {
                mainAxis: 14,
                crossAxis: -100,
            });
            disposables.add(this._framesSettingMenuPopper);
        }
        render() {
            return html `<div class="frame-panel-header">
      <div class="all-frames-setting">
        <span class="all-frames-setting-label">All frames</span>
        <edgeless-tool-icon-button
          class="all-frames-setting-button ${this._settingPopperShow
                ? 'active'
                : ''}"
          .tooltip=${this._settingPopperShow ? '' : 'All Frames Settings'}
          .tipPosition=${'top'}
          .active=${this._settingPopperShow}
          .activeMode=${'background'}
          @click=${() => this._framesSettingMenuPopper?.toggle()}
        >
          ${SettingsIcon}
        </edgeless-tool-icon-button>
      </div>
      <div class="frames-setting-container">
        <affine-frames-setting-menu
          .editorHost=${this.editorHost}
        ></affine-frames-setting-menu>
      </div>
      <div class="presentation-button" @click=${this._enterPresentationMode}>
        ${SmallFrameNavigatorIcon}<span class="presentation-button-label"
          >Presentation</span
        >
      </div>
    </div>`;
        }
        updated(_changedProperties) {
            if (_changedProperties.has('editorHost')) {
                if (this._edgelessRootService) {
                    this._setEdgelessDisposables();
                }
                else {
                    this._clearEdgelessDisposables();
                }
            }
        }
        #_frameSettingButton_accessor_storage;
        get _frameSettingButton() { return this.#_frameSettingButton_accessor_storage; }
        set _frameSettingButton(value) { this.#_frameSettingButton_accessor_storage = value; }
        #_frameSettingMenu_accessor_storage;
        get _frameSettingMenu() { return this.#_frameSettingMenu_accessor_storage; }
        set _frameSettingMenu(value) { this.#_frameSettingMenu_accessor_storage = value; }
        #_settingPopperShow_accessor_storage;
        get _settingPopperShow() { return this.#_settingPopperShow_accessor_storage; }
        set _settingPopperShow(value) { this.#_settingPopperShow_accessor_storage = value; }
        #editorHost_accessor_storage;
        get editorHost() { return this.#editorHost_accessor_storage; }
        set editorHost(value) { this.#editorHost_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._clearEdgelessDisposables = () => {
                this._edgelessDisposables?.dispose();
                this._edgelessDisposables = null;
            };
            this._edgelessDisposables = null;
            this._enterPresentationMode = () => {
                if (!this._edgelessRootService) {
                    this.editorHost.std.get(DocModeProvider).setEditorMode('edgeless');
                }
                setTimeout(() => {
                    this._edgelessRootService?.gfx.tool.setTool({
                        type: 'frameNavigator',
                        mode: this._navigatorMode,
                    });
                }, 100);
            };
            this._framesSettingMenuPopper = null;
            this._navigatorMode = 'fit';
            this._setEdgelessDisposables = () => {
                if (!this._edgelessRootService)
                    return;
                this._clearEdgelessDisposables();
                this._edgelessDisposables = new DisposableGroup();
                this._edgelessDisposables.add(this._edgelessRootService.slots.navigatorSettingUpdated.on(({ fillScreen }) => {
                    this._navigatorMode = fillScreen ? 'fill' : 'fit';
                }));
            };
            this.#_frameSettingButton_accessor_storage = __runInitializers(this, __frameSettingButton_initializers, void 0);
            this.#_frameSettingMenu_accessor_storage = (__runInitializers(this, __frameSettingButton_extraInitializers), __runInitializers(this, __frameSettingMenu_initializers, void 0));
            this.#_settingPopperShow_accessor_storage = (__runInitializers(this, __frameSettingMenu_extraInitializers), __runInitializers(this, __settingPopperShow_initializers, false));
            this.#editorHost_accessor_storage = (__runInitializers(this, __settingPopperShow_extraInitializers), __runInitializers(this, _editorHost_initializers, void 0));
            __runInitializers(this, _editorHost_extraInitializers);
        }
    };
})();
export { FramePanelHeader };
