// @blocksuite/data-view/dist/view-presets/kanban/pc/kanban-view.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { AddCursorIcon } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css } from 'lit';
import { query } from 'lit/decorators.js';
import { repeat } from 'lit/directives/repeat.js';
import { styleMap } from 'lit/directives/style-map.js';
import { html } from 'lit/static-html.js';
import { renderUniLit } from '../../../core/index.js';
import { defaultActivators } from '../../../core/utils/wc-dnd/sensors/index.js';
import { createSortContext, sortable, } from '../../../core/utils/wc-dnd/sort/sort-context.js';
import { horizontalListSortingStrategy } from '../../../core/utils/wc-dnd/sort/strategies/index.js';
import { DataViewBase } from '../../../core/view/data-view-base.js';
import { KanbanClipboardController } from './controller/clipboard.js';
import { KanbanDragController } from './controller/drag.js';
import { KanbanHotkeysController } from './controller/hotkeys.js';
import { KanbanSelectionController } from './controller/selection.js';
const styles = css `
  affine-data-view-kanban {
    user-select: none;
    display: flex;
    flex-direction: column;
  }

  .affine-data-view-kanban-groups {
    position: relative;
    z-index: 1;
    display: flex;
    gap: 20px;
    padding-bottom: 4px;
    overflow-x: scroll;
    overflow-y: hidden;
  }

  .affine-data-view-kanban-groups:hover {
    padding-bottom: 0px;
  }

  .affine-data-view-kanban-groups::-webkit-scrollbar {
    -webkit-appearance: none;
    display: block;
  }

  .affine-data-view-kanban-groups::-webkit-scrollbar:horizontal {
    height: 4px;
  }

  .affine-data-view-kanban-groups::-webkit-scrollbar-thumb {
    border-radius: 2px;
    background-color: transparent;
  }

  .affine-data-view-kanban-groups:hover::-webkit-scrollbar:horizontal {
    height: 8px;
  }

  .affine-data-view-kanban-groups:hover::-webkit-scrollbar-thumb {
    border-radius: 16px;
    background-color: var(--affine-black-30);
  }

  .affine-data-view-kanban-groups:hover::-webkit-scrollbar-track {
    background-color: var(--affine-hover-color);
  }

  .add-group-icon {
    padding: 4px;
    border-radius: 4px;
    display: flex;
    align-items: center;
    cursor: pointer;
  }

  .add-group-icon:hover {
    background-color: var(--affine-hover-color);
  }

  .add-group-icon svg {
    width: 16px;
    height: 16px;
    fill: var(--affine-icon-color);
    color: var(--affine-icon-color);
  }
`;
let DataViewKanban = (() => {
    let _classSuper = DataViewBase;
    let _groups_decorators;
    let _groups_initializers = [];
    let _groups_extraInitializers = [];
    return class DataViewKanban extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _groups_decorators = [query('.affine-data-view-kanban-groups')];
            __esDecorate(this, null, _groups_decorators, { kind: "accessor", name: "groups", static: false, private: false, access: { has: obj => "groups" in obj, get: obj => obj.groups, set: (obj, value) => { obj.groups = value; } }, metadata: _metadata }, _groups_initializers, _groups_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = styles; }
        get expose() {
            return {
                clearSelection: () => {
                    this.selectionController.clear();
                },
                addRow: position => {
                    if (this.props.view.readonly$.value)
                        return;
                    const rowId = this.props.view.rowAdd(position);
                    if (rowId) {
                        this.props.dataViewEle.openDetailPanel({
                            view: this.props.view,
                            rowId,
                        });
                    }
                    return rowId;
                },
                focusFirstCell: () => {
                    this.selectionController.focusFirstCell();
                },
                getSelection: () => {
                    return this.selectionController.selection;
                },
                hideIndicator: () => {
                    this.dragController.dropPreview.remove();
                },
                moveTo: (id, evt) => {
                    const position = this.dragController.getInsertPosition(evt);
                    if (position) {
                        position.group.group.manager.moveCardTo(id, '', position.group.group.key, position.position);
                    }
                },
                showIndicator: evt => {
                    return this.dragController.shooIndicator(evt, undefined) != null;
                },
                view: this.props.view,
                eventTrace: this.props.eventTrace,
            };
        }
        get groupManager() {
            return this.props.view.groupTrait;
        }
        render() {
            const groups = this.groupManager.groupsDataList$.value;
            if (!groups) {
                return html ``;
            }
            const vPadding = this.props.virtualPadding$.value;
            const wrapperStyle = styleMap({
                marginLeft: `-${vPadding}px`,
                marginRight: `-${vPadding}px`,
                paddingLeft: `${vPadding}px`,
                paddingRight: `${vPadding}px`,
            });
            return html `
      ${renderUniLit(this.props.headerWidget, {
                dataViewInstance: this.expose,
            })}
      <div
        class="affine-data-view-kanban-groups"
        style="${wrapperStyle}"
        @wheel="${this.onWheel}"
      >
        ${repeat(groups, group => group.key, group => {
                return html ` <affine-data-view-kanban-group
              ${sortable(group.key)}
              data-key="${group.key}"
              .dataViewEle="${this.props.dataViewEle}"
              .view="${this.props.view}"
              .group="${group}"
            ></affine-data-view-kanban-group>`;
            })}
        ${this.renderAddGroup()}
      </div>
    `;
        }
        #groups_accessor_storage;
        get groups() { return this.#groups_accessor_storage; }
        set groups(value) { this.#groups_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.dragController = new KanbanDragController(this);
            this.clipboardController = new KanbanClipboardController(this);
            this.hotkeysController = new KanbanHotkeysController(this);
            this.onWheel = (event) => {
                if (event.metaKey || event.ctrlKey) {
                    return;
                }
                const ele = event.currentTarget;
                if (ele instanceof HTMLElement) {
                    if (ele.scrollWidth === ele.clientWidth) {
                        return;
                    }
                    event.stopPropagation();
                }
            };
            this.renderAddGroup = () => {
                const addGroup = this.groupManager.addGroup;
                if (!addGroup) {
                    return;
                }
                const add = (e) => {
                    const ele = e.currentTarget;
                    popMenu(popupTargetFromElement(ele), {
                        options: {
                            items: [
                                menu.input({
                                    onComplete: text => {
                                        const column = this.groupManager.property$.value;
                                        if (column) {
                                            column.dataUpdate(() => addGroup({
                                                text,
                                                oldData: column.data$.value,
                                                dataSource: this.props.view.manager.dataSource,
                                            }));
                                        }
                                    },
                                }),
                            ],
                        },
                    });
                };
                return html ` <div
      style="height: 32px;flex-shrink:0;display:flex;align-items:center;"
      @click="${add}"
    >
      <div class="add-group-icon">${AddCursorIcon()}</div>
    </div>`;
            };
            this.selectionController = new KanbanSelectionController(this);
            this.sortContext = createSortContext({
                activators: defaultActivators,
                container: this,
                onDragEnd: evt => {
                    const over = evt.over;
                    const activeId = evt.active.id;
                    const groups = this.groupManager.groupsDataList$.value;
                    if (over && over.id !== activeId && groups) {
                        const activeIndex = groups.findIndex(data => data.key === activeId);
                        const overIndex = groups.findIndex(data => data.key === over.id);
                        this.groupManager.moveGroupTo(activeId, activeIndex > overIndex
                            ? {
                                before: true,
                                id: over.id,
                            }
                            : {
                                before: false,
                                id: over.id,
                            });
                    }
                },
                modifiers: [
                    ({ transform }) => {
                        return {
                            ...transform,
                            y: 0,
                        };
                    },
                ],
                items: computed(() => {
                    return this.groupManager.groupsDataList$.value?.map(v => v.key) ?? [];
                }),
                strategy: horizontalListSortingStrategy,
            });
            this.#groups_accessor_storage = __runInitializers(this, _groups_initializers, void 0);
            __runInitializers(this, _groups_extraInitializers);
        }
    };
})();
export { DataViewKanban };
