// react-arborist/dist/main/components/list-outer-element.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListOuterElement = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const context_1 = require("../context");
const cursor_1 = require("./cursor");
exports.ListOuterElement = (0, react_1.forwardRef)(function Outer(props, ref) {
    const { children } = props, rest = __rest(props, ["children"]);
    const tree = (0, context_1.useTreeApi)();
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ 
        // @ts-ignore
        ref: ref }, rest, { onClick: (e) => {
            if (e.currentTarget === e.target)
                tree.deselectAll();
        }, children: [(0, jsx_runtime_1.jsx)(DropContainer, {}), children] })));
});
const DropContainer = () => {
    const tree = (0, context_1.useTreeApi)();
    return ((0, jsx_runtime_1.jsx)("div", { style: {
            height: tree.visibleNodes.length * tree.rowHeight,
            width: "100%",
            position: "absolute",
            left: "0",
            right: "0",
        }, children: (0, jsx_runtime_1.jsx)(cursor_1.Cursor, {}) }));
};
