// @blocksuite/data-view/dist/core/group-by/setting.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, } from '@blocksuite/affine-components/context-menu';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { DeleteIcon } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css, html, unsafeCSS } from 'lit';
import { property, query } from 'lit/decorators.js';
import { repeat } from 'lit/directives/repeat.js';
import { KanbanSingleView } from '../../view-presets/kanban/kanban-view-manager.js';
import { TableSingleView } from '../../view-presets/table/table-view-manager.js';
import { dataViewCssVariable } from '../common/css-variable.js';
import { renderUniLit } from '../utils/uni-component/uni-component.js';
import { dragHandler } from '../utils/wc-dnd/dnd-context.js';
import { defaultActivators } from '../utils/wc-dnd/sensors/index.js';
import { createSortContext, sortable, } from '../utils/wc-dnd/sort/sort-context.js';
import { verticalListSortingStrategy } from '../utils/wc-dnd/sort/strategies/index.js';
import { groupByMatcher } from './matcher.js';
let GroupSetting = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _groupTrait_decorators;
    let _groupTrait_initializers = [];
    let _groupTrait_extraInitializers = [];
    let _groupContainer_decorators;
    let _groupContainer_initializers = [];
    let _groupContainer_extraInitializers = [];
    return class GroupSetting extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _groupTrait_decorators = [property({ attribute: false })];
            _groupContainer_decorators = [query('.group-sort-setting')];
            __esDecorate(this, null, _groupTrait_decorators, { kind: "accessor", name: "groupTrait", static: false, private: false, access: { has: obj => "groupTrait" in obj, get: obj => obj.groupTrait, set: (obj, value) => { obj.groupTrait = value; } }, metadata: _metadata }, _groupTrait_initializers, _groupTrait_extraInitializers);
            __esDecorate(this, null, _groupContainer_decorators, { kind: "accessor", name: "groupContainer", static: false, private: false, access: { has: obj => "groupContainer" in obj, get: obj => obj.groupContainer, set: (obj, value) => { obj.groupContainer = value; } }, metadata: _metadata }, _groupContainer_initializers, _groupContainer_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    data-view-group-setting {
      display: flex;
      flex-direction: column;
      gap: 4px;
      ${unsafeCSS(dataViewCssVariable())};
    }

    .group-item {
      display: flex;
      padding: 4px 12px;
      position: relative;
      cursor: grab;
    }

    .group-item-drag-bar {
      width: 4px;
      height: 12px;
      border-radius: 1px;
      background-color: #efeff0;
      position: absolute;
      left: 4px;
      top: 0;
      bottom: 0;
      margin: auto;
    }

    .group-item:hover .group-item-drag-bar {
      background-color: #c0bfc1;
    }
  `; }
        #groupTrait_accessor_storage;
        get groupTrait() { return this.#groupTrait_accessor_storage; }
        set groupTrait(value) { this.#groupTrait_accessor_storage = value; }
        connectedCallback() {
            super.connectedCallback();
            this._disposables.addFromEvent(this, 'pointerdown', e => {
                e.stopPropagation();
            });
        }
        render() {
            const groups = this.groupTrait.groupsDataList$.value;
            if (!groups) {
                return;
            }
            return html `
      <div style="padding: 7px 0;">
        <div
          style="padding: 0 4px; font-size: 12px;color: var(--affine-text-secondary-color);line-height: 20px;"
        >
          Groups
        </div>
        <div></div>
      </div>
      <div
        style="display:flex;flex-direction: column;gap: 4px;"
        class="group-sort-setting"
      >
        ${repeat(groups, group => group.key, group => {
                const props = {
                    value: group.value,
                    data: group.property.data$.value,
                    readonly: true,
                };
                const config = group.manager.config$.value;
                return html ` <div
              ${sortable(group.key)}
              ${dragHandler(group.key)}
              class="dv-hover dv-round-4 group-item"
            >
              <div class="group-item-drag-bar"></div>
              <div style="padding: 0 4px;position:relative;">
                ${renderUniLit(config?.view, props)}
                <div
                  style="position:absolute;left: 0;top: 0;right: 0;bottom: 0;"
                ></div>
              </div>
            </div>`;
            })}
      </div>
    `;
        }
        #groupContainer_accessor_storage;
        get groupContainer() { return this.#groupContainer_accessor_storage; }
        set groupContainer(value) { this.#groupContainer_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.#groupTrait_accessor_storage = __runInitializers(this, _groupTrait_initializers, void 0);
            this.groups$ = (__runInitializers(this, _groupTrait_extraInitializers), computed(() => {
                return this.groupTrait.groupsDataList$.value;
            }));
            this.sortContext = createSortContext({
                activators: defaultActivators,
                container: this,
                onDragEnd: evt => {
                    const over = evt.over;
                    const activeId = evt.active.id;
                    const groups = this.groups$.value;
                    if (over && over.id !== activeId && groups) {
                        const activeIndex = groups.findIndex(data => data.key === activeId);
                        const overIndex = groups.findIndex(data => data.key === over.id);
                        this.groupTrait.moveGroupTo(activeId, activeIndex > overIndex
                            ? {
                                before: true,
                                id: over.id,
                            }
                            : {
                                before: false,
                                id: over.id,
                            });
                    }
                },
                modifiers: [
                    ({ transform }) => {
                        return {
                            ...transform,
                            x: 0,
                        };
                    },
                ],
                items: computed(() => {
                    return this.groupTrait.groupsDataList$.value?.map(v => v.key) ?? [];
                }),
                strategy: verticalListSortingStrategy,
            });
            this.#groupContainer_accessor_storage = __runInitializers(this, _groupContainer_initializers, void 0);
            __runInitializers(this, _groupContainer_extraInitializers);
        }
    };
})();
export { GroupSetting };
export const selectGroupByProperty = (group, ops) => {
    const view = group.view;
    return {
        onClose: ops?.onClose,
        title: {
            text: 'Group by',
            onBack: ops?.onBack,
        },
        items: [
            menu.group({
                items: view.propertiesWithoutFilter$.value
                    .filter(id => {
                    if (view.propertyGet(id).type$.value === 'title') {
                        return false;
                    }
                    return !!groupByMatcher.match(view.propertyGet(id).dataType$.value);
                })
                    .map(id => {
                    const property = view.propertyGet(id);
                    return menu.action({
                        name: property.name$.value,
                        isSelected: group.property$.value?.id === id,
                        prefix: html ` <uni-lit .uni="${property.icon}"></uni-lit>`,
                        select: () => {
                            group.changeGroup(id);
                            ops?.onSelect?.(id);
                        },
                    });
                }),
            }),
            menu.group({
                items: [
                    menu.action({
                        prefix: DeleteIcon(),
                        hide: () => view instanceof KanbanSingleView || group.property$.value == null,
                        class: { 'delete-item': true },
                        name: 'Remove Grouping',
                        select: () => {
                            group.changeGroup(undefined);
                            ops?.onSelect?.();
                        },
                    }),
                ],
            }),
        ],
    };
};
export const popSelectGroupByProperty = (target, group, ops) => {
    popMenu(target, {
        options: selectGroupByProperty(group, ops),
    });
};
export const popGroupSetting = (target, group, onBack) => {
    const view = group.view;
    const groupProperty = group.property$.value;
    if (groupProperty == null) {
        return;
    }
    const type = groupProperty.type$.value;
    if (!type) {
        return;
    }
    const icon = view.propertyIconGet(type);
    const menuHandler = popMenu(target, {
        options: {
            title: {
                text: 'Group',
                onBack: onBack,
            },
            items: [
                menu.group({
                    items: [
                        menu.subMenu({
                            name: 'Group By',
                            postfix: html `
                <div
                  style="display:flex;align-items:center;gap: 4px;font-size: 12px;line-height: 20px;color: var(--affine-text-secondary-color);margin-right: 4px;margin-left: 8px;"
                  class="dv-icon-16"
                >
                  ${renderUniLit(icon, {})} ${groupProperty.name$.value}
                </div>
              `,
                            label: () => html `
                <div style="color: var(--affine-text-secondary-color);">
                  Group By
                </div>
              `,
                            options: selectGroupByProperty(group, {
                                onSelect: () => {
                                    menuHandler.close();
                                    popGroupSetting(target, group, onBack);
                                },
                            }),
                        }),
                    ],
                }),
                menu.group({
                    items: [
                        menu => html ` <data-view-group-setting
                @mouseenter="${() => menu.closeSubMenu()}"
                .groupTrait="${group}"
                .columnId="${groupProperty.id}"
              ></data-view-group-setting>`,
                    ],
                }),
                menu.group({
                    items: [
                        menu.action({
                            name: 'Remove grouping',
                            prefix: DeleteIcon(),
                            class: { 'delete-item': true },
                            hide: () => !(view instanceof TableSingleView),
                            select: () => {
                                group.changeGroup(undefined);
                            },
                        }),
                    ],
                }),
            ],
        },
    });
};
