// @blocksuite/data-view/dist/core/common/properties.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, } from '@blocksuite/affine-components/context-menu';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { InvisibleIcon, ViewIcon } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css, html } from 'lit';
import { property, query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { repeat } from 'lit/directives/repeat.js';
import { dragHandler } from '../utils/wc-dnd/dnd-context.js';
import { defaultActivators } from '../utils/wc-dnd/sensors/index.js';
import { createSortContext, sortable, } from '../utils/wc-dnd/sort/sort-context.js';
import { verticalListSortingStrategy } from '../utils/wc-dnd/sort/strategies/index.js';
let DataViewPropertiesSettingView = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _view_decorators;
    let _view_initializers = [];
    let _view_extraInitializers = [];
    let _groupContainer_decorators;
    let _groupContainer_initializers = [];
    let _groupContainer_extraInitializers = [];
    let _onBack_decorators;
    let _onBack_initializers = [];
    let _onBack_extraInitializers = [];
    return class DataViewPropertiesSettingView extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _view_decorators = [property({ attribute: false })];
            _groupContainer_decorators = [query('.properties-group')];
            _onBack_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _view_decorators, { kind: "accessor", name: "view", static: false, private: false, access: { has: obj => "view" in obj, get: obj => obj.view, set: (obj, value) => { obj.view = value; } }, metadata: _metadata }, _view_initializers, _view_extraInitializers);
            __esDecorate(this, null, _groupContainer_decorators, { kind: "accessor", name: "groupContainer", static: false, private: false, access: { has: obj => "groupContainer" in obj, get: obj => obj.groupContainer, set: (obj, value) => { obj.groupContainer = value; } }, metadata: _metadata }, _groupContainer_initializers, _groupContainer_extraInitializers);
            __esDecorate(this, null, _onBack_decorators, { kind: "accessor", name: "onBack", static: false, private: false, access: { has: obj => "onBack" in obj, get: obj => obj.onBack, set: (obj, value) => { obj.onBack = value; } }, metadata: _metadata }, _onBack_initializers, _onBack_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .properties-group-header {
      user-select: none;
      padding: 4px 12px 12px 12px;
      margin-bottom: 12px;
      display: flex;
      align-items: center;
      justify-content: space-between;
      border-bottom: 1px solid var(--affine-divider-color);
    }

    .properties-group-title {
      font-size: 12px;
      line-height: 20px;
      color: var(--affine-text-secondary-color);
      display: flex;
      align-items: center;
      gap: 8px;
    }

    .properties-group-op {
      padding: 4px 8px;
      font-size: 12px;
      line-height: 20px;
      font-weight: 500;
      border-radius: 4px;
      cursor: pointer;
    }

    .properties-group-op:hover {
      background-color: var(--affine-hover-color);
    }

    .properties-group {
      min-height: 40px;
    }

    .property-item {
      padding: 4px;
      display: flex;
      align-items: center;
      gap: 8px;
      user-select: none;
      cursor: pointer;
      border-radius: 4px;
    }

    .property-item-drag-bar {
      width: 4px;
      height: 12px;
      border-radius: 1px;
      background-color: #efeff0;
    }

    .property-item:hover .property-item-drag-bar {
      background-color: #c0bfc1;
    }

    .property-item-icon {
      display: flex;
      align-items: center;
    }

    .property-item-icon svg {
      color: var(--affine-icon-color);
      fill: var(--affine-icon-color);
      width: 20px;
      height: 20px;
    }

    .property-item-op-icon {
      display: flex;
      align-items: center;
      border-radius: 4px;
    }

    .property-item-op-icon:hover {
      background-color: var(--affine-hover-color);
    }

    .property-item-op-icon.disabled:hover {
      background-color: transparent;
    }

    .property-item-op-icon svg {
      fill: var(--affine-icon-color);
      color: var(--affine-icon-color);
      width: 20px;
      height: 20px;
    }

    .property-item-op-icon.disabled svg {
      fill: var(--affine-text-disable-color);
      color: var(--affine-text-disable-color);
    }

    .property-item-name {
      font-size: 14px;
      line-height: 22px;
      flex: 1;
    }
  `; }
        #view_accessor_storage;
        get view() { return this.#view_accessor_storage; }
        set view(value) { this.#view_accessor_storage = value; }
        itemsGroup() {
            return this.view.propertiesWithoutFilter$.value.map(id => this.view.propertyGet(id));
        }
        connectedCallback() {
            super.connectedCallback();
            this._disposables.addFromEvent(this, 'pointerdown', e => {
                e.stopPropagation();
            });
        }
        render() {
            const items = this.itemsGroup();
            return html `
      <div class="properties-group">
        ${repeat(items, v => v.id, this.renderProperty)}
      </div>
    `;
        }
        #groupContainer_accessor_storage;
        get groupContainer() { return this.#groupContainer_accessor_storage; }
        set groupContainer(value) { this.#groupContainer_accessor_storage = value; }
        #onBack_accessor_storage;
        get onBack() { return this.#onBack_accessor_storage; }
        set onBack(value) { this.#onBack_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.#view_accessor_storage = __runInitializers(this, _view_initializers, void 0);
            this.items$ = (__runInitializers(this, _view_extraInitializers), computed(() => {
                return this.view.propertiesWithoutFilter$.value;
            }));
            this.renderProperty = (property) => {
                const isTitle = property.type$.value === 'title';
                const icon = property.hide$.value ? InvisibleIcon() : ViewIcon();
                const changeVisible = () => {
                    if (property.type$.value !== 'title') {
                        property.hideSet(!property.hide$.value);
                    }
                };
                const classList = classMap({
                    'property-item-op-icon': true,
                    disabled: isTitle,
                });
                return html ` <div
      ${dragHandler(property.id)}
      ${sortable(property.id)}
      class="property-item"
    >
      <div class="property-item-drag-bar"></div>
      <uni-lit class="property-item-icon" .uni="${property.icon}"></uni-lit>
      <div class="property-item-name">${property.name$.value}</div>
      <div class="${classList}" @click="${changeVisible}">${icon}</div>
    </div>`;
            };
            this.sortContext = createSortContext({
                activators: defaultActivators,
                container: this,
                onDragEnd: evt => {
                    const over = evt.over;
                    const activeId = evt.active.id;
                    if (over && over.id !== activeId) {
                        const properties = this.items$.value;
                        const activeIndex = properties.findIndex(id => id === activeId);
                        const overIndex = properties.findIndex(id => id === over.id);
                        this.view.propertyMove(activeId, activeIndex > overIndex
                            ? {
                                before: true,
                                id: over.id,
                            }
                            : {
                                before: false,
                                id: over.id,
                            });
                    }
                },
                modifiers: [
                    ({ transform }) => {
                        return {
                            ...transform,
                            x: 0,
                        };
                    },
                ],
                items: this.items$,
                strategy: verticalListSortingStrategy,
            });
            this.#groupContainer_accessor_storage = __runInitializers(this, _groupContainer_initializers, void 0);
            this.#onBack_accessor_storage = (__runInitializers(this, _groupContainer_extraInitializers), __runInitializers(this, _onBack_initializers, undefined));
            __runInitializers(this, _onBack_extraInitializers);
        }
    };
})();
export { DataViewPropertiesSettingView };
export const popPropertiesSetting = (target, props) => {
    popMenu(target, {
        options: {
            title: {
                text: 'Properties',
                onBack: props.onBack,
                postfix: () => {
                    const items = props.view.propertiesWithoutFilter$.value.map(id => props.view.propertyGet(id));
                    const isAllShowed = items.every(v => !v.hide$.value);
                    const clickChangeAll = () => {
                        props.view.propertiesWithoutFilter$.value.forEach(id => {
                            if (props.view.propertyTypeGet(id) !== 'title') {
                                props.view.propertyHideSet(id, isAllShowed);
                            }
                        });
                    };
                    return html ` <div
            class="properties-group-op"
            @click="${clickChangeAll}"
          >
            ${isAllShowed ? 'Hide All' : 'Show All'}
          </div>`;
                },
            },
            items: [
                menu.group({
                    items: [
                        () => html ` <data-view-properties-setting
                .view="${props.view}"
              ></data-view-properties-setting>`,
                    ],
                }),
            ],
        },
    });
    // const view = new DataViewPropertiesSettingView();
    // view.view = props.view;
    // view.onBack = () => {
    //   close();
    //   props.onBack?.();
    // };
    // const close = createPopup(target, view, { onClose: props.onClose });
};
