// @blocksuite/block-std/dist/gfx/controller.js
import { assertType, Bound, DisposableGroup, getCommonBoundWithRotation, last, } from '@blocksuite/global/utils';
import { Signal } from '@preact/signals-core';
import { LifeCycleWatcher } from '../extension/lifecycle-watcher.js';
import { onSurfaceAdded } from '../utils/gfx.js';
import { GfxClassExtenderIdentifier, GfxExtensionIdentifier, } from './extension.js';
import { GridManager } from './grid.js';
import { gfxControllerKey } from './identifiers.js';
import { KeyboardController } from './keyboard.js';
import { LayerManager } from './layer.js';
import { GfxBlockElementModel } from './model/gfx-block-model.js';
import { GfxGroupLikeElementModel, GfxPrimitiveElementModel, } from './model/surface/element-model.js';
import { Viewport } from './viewport.js';
export class GfxController extends LifeCycleWatcher {
    static { this.key = gfxControllerKey; }
    get doc() {
        return this.std.doc;
    }
    get elementsBound() {
        return getCommonBoundWithRotation(this.gfxElements);
    }
    get gfxElements() {
        return [...this.layer.blocks, ...this.layer.canvasElements];
    }
    get surface() {
        return this._surface;
    }
    get surfaceComponent() {
        return this.surface
            ? (this.std.view.getBlock(this.surface.id) ?? null)
            : null;
    }
    constructor(std) {
        super(std);
        this._disposables = new DisposableGroup();
        this._surface = null;
        this.cursor$ = new Signal();
        this.viewport = new Viewport();
        this.grid = new GridManager();
        this.layer = new LayerManager(this.doc, null);
        this.keyboard = new KeyboardController(std);
        this._disposables.add(onSurfaceAdded(this.doc, surface => {
            this._surface = surface;
            if (surface) {
                this._disposables.add(this.grid.watch({ surface }));
                this.layer.watch({ surface });
            }
        }));
        this._disposables.add(this.grid.watch({ doc: this.doc }));
        this._disposables.add(this.layer);
        this._disposables.add(this.viewport);
        this._disposables.add(this.keyboard);
        this.std.provider.getAll(GfxClassExtenderIdentifier).forEach(ext => {
            ext.extendFn(this);
        });
    }
    deleteElement(element) {
        element = typeof element === 'string' ? element : element.id;
        assertType(element);
        if (this.surface?.hasElementById(element)) {
            this.surface.deleteElement(element);
        }
        else {
            const block = this.doc.getBlock(element)?.model;
            block && this.doc.deleteBlock(block);
        }
    }
    /**
     * Get a block or element by its id.
     * Note that non-gfx block can also be queried in this method.
     * @param id
     * @returns
     */
    getElementById(id) {
        // @ts-ignore
        return (this.surface?.getElementById(id) ?? this.doc.getBlock(id)?.model ?? null);
    }
    getElementByPoint(x, y, options = { all: false, hitThreshold: 10 }) {
        options.zoom = this.viewport.zoom;
        options.hitThreshold ??= 10;
        const hitThreshold = options.hitThreshold;
        const responsePadding = options.responsePadding ?? [
            hitThreshold / 2,
            hitThreshold / 2,
        ];
        const all = options.all ?? false;
        const hitTestBound = {
            x: x - responsePadding[0],
            y: y - responsePadding[1],
            w: responsePadding[0] * 2,
            h: responsePadding[1] * 2,
        };
        const candidates = this.grid.search(hitTestBound);
        const picked = candidates.filter(elm => elm.includesPoint(x, y, options, this.std.host) ||
            elm.externalBound?.isPointInBound([x, y]));
        picked.sort(this.layer.compare);
        if (all) {
            return picked;
        }
        return last(picked) ?? null;
    }
    getElementInGroup(x, y, options) {
        const selectionManager = this.selection;
        const results = this.getElementByPoint(x, y, {
            ...options,
            all: true,
        });
        let picked = last(results) ?? null;
        const { activeGroup } = selectionManager;
        const first = picked;
        if (activeGroup && picked && activeGroup.hasDescendant(picked)) {
            let index = results.length - 1;
            while (picked === activeGroup ||
                (picked instanceof GfxGroupLikeElementModel &&
                    picked.hasDescendant(activeGroup))) {
                picked = results[--index];
            }
        }
        else if (picked) {
            let index = results.length - 1;
            while (picked.group instanceof GfxGroupLikeElementModel) {
                if (--index < 0) {
                    picked = null;
                    break;
                }
                picked = results[index];
            }
        }
        return (picked ?? first);
    }
    getElementsByBound(bound, options = {
        type: 'all',
    }) {
        bound = bound instanceof Bound ? bound : Bound.from(bound);
        let candidates = this.grid.search(bound);
        if (options.type !== 'all') {
            const filter = options.type === 'block'
                ? (elm) => elm instanceof GfxBlockElementModel
                : (elm) => elm instanceof GfxPrimitiveElementModel;
            candidates = candidates.filter(filter);
        }
        candidates.sort(this.layer.compare);
        return candidates;
    }
    getElementsByType(type) {
        return (this.surface?.getElementsByType(type) ??
            this.doc.getBlocksByFlavour(type).map(b => b.model));
    }
    mounted() {
        this.viewport.setViewportElement(this.std.host);
        this.std.provider.getAll(GfxExtensionIdentifier).forEach(ext => {
            ext.mounted();
        });
    }
    unmounted() {
        this.std.provider.getAll(GfxExtensionIdentifier).forEach(ext => {
            ext.unmounted();
        });
        this.viewport.clearViewportElement();
        this._disposables.dispose();
    }
    updateElement(element, props) {
        const elemId = typeof element === 'string' ? element : element.id;
        if (this.surface?.hasElementById(elemId)) {
            this.surface.updateElement(elemId, props);
        }
        else {
            const block = this.doc.getBlock(elemId);
            block && this.doc.updateBlock(block.model, props);
        }
    }
}
