// @blocksuite/block-std/dist/gfx/extension.js
import { createIdentifier } from '@blocksuite/global/di';
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { Extension } from '../extension/extension.js';
import { GfxControllerIdentifier } from './identifiers.js';
export const GfxExtensionIdentifier = createIdentifier('GfxExtension');
export const GfxClassExtenderIdentifier = createIdentifier('GfxClassExtender');
export class GfxExtension extends Extension {
    get std() {
        return this.gfx.std;
    }
    constructor(gfx) {
        super();
        this.gfx = gfx;
    }
    // This method is used to extend the GfxController
    static extendGfx(_) { }
    static setup(di) {
        if (!this.key) {
            throw new BlockSuiteError(ErrorCode.ValueNotExists, 'key is not defined in the GfxExtension');
        }
        di.addImpl(GfxClassExtenderIdentifier(this.key), {
            extendFn: this.extendGfx,
        });
        di.add(this, [
            GfxControllerIdentifier,
        ]);
        di.addImpl(GfxExtensionIdentifier(this.key), provider => provider.get(this));
    }
    mounted() { }
    unmounted() { }
}
