// @blocksuite/block-std/dist/gfx/surface-middleware.js
import { createIdentifier } from '@blocksuite/global/di';
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { Extension } from '../extension/extension.js';
import { LifeCycleWatcher } from '../extension/lifecycle-watcher.js';
import { StdIdentifier } from '../identifier.js';
import { onSurfaceAdded } from '../utils/gfx.js';
import { GfxControllerIdentifier } from './identifiers.js';
export class SurfaceMiddlewareBuilder extends Extension {
    static { this.key = ''; }
    get gfx() {
        return this.std.provider.get(GfxControllerIdentifier);
    }
    constructor(std) {
        super();
        this.std = std;
    }
    static setup(di) {
        if (!this.key) {
            throw new BlockSuiteError(ErrorCode.ValueNotExists, 'The surface middleware builder should have a static key property.');
        }
        di.addImpl(SurfaceMiddlewareBuilderIdentifier(this.key), this, [
            StdIdentifier,
        ]);
    }
    mounted() { }
    unmounted() { }
}
export const SurfaceMiddlewareBuilderIdentifier = createIdentifier('SurfaceMiddlewareBuilder');
export class SurfaceMiddlewareExtension extends LifeCycleWatcher {
    static { this.key = 'surfaceMiddleware'; }
    mounted() {
        const builders = Array.from(this.std.provider.getAll(SurfaceMiddlewareBuilderIdentifier).values());
        const dispose = onSurfaceAdded(this.std.doc, surface => {
            if (surface) {
                surface.applyMiddlewares(builders.map(builder => builder.middleware));
                queueMicrotask(() => dispose());
            }
        });
    }
}
