// @blocksuite/affine-components/dist/drag-indicator/index.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { css, html, LitElement } from 'lit';
import { property } from 'lit/decorators.js';
import { styleMap } from 'lit/directives/style-map.js';
let DragIndicator = (() => {
    let _classSuper = LitElement;
    let _rect_decorators;
    let _rect_initializers = [];
    let _rect_extraInitializers = [];
    return class DragIndicator extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _rect_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _rect_decorators, { kind: "accessor", name: "rect", static: false, private: false, access: { has: obj => "rect" in obj, get: obj => obj.rect, set: (obj, value) => { obj.rect = value; } }, metadata: _metadata }, _rect_initializers, _rect_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .affine-drag-indicator {
      position: absolute;
      top: 0;
      left: 0;
      background: var(--affine-primary-color);
      transition-property: width, height, transform;
      transition-duration: 100ms;
      transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
      transition-delay: 0s;
      transform-origin: 0 0;
      pointer-events: none;
      z-index: 2;
    }
  `; }
        render() {
            if (!this.rect) {
                return null;
            }
            const { left, top, width, height } = this.rect;
            const style = styleMap({
                width: `${width}px`,
                height: `${height}px`,
                transform: `translate(${left}px, ${top}px)`,
            });
            return html `<div class="affine-drag-indicator" style=${style}></div>`;
        }
        #rect_accessor_storage = __runInitializers(this, _rect_initializers, null);
        get rect() { return this.#rect_accessor_storage; }
        set rect(value) { this.#rect_accessor_storage = value; }
        constructor() {
            super(...arguments);
            __runInitializers(this, _rect_extraInitializers);
        }
    };
})();
export { DragIndicator };
export function effects() {
    customElements.define('affine-drag-indicator', DragIndicator);
}
