// @blocksuite/affine-components/dist/rich-text/inline/presets/nodes/reference-node/reference-config.js
import { createIdentifier } from '@blocksuite/global/di';
export const ReferenceNodeConfigIdentifier = createIdentifier('AffineReferenceNodeConfig');
export function ReferenceNodeConfigExtension(config) {
    return {
        setup: di => {
            di.addImpl(ReferenceNodeConfigIdentifier, () => ({ ...config }));
        },
    };
}
export class ReferenceNodeConfigProvider {
    get customContent() {
        return this._customContent;
    }
    get doc() {
        return this.std.doc;
    }
    get hidePopup() {
        return this._hidePopup;
    }
    get interactable() {
        return this._interactable;
    }
    constructor(std) {
        this.std = std;
        this._customContent = undefined;
        this._hidePopup = false;
        this._interactable = true;
    }
    setCustomContent(content) {
        this._customContent = content;
    }
    setHidePopup(hidePopup) {
        this._hidePopup = hidePopup;
    }
    setInteractable(interactable) {
        this._interactable = interactable;
    }
}
