// @blocksuite/data-view/dist/view-presets/table/table-view-manager.js
import { insertPositionToIndex, } from '@blocksuite/affine-shared/utils';
import { computed } from '@preact/signals-core';
import { evalFilter } from '../../core/filter/eval.js';
import { generateDefaultValues } from '../../core/filter/generate-default-values.js';
import { FilterTrait, filterTraitKey } from '../../core/filter/trait.js';
import { emptyFilterGroup } from '../../core/filter/utils.js';
import { GroupTrait, groupTraitKey, sortByManually, } from '../../core/group-by/trait.js';
import { SortManager, sortTraitKey } from '../../core/sort/manager.js';
import { PropertyBase } from '../../core/view-manager/property.js';
import { SingleViewBase, } from '../../core/view-manager/single-view.js';
import { DEFAULT_COLUMN_MIN_WIDTH, DEFAULT_COLUMN_WIDTH } from './consts.js';
export class TableSingleView extends SingleViewBase {
    get groupProperties() {
        return this.data$.value?.groupProperties ?? [];
    }
    get name() {
        return this.data$.value?.name ?? '';
    }
    get type() {
        return this.data$.value?.mode ?? 'table';
    }
    constructor(viewManager, viewId) {
        super(viewManager, viewId);
        this.propertiesWithoutFilter$ = computed(() => {
            const needShow = new Set(this.dataSource.properties$.value);
            const result = [];
            this.data$.value?.columns.forEach(v => {
                if (needShow.has(v.id)) {
                    result.push(v.id);
                    needShow.delete(v.id);
                }
            });
            result.push(...needShow);
            return result;
        });
        this.computedColumns$ = computed(() => {
            return this.propertiesWithoutFilter$.value.map(id => {
                const column = this.propertyGet(id);
                return {
                    id: column.id,
                    hide: column.hide$.value,
                    width: column.width$.value,
                    statCalcType: column.statCalcOp$.value,
                };
            });
        });
        this.filter$ = computed(() => {
            return this.data$.value?.filter ?? emptyFilterGroup;
        });
        this.groupBy$ = computed(() => {
            return this.data$.value?.groupBy;
        });
        this.sortList$ = computed(() => {
            return this.data$.value?.sort;
        });
        this.sortManager = this.traitSet(sortTraitKey, new SortManager(this.sortList$, this, {
            setSortList: sortList => {
                this.dataUpdate(data => {
                    return {
                        sort: {
                            ...data.sort,
                            ...sortList,
                        },
                    };
                });
            },
        }));
        this.detailProperties$ = computed(() => {
            return this.propertiesWithoutFilter$.value.filter(id => this.propertyTypeGet(id) !== 'title');
        });
        this.filterTrait = this.traitSet(filterTraitKey, new FilterTrait(this.filter$, this, {
            filterSet: (filter) => {
                this.dataUpdate(() => {
                    return {
                        filter,
                    };
                });
            },
        }));
        this.groupTrait = this.traitSet(groupTraitKey, new GroupTrait(this.groupBy$, this, {
            groupBySet: groupBy => {
                this.dataUpdate(() => {
                    return {
                        groupBy,
                    };
                });
            },
            sortGroup: ids => sortByManually(ids, v => v, this.groupProperties.map(v => v.key)),
            sortRow: (key, ids) => {
                const property = this.groupProperties.find(v => v.key === key);
                return sortByManually(ids, v => v, property?.manuallyCardSort ?? []);
            },
            changeGroupSort: keys => {
                const map = new Map(this.groupProperties.map(v => [v.key, v]));
                this.dataUpdate(() => {
                    return {
                        groupProperties: keys.map(key => {
                            const property = map.get(key);
                            if (property) {
                                return property;
                            }
                            return {
                                key,
                                hide: false,
                                manuallyCardSort: [],
                            };
                        }),
                    };
                });
            },
            changeRowSort: (groupKeys, groupKey, keys) => {
                const map = new Map(this.groupProperties.map(v => [v.key, v]));
                this.dataUpdate(() => {
                    return {
                        groupProperties: groupKeys.map(key => {
                            if (key === groupKey) {
                                const group = map.get(key);
                                return group
                                    ? {
                                        ...group,
                                        manuallyCardSort: keys,
                                    }
                                    : {
                                        key,
                                        hide: false,
                                        manuallyCardSort: keys,
                                    };
                            }
                            else {
                                return (map.get(key) ?? {
                                    key,
                                    hide: false,
                                    manuallyCardSort: [],
                                });
                            }
                        }),
                    };
                });
            },
        }));
        this.mainProperties$ = computed(() => {
            return (this.data$.value?.header ?? {
                titleColumn: this.propertiesWithoutFilter$.value.find(id => this.propertyTypeGet(id) === 'title'),
                iconColumn: 'type',
            });
        });
        this.propertyIds$ = computed(() => {
            return this.propertiesWithoutFilter$.value.filter(id => !this.propertyHideGet(id));
        });
        this.readonly$ = computed(() => {
            return this.manager.readonly$.value;
        });
    }
    columnGetStatCalcOp(columnId) {
        return this.data$.value?.columns.find(v => v.id === columnId)?.statCalcType;
    }
    columnGetWidth(columnId) {
        const column = this.data$.value?.columns.find(v => v.id === columnId);
        if (column?.width != null) {
            return column.width;
        }
        const type = this.propertyTypeGet(columnId);
        if (type === 'title') {
            return 260;
        }
        return DEFAULT_COLUMN_WIDTH;
    }
    columnUpdateStatCalcOp(columnId, op) {
        this.dataUpdate(() => {
            return {
                columns: this.computedColumns$.value.map(v => v.id === columnId
                    ? {
                        ...v,
                        statCalcType: op,
                    }
                    : v),
            };
        });
    }
    columnUpdateWidth(columnId, width) {
        this.dataUpdate(() => {
            return {
                columns: this.computedColumns$.value.map(v => v.id === columnId
                    ? {
                        ...v,
                        width: width,
                    }
                    : v),
            };
        });
    }
    isShow(rowId) {
        if (this.filter$.value?.conditions.length) {
            const rowMap = Object.fromEntries(this.properties$.value.map(column => [
                column.id,
                column.cellGet(rowId).jsonValue$.value,
            ]));
            return evalFilter(this.filter$.value, rowMap);
        }
        return true;
    }
    minWidthGet(type) {
        return (this.propertyMetaGet(type)?.config.minWidth ?? DEFAULT_COLUMN_MIN_WIDTH);
    }
    propertyGet(columnId) {
        return new TableColumn(this, columnId);
    }
    propertyHideGet(columnId) {
        return (this.data$.value?.columns.find(v => v.id === columnId)?.hide ?? false);
    }
    propertyHideSet(columnId, hide) {
        this.dataUpdate(() => {
            return {
                columns: this.computedColumns$.value.map(v => v.id === columnId
                    ? {
                        ...v,
                        hide,
                    }
                    : v),
            };
        });
    }
    propertyMove(columnId, toAfterOfColumn) {
        this.dataUpdate(() => {
            const columnIndex = this.computedColumns$.value.findIndex(v => v.id === columnId);
            if (columnIndex < 0) {
                return {};
            }
            const columns = [...this.computedColumns$.value];
            const [column] = columns.splice(columnIndex, 1);
            const index = insertPositionToIndex(toAfterOfColumn, columns);
            columns.splice(index, 0, column);
            return {
                columns,
            };
        });
    }
    rowAdd(insertPosition, groupKey) {
        const id = super.rowAdd(insertPosition);
        const filter = this.filter$.value;
        if (filter.conditions.length > 0) {
            const defaultValues = generateDefaultValues(filter, this.vars$.value);
            Object.entries(defaultValues).forEach(([propertyId, jsonValue]) => {
                const property = this.propertyGet(propertyId);
                const propertyMeta = this.propertyMetaGet(property.type$.value);
                if (propertyMeta?.config.cellFromJson) {
                    const value = propertyMeta.config.cellFromJson({
                        value: jsonValue,
                        data: property.data$.value,
                        dataSource: this.dataSource,
                    });
                    this.cellValueSet(id, propertyId, value);
                }
            });
        }
        if (groupKey) {
            this.groupTrait.addToGroup(id, groupKey);
        }
        return id;
    }
    rowMove(rowId, position, fromGroup, toGroup) {
        if (toGroup == null) {
            super.rowMove(rowId, position);
            return;
        }
        this.groupTrait.moveCardTo(rowId, fromGroup, toGroup, position);
    }
    rowNextGet(rowId) {
        const index = this.rows$.value.indexOf(rowId);
        return this.rows$.value[index + 1];
    }
    rowPrevGet(rowId) {
        const index = this.rows$.value.indexOf(rowId);
        return this.rows$.value[index - 1];
    }
    rowsMapping(rows) {
        return this.sortManager.sort(super.rowsMapping(rows));
    }
}
export class TableColumn extends PropertyBase {
    get minWidth() {
        return this.tableView.minWidthGet(this.type$.value);
    }
    constructor(tableView, columnId) {
        super(tableView, columnId);
        this.tableView = tableView;
        this.statCalcOp$ = computed(() => {
            return this.tableView.columnGetStatCalcOp(this.id);
        });
        this.width$ = computed(() => {
            return this.tableView.columnGetWidth(this.id);
        });
    }
    updateStatCalcOp(type) {
        return this.tableView.columnUpdateStatCalcOp(this.id, type);
    }
    updateWidth(width) {
        this.tableView.columnUpdateWidth(this.id, width);
    }
}
