// @blocksuite/data-view/dist/widget-presets/tools/presets/search/search.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { unsafeCSSVarV2 } from '@blocksuite/affine-shared/theme';
import { IS_MOBILE } from '@blocksuite/global/env';
import { CloseIcon, SearchIcon } from '@blocksuite/icons/lit';
import { baseTheme } from '@toeverything/theme';
import { css, html, unsafeCSS } from 'lit';
import { query, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { styleMap } from 'lit/directives/style-map.js';
import { stopPropagation } from '../../../../core/utils/event.js';
import { WidgetBase } from '../../../../core/widget/widget-base.js';
const styles = css `
  .affine-database-search-container {
    position: relative;
    display: flex;
    align-items: center;
    gap: 8px;
    width: 24px;
    height: 24px;
    border-radius: 4px;
    transition: width 0.3s ease;
    overflow: hidden;
  }

  .search-container-expand {
    overflow: visible;
    width: 138px;
    background-color: var(--affine-hover-color);
  }

  .search-input-container {
    display: flex;
    align-items: center;
  }

  .close-icon {
    display: flex;
    align-items: center;
    padding-right: 8px;
    height: 100%;
    cursor: pointer;
  }

  .affine-database-search-input-icon {
    position: absolute;
    left: 0;
    font-size: 20px;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
    cursor: pointer;
    padding: 2px;
    border-radius: 4px;
    color: ${unsafeCSSVarV2('icon/primary')};
  }

  .affine-database-search-input-icon:hover {
    background: var(--affine-hover-color);
  }

  .search-container-expand .affine-database-search-input-icon {
    left: 4px;
    pointer-events: none;
  }

  .affine-database-search-input {
    flex: 1;
    width: 100%;
    padding: 0 2px 0 30px;
    border: none;
    font-family: ${unsafeCSS(baseTheme.fontSansFamily)};
    font-size: var(--affine-font-sm);
    box-sizing: border-box;
    color: inherit;
    background: transparent;
    outline: none;
  }

  .affine-database-search-input::placeholder {
    color: var(--affine-placeholder-color);
    font-size: var(--affine-font-sm);
  }
`;
let DataViewHeaderToolsSearch = (() => {
    let _classSuper = WidgetBase;
    let __searchInput_decorators;
    let __searchInput_initializers = [];
    let __searchInput_extraInitializers = [];
    let _showSearch_decorators;
    let _showSearch_initializers = [];
    let _showSearch_extraInitializers = [];
    return class DataViewHeaderToolsSearch extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            __searchInput_decorators = [query('.affine-database-search-input')];
            _showSearch_decorators = [state()];
            __esDecorate(this, null, __searchInput_decorators, { kind: "accessor", name: "_searchInput", static: false, private: false, access: { has: obj => "_searchInput" in obj, get: obj => obj._searchInput, set: (obj, value) => { obj._searchInput = value; } }, metadata: _metadata }, __searchInput_initializers, __searchInput_extraInitializers);
            __esDecorate(this, null, _showSearch_decorators, { kind: "accessor", name: "showSearch", static: false, private: false, access: { has: obj => "showSearch" in obj, get: obj => obj.showSearch, set: (obj, value) => { obj.showSearch = value; } }, metadata: _metadata }, _showSearch_initializers, _showSearch_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = styles; }
        connectedCallback() {
            super.connectedCallback();
            this.style.display = IS_MOBILE ? 'none' : 'flex';
        }
        render() {
            const searchToolClassMap = classMap({
                'affine-database-search-container': true,
                'search-container-expand': this.showSearch,
                active: this.showSearch,
            });
            return html `
      <label class="${searchToolClassMap}" @click="${this._clickSearch}">
        <div class="affine-database-search-input-icon">${SearchIcon()}</div>
        <input
          placeholder="Search..."
          class="affine-database-search-input"
          @input="${this._onSearch}"
          @click="${(event) => event.stopPropagation()}"
          @keydown="${this._onSearchKeydown}"
          @pointerdown="${stopPropagation}"
          @blur="${this._onSearchBlur}"
        />
        <div class="close-icon" @mousedown="${this._clearSearch}">
          ${CloseIcon()}
          <affine-tooltip>
            <span
              style=${styleMap({
                display: 'flex',
                alignItems: 'center',
                boxSizing: 'border-box',
                padding: '2px 6px',
                borderRadius: '4px',
                background: 'var(--affine-white-10)',
            })}
              >Esc</span
            >
            to clear all
          </affine-tooltip>
        </div>
      </label>
    `;
        }
        #_searchInput_accessor_storage;
        get _searchInput() { return this.#_searchInput_accessor_storage; }
        set _searchInput(value) { this.#_searchInput_accessor_storage = value; }
        #showSearch_accessor_storage;
        get showSearch() { return this.#showSearch_accessor_storage; }
        set showSearch(value) { this.#showSearch_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._clearSearch = () => {
                this._searchInput.value = '';
                this.view.setSearch('');
                this.preventBlur = true;
                setTimeout(() => {
                    this.preventBlur = false;
                });
            };
            this._clickSearch = (e) => {
                e.stopPropagation();
                this.showSearch = true;
            };
            this._onSearch = (event) => {
                const el = event.target;
                const inputValue = el.value.trim();
                this.view.setSearch(inputValue);
            };
            this._onSearchBlur = () => {
                if (this._searchInput.value || this.preventBlur) {
                    return;
                }
                this.showSearch = false;
            };
            this._onSearchKeydown = (event) => {
                if (event.key === 'Escape') {
                    if (this._searchInput.value) {
                        this._searchInput.value = '';
                        this.view.setSearch('');
                    }
                    else {
                        this.showSearch = false;
                    }
                }
            };
            this.preventBlur = false;
            this.#_searchInput_accessor_storage = __runInitializers(this, __searchInput_initializers, void 0);
            this.#showSearch_accessor_storage = (__runInitializers(this, __searchInput_extraInitializers), __runInitializers(this, _showSearch_initializers, false));
            __runInitializers(this, _showSearch_extraInitializers);
        }
    };
})();
export { DataViewHeaderToolsSearch };
