// @blocksuite/data-view/dist/view-presets/table/pc/header/vertical-indicator.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { unsafeCSSVarV2 } from '@blocksuite/affine-shared/theme';
import { ShadowlessElement } from '@blocksuite/block-std';
import { WithDisposable } from '@blocksuite/global/utils';
import { css, html } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { styleMap } from 'lit/directives/style-map.js';
import { startDrag } from '../../../../core/utils/drag.js';
import { getResultInRange } from '../../../../core/utils/utils.js';
let TableVerticalIndicator = (() => {
    let _classSuper = WithDisposable(ShadowlessElement);
    let _height_decorators;
    let _height_initializers = [];
    let _height_extraInitializers = [];
    let _left_decorators;
    let _left_initializers = [];
    let _left_extraInitializers = [];
    let _shadow_decorators;
    let _shadow_initializers = [];
    let _shadow_extraInitializers = [];
    let _top_decorators;
    let _top_initializers = [];
    let _top_extraInitializers = [];
    let _width_decorators;
    let _width_initializers = [];
    let _width_extraInitializers = [];
    return class TableVerticalIndicator extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _height_decorators = [property({ attribute: false })];
            _left_decorators = [property({ attribute: false })];
            _shadow_decorators = [property({ attribute: false })];
            _top_decorators = [property({ attribute: false })];
            _width_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _height_decorators, { kind: "accessor", name: "height", static: false, private: false, access: { has: obj => "height" in obj, get: obj => obj.height, set: (obj, value) => { obj.height = value; } }, metadata: _metadata }, _height_initializers, _height_extraInitializers);
            __esDecorate(this, null, _left_decorators, { kind: "accessor", name: "left", static: false, private: false, access: { has: obj => "left" in obj, get: obj => obj.left, set: (obj, value) => { obj.left = value; } }, metadata: _metadata }, _left_initializers, _left_extraInitializers);
            __esDecorate(this, null, _shadow_decorators, { kind: "accessor", name: "shadow", static: false, private: false, access: { has: obj => "shadow" in obj, get: obj => obj.shadow, set: (obj, value) => { obj.shadow = value; } }, metadata: _metadata }, _shadow_initializers, _shadow_extraInitializers);
            __esDecorate(this, null, _top_decorators, { kind: "accessor", name: "top", static: false, private: false, access: { has: obj => "top" in obj, get: obj => obj.top, set: (obj, value) => { obj.top = value; } }, metadata: _metadata }, _top_initializers, _top_extraInitializers);
            __esDecorate(this, null, _width_decorators, { kind: "accessor", name: "width", static: false, private: false, access: { has: obj => "width" in obj, get: obj => obj.width, set: (obj, value) => { obj.width = value; } }, metadata: _metadata }, _width_initializers, _width_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    data-view-table-vertical-indicator {
      position: fixed;
      left: 0;
      top: 0;
      z-index: 1;
      pointer-events: none;
    }

    .vertical-indicator {
      position: absolute;
      pointer-events: none;
      width: 1px;
      background-color: ${unsafeCSSVarV2('layer/background/hoverOverlay')};
    }

    .vertical-indicator::after {
      position: absolute;
      z-index: 1;
      width: 2px;
      height: 100%;
      content: '';
      right: 0;
      background-color: var(--affine-primary-color);
      border-radius: 1px;
    }

    .with-shadow.vertical-indicator::after {
      box-shadow: 0px 0px 8px 0px rgba(30, 150, 235, 0.35);
    }
  `; }
        render() {
            const style = styleMap({
                top: `${this.top}px`,
                left: `${this.left}px`,
                height: `${this.height}px`,
                width: `${this.width}px`,
            });
            const className = classMap({
                'with-shadow': this.shadow,
                'vertical-indicator': true,
            });
            return html ` <div class="${className}" style=${style}></div> `;
        }
        #height_accessor_storage = __runInitializers(this, _height_initializers, void 0);
        get height() { return this.#height_accessor_storage; }
        set height(value) { this.#height_accessor_storage = value; }
        #left_accessor_storage = (__runInitializers(this, _height_extraInitializers), __runInitializers(this, _left_initializers, void 0));
        get left() { return this.#left_accessor_storage; }
        set left(value) { this.#left_accessor_storage = value; }
        #shadow_accessor_storage = (__runInitializers(this, _left_extraInitializers), __runInitializers(this, _shadow_initializers, false));
        get shadow() { return this.#shadow_accessor_storage; }
        set shadow(value) { this.#shadow_accessor_storage = value; }
        #top_accessor_storage = (__runInitializers(this, _shadow_extraInitializers), __runInitializers(this, _top_initializers, void 0));
        get top() { return this.#top_accessor_storage; }
        set top(value) { this.#top_accessor_storage = value; }
        #width_accessor_storage = (__runInitializers(this, _top_extraInitializers), __runInitializers(this, _width_initializers, void 0));
        get width() { return this.#width_accessor_storage; }
        set width(value) { this.#width_accessor_storage = value; }
        constructor() {
            super(...arguments);
            __runInitializers(this, _width_extraInitializers);
        }
    };
})();
export { TableVerticalIndicator };
export const getTableGroupRect = (ele) => {
    const group = ele.closest('affine-data-view-table-group');
    if (!group) {
        return;
    }
    const groupRect = group?.getBoundingClientRect();
    const top = group
        .querySelector('.affine-database-column-header')
        ?.getBoundingClientRect().top ?? groupRect.top;
    const bottom = group.querySelector('.affine-database-block-rows')?.getBoundingClientRect()
        .bottom ?? groupRect.bottom;
    return {
        top: top,
        bottom: bottom,
    };
};
export const startDragWidthAdjustmentBar = (evt, ele, width, column) => {
    const scale = width / column.width$.value;
    const left = ele.getBoundingClientRect().left;
    const rect = getTableGroupRect(ele);
    if (!rect) {
        return;
    }
    const preview = getVerticalIndicator();
    preview.display(left, rect.top, rect.bottom - rect.top, width * scale);
    startDrag(evt, {
        onDrag: () => ({ width: column.width$.value }),
        onMove: ({ x }) => {
            const width = Math.round(getResultInRange((x - left) / scale, column.minWidth, Infinity));
            preview.display(left, rect.top, rect.bottom - rect.top, width * scale);
            return {
                width,
            };
        },
        onDrop: ({ width }) => {
            column.updateWidth(width);
        },
        onClear: () => {
            preview.remove();
        },
    });
};
let preview = null;
export const getVerticalIndicator = () => {
    if (!preview) {
        const dragBar = new TableVerticalIndicator();
        preview = {
            display(left, top, height, width = 1, shadow = false) {
                document.body.append(dragBar);
                dragBar.left = left;
                dragBar.height = height;
                dragBar.top = top;
                dragBar.width = width;
                dragBar.shadow = shadow;
            },
            remove() {
                dragBar.remove();
            },
        };
    }
    return preview;
};
