// @blocksuite/data-view/dist/core/filter/add-filter.js
import { menu, popMenu, } from '@blocksuite/affine-components/context-menu';
import { AddCursorIcon } from '@blocksuite/icons/lit';
import { renderUniLit } from '../utils/index.js';
import { firstFilterByRef, firstFilterInGroup } from './utils.js';
export const popCreateFilter = (target, props, ops) => {
    popMenu(target, {
        middleware: ops?.middleware,
        options: {
            onClose: props.onClose,
            title: {
                onBack: props.onBack,
                text: 'New filter',
            },
            items: [
                menu.group({
                    items: props.vars.value.map(v => menu.action({
                        name: v.name,
                        prefix: renderUniLit(v.icon, {}),
                        select: () => {
                            props.onSelect(firstFilterByRef(props.vars.value, {
                                type: 'ref',
                                name: v.id,
                            }));
                        },
                    })),
                }),
                menu.group({
                    name: '',
                    items: [
                        menu.action({
                            name: 'Add filter group',
                            prefix: AddCursorIcon(),
                            select: () => {
                                props.onSelect(firstFilterInGroup(props.vars.value));
                            },
                        }),
                    ],
                }),
            ],
        },
    });
};
