// @blocksuite/data-view/dist/core/filter/utils.js
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { getRefType } from '../expression/ref/ref.js';
import { filterMatcher } from './filter-fn/matcher.js';
export const firstFilterName = (vars, ref) => {
    const type = getRefType(vars, ref);
    if (!type) {
        throw new BlockSuiteError(ErrorCode.DatabaseBlockError, `can't resolve ref type`);
    }
    return filterMatcher.firstMatchedBySelfType(type)?.name;
};
export const firstFilterByRef = (vars, ref) => {
    return {
        type: 'filter',
        left: ref,
        function: firstFilterName(vars, ref),
        args: [],
    };
};
export const firstFilter = (vars) => {
    const ref = {
        type: 'ref',
        name: vars[0].id,
    };
    const filter = firstFilterName(vars, ref);
    if (!filter) {
        throw new BlockSuiteError(ErrorCode.DatabaseBlockError, `can't match any filter`);
    }
    return {
        type: 'filter',
        left: ref,
        function: filter,
        args: [],
    };
};
export const firstFilterInGroup = (vars) => {
    return {
        type: 'group',
        op: 'and',
        conditions: [firstFilter(vars)],
    };
};
export const emptyFilterGroup = {
    type: 'group',
    op: 'and',
    conditions: [],
};
