// @toeverything/theme/dist/presets/typography.js
var i = "xn1iiob", a = "xn1iiod", o = "xn1iioc", e = "xn1iioa", n = "xn1iiof", t = "xn1iioh", l = "xn1iiog", r = "xn1iioe", v = "xn1iior", x = "xn1iiot", c = "xn1iios", p = "xn1iioq", d = "xn1iiov", s = "xn1iiox", h = "xn1iiow", u = "xn1iiou", m = "xn1iion", z = "xn1iiop", E = "xn1iioo", g = "xn1iiom", I = "xn1iio9", R = "xn1iio8", b = "xn1iio1", y = "xn1iio0", f = "xn1iioj", H = "xn1iiol", _ = "xn1iiok", T = "xn1iioi", j = "xn1iio3", k = "xn1iio2", q = "xn1iio5", w = "xn1iio4", A = "xn1iio7", B = "xn1iio6";
export {
  i as bodyEmphasized,
  a as bodyEmphasizedItalic,
  o as bodyItalic,
  e as bodyRegular,
  n as calloutEmphasized,
  t as calloutEmphasizedItalic,
  l as calloutItalic,
  r as calloutRegular,
  v as caption1Emphasized,
  x as caption1EmphasizedItalic,
  c as caption1Italic,
  p as caption1Regular,
  d as caption2Emphasized,
  s as caption2EmphasizedItalic,
  h as caption2Italic,
  u as caption2Regular,
  m as footnoteEmphasized,
  z as footnoteEmphasizedItalic,
  E as footnoteItalic,
  g as footnoteRegular,
  I as headlineItalic,
  R as headlineRegular,
  b as largeTitleEmphasized,
  y as largeTitleRegular,
  f as subHeadlineEmphasized,
  H as subHeadlineEmphasizedItalic,
  _ as subHeadlineItalic,
  T as subHeadlineRegular,
  j as title1Emphasized,
  k as title1Regular,
  q as title2Emphasized,
  w as title2Regular,
  A as title3Emphasized,
  B as title3Regular
};
