// @blocksuite/data-view/dist/property-presets/select/cell-renderer.js
import { popupTargetFromElement } from '@blocksuite/affine-components/context-menu';
import { computed, signal } from '@preact/signals-core';
import { html } from 'lit/static-html.js';
import { popTagSelect } from '../../core/component/tags/multi-tag-select.js';
import { BaseCellRenderer } from '../../core/property/index.js';
import { createFromBaseCellRenderer } from '../../core/property/renderer.js';
import { createIcon } from '../../core/utils/uni-icon.js';
import { selectPropertyModelConfig, } from './define.js';
export class SelectCell extends BaseCellRenderer {
    render() {
        const value = this.value ? [this.value] : [];
        return html `
      <affine-multi-tag-view
        .value="${value}"
        .options="${this.property.data$.value.options}"
      ></affine-multi-tag-view>
    `;
    }
}
export class SelectCellEditing extends BaseCellRenderer {
    constructor() {
        super(...arguments);
        this.popTagSelect = () => {
            const value = signal(this._value);
            this._disposables.add({
                dispose: popTagSelect(popupTargetFromElement(this.querySelector('affine-multi-tag-view') ?? this), {
                    name: this.cell.property.name$.value,
                    mode: 'single',
                    options: this.options$,
                    onOptionsChange: this._onOptionsChange,
                    value: signal(this._value),
                    onChange: v => {
                        this._onChange(v);
                        value.value = v;
                    },
                    onComplete: this._editComplete,
                    minWidth: 400,
                }),
            });
        };
        this._editComplete = () => {
            this.selectCurrentCell(false);
        };
        this._onChange = ([id]) => {
            this.onChange(id);
        };
        this._onOptionsChange = (options) => {
            this.property.dataUpdate(data => {
                return {
                    ...data,
                    options,
                };
            });
        };
        this.options$ = computed(() => {
            return this.property.data$.value.options;
        });
    }
    get _value() {
        const value = this.value;
        return value ? [value] : [];
    }
    firstUpdated() {
        this.popTagSelect();
    }
    render() {
        return html `
      <affine-multi-tag-view
        .value="${this._value}"
        .options="${this.options$.value}"
      ></affine-multi-tag-view>
    `;
    }
}
export const selectPropertyConfig = selectPropertyModelConfig.createPropertyMeta({
    icon: createIcon('SingleSelectIcon'),
    cellRenderer: {
        view: createFromBaseCellRenderer(SelectCell),
        edit: createFromBaseCellRenderer(SelectCellEditing),
    },
});
