// @blocksuite/data-view/dist/core/group-by/renderer/string-group.js
import { menu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { css, html } from 'lit';
import { BaseGroup } from './base.js';
export class StringGroupView extends BaseGroup {
    constructor() {
        super(...arguments);
        this._click = () => {
            if (this.readonly) {
                return;
            }
            popMenu(popupTargetFromElement(this), {
                options: {
                    items: [
                        menu.input({
                            initialValue: this.value ?? '',
                            onChange: text => {
                                this.updateValue?.(text);
                            },
                        }),
                    ],
                },
            });
        };
    }
    static { this.styles = css `
    .data-view-group-title-string-view {
      border-radius: 8px;
      padding: 4px 8px;
      width: max-content;
      cursor: pointer;
    }

    .data-view-group-title-string-view:hover {
      background-color: var(--affine-hover-color);
    }
  `; }
    render() {
        if (!this.value) {
            return html ` <div>Ungroups</div>`;
        }
        return html ` <div
      @click="${this._click}"
      class="data-view-group-title-string-view"
    >
      ${this.value}
    </div>`;
    }
}
