// @blocksuite/affine-components/dist/caption/captioned-block-component.js
import { ThemeProvider } from '@blocksuite/affine-shared/services';
import { BlockComponent } from '@blocksuite/block-std';
import { html, nothing } from 'lit';
import { classMap } from 'lit/directives/class-map.js';
import { createRef, ref } from 'lit/directives/ref.js';
import { styleMap } from 'lit/directives/style-map.js';
import { styles } from './styles.js';
export var SelectedStyle;
(function (SelectedStyle) {
    SelectedStyle["Background"] = "Background";
    SelectedStyle["Border"] = "Border";
})(SelectedStyle || (SelectedStyle = {}));
export class CaptionedBlockComponent extends BlockComponent {
    static { this.styles = styles; }
    get captionEditor() {
        if (!this.useCaptionEditor || !this._captionEditorRef.value) {
            console.error('Oops! Please enable useCaptionEditor before accessing captionEditor');
        }
        return this._captionEditorRef.value;
    }
    constructor() {
        super();
        this.addRenderer(this._renderWithWidget);
    }
    _renderWithWidget(content) {
        const style = styleMap({
            position: 'relative',
            ...this.blockContainerStyles,
        });
        const theme = this.std.get(ThemeProvider).theme;
        const isBorder = this.selectedStyle === SelectedStyle.Border;
        return html `<div
      style=${style}
      class=${classMap({
            'affine-block-component': true,
            [theme]: true,
            border: isBorder,
        })}
    >
      ${content}
      ${this.useCaptionEditor
            ? html `<block-caption-editor
            ${ref(this._captionEditorRef)}
          ></block-caption-editor>`
            : nothing}
      ${this.selectedStyle === SelectedStyle.Background
            ? html `<affine-block-selection .block=${this}></affine-block-selection>`
            : null}
      ${this.useZeroWidth && !this.doc.readonly
            ? html `<block-zero-width .block=${this}></block-zero-width>`
            : nothing}
    </div>`;
    }
    #_captionEditorRef_accessor_storage = createRef();
    // There may be multiple block-caption-editors in a nested structure.
    get _captionEditorRef() { return this.#_captionEditorRef_accessor_storage; }
    set _captionEditorRef(value) { this.#_captionEditorRef_accessor_storage = value; }
    #blockContainerStyles_accessor_storage = undefined;
    get blockContainerStyles() { return this.#blockContainerStyles_accessor_storage; }
    set blockContainerStyles(value) { this.#blockContainerStyles_accessor_storage = value; }
    #selectedStyle_accessor_storage = SelectedStyle.Background;
    get selectedStyle() { return this.#selectedStyle_accessor_storage; }
    set selectedStyle(value) { this.#selectedStyle_accessor_storage = value; }
    #useCaptionEditor_accessor_storage = false;
    get useCaptionEditor() { return this.#useCaptionEditor_accessor_storage; }
    set useCaptionEditor(value) { this.#useCaptionEditor_accessor_storage = value; }
    #useZeroWidth_accessor_storage = false;
    get useZeroWidth() { return this.#useZeroWidth_accessor_storage; }
    set useZeroWidth(value) { this.#useZeroWidth_accessor_storage = value; }
}
