// @blocksuite/affine-block-embed/dist/embed-synced-doc-block/utils.js
import { DarkLoadingIcon, EmbedEdgelessIcon, EmbedPageIcon, LightLoadingIcon, ReloadIcon, } from '@blocksuite/affine-components/icons';
import { ColorScheme } from '@blocksuite/affine-model';
import { DarkSyncedDocDeletedBanner, DarkSyncedDocEmptyBanner, DarkSyncedDocErrorBanner, LightSyncedDocDeletedBanner, LightSyncedDocEmptyBanner, LightSyncedDocErrorBanner, SyncedDocDeletedIcon, SyncedDocErrorIcon, } from './styles.js';
export function getSyncedDocIcons(theme, editorMode) {
    if (theme === ColorScheme.Light) {
        return {
            LoadingIcon: LightLoadingIcon,
            SyncedDocIcon: editorMode === 'page' ? EmbedPageIcon : EmbedEdgelessIcon,
            SyncedDocErrorIcon,
            SyncedDocDeletedIcon,
            ReloadIcon,
            SyncedDocEmptyBanner: LightSyncedDocEmptyBanner,
            SyncedDocErrorBanner: LightSyncedDocErrorBanner,
            SyncedDocDeletedBanner: LightSyncedDocDeletedBanner,
        };
    }
    else {
        return {
            LoadingIcon: DarkLoadingIcon,
            SyncedDocIcon: editorMode === 'page' ? EmbedPageIcon : EmbedEdgelessIcon,
            SyncedDocErrorIcon,
            SyncedDocDeletedIcon,
            ReloadIcon,
            SyncedDocEmptyBanner: DarkSyncedDocEmptyBanner,
            SyncedDocErrorBanner: DarkSyncedDocErrorBanner,
            SyncedDocDeletedBanner: DarkSyncedDocDeletedBanner,
        };
    }
}
