// @blocksuite/affine-block-surface/dist/renderer/overlay.js
import { Extension } from '@blocksuite/block-std';
import { GfxControllerIdentifier, } from '@blocksuite/block-std/gfx';
import { createIdentifier } from '@blocksuite/global/di';
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
/**
 * An overlay is a layer covered on top of elements,
 * can be used for rendering non-CRDT state indicators.
 */
export class Overlay extends Extension {
    static { this.overlayName = ''; }
    constructor(gfx) {
        super();
        this.gfx = gfx;
        this._renderer = null;
    }
    static setup(di) {
        if (!this.overlayName) {
            throw new BlockSuiteError(ErrorCode.ValueNotExists, `The overlay constructor '${this.name}' should have a static 'overlayName' property.`);
        }
        di.addImpl(OverlayIdentifier(this.overlayName), this, [
            GfxControllerIdentifier,
        ]);
    }
    clear() { }
    dispose() { }
    refresh() {
        if (this._renderer) {
            this._renderer.refresh();
        }
    }
    setRenderer(renderer) {
        this._renderer = renderer;
    }
}
export const OverlayIdentifier = createIdentifier('Overlay');
