// @blocksuite/affine-components/dist/rich-text/inline/presets/nodes/link-node/affine-link.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { ParseDocUrlProvider } from '@blocksuite/affine-shared/services';
import { BLOCK_ID_ATTR, ShadowlessElement } from '@blocksuite/block-std';
import { INLINE_ROOT_ATTR, ZERO_WIDTH_SPACE, } from '@blocksuite/inline';
import { css, html } from 'lit';
import { property } from 'lit/decorators.js';
import { ref } from 'lit/directives/ref.js';
import { styleMap } from 'lit/directives/style-map.js';
import { HoverController } from '../../../../../hover/index.js';
import { RefNodeSlotsProvider } from '../../../../extension/index.js';
import { affineTextStyles } from '../affine-text.js';
import { toggleLinkPopup } from './link-popup/toggle-link-popup.js';
let AffineLink = (() => {
    let _classSuper = ShadowlessElement;
    let _delta_decorators;
    let _delta_initializers = [];
    let _delta_extraInitializers = [];
    return class AffineLink extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _delta_decorators = [property({ type: Object })];
            __esDecorate(this, null, _delta_decorators, { kind: "accessor", name: "delta", static: false, private: false, access: { has: obj => "delta" in obj, get: obj => obj.delta, set: (obj, value) => { obj.delta = value; } }, metadata: _metadata }, _delta_initializers, _delta_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    affine-link a:hover [data-v-text='true'] {
      text-decoration: underline;
    }
  `; }
        // Workaround for links not working in contenteditable div
        // see also https://stackoverflow.com/questions/12059211/how-to-make-clickable-anchor-in-contenteditable-div
        //
        // Note: We cannot use JS to directly open a new page as this may be blocked by the browser.
        //
        // Please also note that when readonly mode active,
        // this workaround is not necessary and links work normally.
        get block() {
            const block = this.inlineEditor?.rootElement.closest(`[${BLOCK_ID_ATTR}]`);
            return block;
        }
        get inlineEditor() {
            const inlineRoot = this.closest(`[${INLINE_ROOT_ATTR}]`);
            return inlineRoot?.inlineEditor;
        }
        get link() {
            return this.delta.attributes?.link ?? '';
        }
        get selfInlineRange() {
            const selfInlineRange = this.inlineEditor?.getInlineRangeFromElement(this);
            return selfInlineRange;
        }
        get std() {
            const std = this.block?.std;
            return std;
        }
        // Identify if url is an internal link
        _identify() {
            const link = this.link;
            if (!link)
                return;
            const result = this.std
                ?.getOptional(ParseDocUrlProvider)
                ?.parseDocUrl(link);
            if (!result)
                return;
            const { docId: pageId, ...params } = result;
            this._referenceInfo = { pageId, params };
        }
        _renderLink(style) {
            return html `<a
      ${ref(this._whenHover.setReference)}
      href=${this.link}
      rel="noopener noreferrer"
      target="_blank"
      style=${styleMap(style)}
      @click=${this.openLink}
      @mouseup=${this._onMouseUp}
      ><v-text .str=${this.delta.insert}></v-text
    ></a>`;
        }
        render() {
            const linkStyle = {
                color: 'var(--affine-link-color)',
                fill: 'var(--affine-link-color)',
                'text-decoration': 'none',
                cursor: 'pointer',
            };
            if (this.delta.attributes && this.delta.attributes?.code) {
                const codeStyle = affineTextStyles(this.delta.attributes);
                return html `<code style=${styleMap(codeStyle)}>
        ${this._renderLink(linkStyle)}
      </code>`;
            }
            const style = this.delta.attributes
                ? affineTextStyles(this.delta.attributes, linkStyle)
                : {};
            return this._renderLink(style);
        }
        #delta_accessor_storage;
        get delta() { return this.#delta_accessor_storage; }
        set delta(value) { this.#delta_accessor_storage = value; }
        constructor() {
            super(...arguments);
            // The link has been identified.
            this._identified = false;
            // see https://github.com/toeverything/AFFiNE/issues/1540
            this._onMouseUp = () => {
                const anchorElement = this.querySelector('a');
                if (!anchorElement || !anchorElement.isContentEditable)
                    return;
                anchorElement.contentEditable = 'false';
                setTimeout(() => {
                    anchorElement.removeAttribute('contenteditable');
                }, 0);
            };
            this._referenceInfo = null;
            this.openLink = (e) => {
                if (!this._identified) {
                    this._identified = true;
                    this._identify();
                }
                const referenceInfo = this._referenceInfo;
                if (!referenceInfo)
                    return;
                const refNodeSlotsProvider = this.std?.getOptional(RefNodeSlotsProvider);
                if (!refNodeSlotsProvider)
                    return;
                e?.preventDefault();
                refNodeSlotsProvider.docLinkClicked.emit(referenceInfo);
            };
            this._whenHover = new HoverController(this, ({ abortController }) => {
                if (this.block?.doc.readonly) {
                    return null;
                }
                if (!this.inlineEditor || !this.selfInlineRange) {
                    return null;
                }
                const selection = this.std?.selection;
                const textSelection = selection?.find('text');
                if (!!textSelection && !textSelection.isCollapsed()) {
                    return null;
                }
                const blockSelections = selection?.filter('block');
                if (blockSelections?.length) {
                    return null;
                }
                return {
                    template: toggleLinkPopup(this.inlineEditor, 'view', this.selfInlineRange, abortController, (e) => {
                        this.openLink(e);
                        abortController.abort();
                    }),
                };
            }, { enterDelay: 500 });
            this.#delta_accessor_storage = __runInitializers(this, _delta_initializers, {
                insert: ZERO_WIDTH_SPACE,
            });
            __runInitializers(this, _delta_extraInitializers);
        }
    };
})();
export { AffineLink };
