// @blocksuite/data-view/dist/widget-presets/views-bar/views-view.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popFilterableSimpleMenu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { DeleteIcon, DuplicateIcon, InfoIcon, MoreHorizontalIcon, MoveLeftIcon, MoveRightIcon, PlusIcon, } from '@blocksuite/icons/lit';
import { css, html } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { WidgetBase } from '../../core/widget/widget-base.js';
let DataViewHeaderViews = (() => {
    let _classSuper = WidgetBase;
    let _onChangeView_decorators;
    let _onChangeView_initializers = [];
    let _onChangeView_extraInitializers = [];
    return class DataViewHeaderViews extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _onChangeView_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _onChangeView_decorators, { kind: "accessor", name: "onChangeView", static: false, private: false, access: { has: obj => "onChangeView" in obj, get: obj => obj.onChangeView, set: (obj, value) => { obj.onChangeView = value; } }, metadata: _metadata }, _onChangeView_initializers, _onChangeView_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    data-view-header-views {
      height: 28px;
      display: flex;
      user-select: none;
      gap: 4px;
    }
    data-view-header-views::-webkit-scrollbar-thumb {
      width: 1px;
    }

    .database-view-button {
      height: 100%;
      cursor: pointer;
      padding: 2px 4px;
      border-radius: 4px;
      font-size: 14px;
      display: flex;
      align-items: center;
      justify-content: center;
      color: var(--affine-text-secondary-color);
      white-space: nowrap;
      max-width: 200px;
      min-width: 28px;
    }

    .database-view-button .name {
      align-items: center;
      font-size: 15px;
      line-height: 24px;
      overflow: hidden;
      text-overflow: ellipsis;
      font-weight: 500;
      padding-right: 2px;
    }

    .database-view-button .icon {
      margin-right: 6px;
      display: block;
      flex-shrink: 0;
    }

    .database-view-button .icon svg {
      width: 16px;
      height: 16px;
    }

    .database-view-button.selected {
      color: var(--affine-text-primary-color);
      background-color: var(--affine-hover-color-filled);
    }
  `; }
        get readonly() {
            return this.viewManager.readonly$.value;
        }
        getRenderer(viewId) {
            return this.dataSource.viewMetaGetById(viewId).renderer;
        }
        _clickView(event, id) {
            if (this.viewManager.currentViewId$.value !== id) {
                this.viewManager.setCurrentView(id);
                this.onChangeView?.(id);
                return;
            }
            this.openViewOption(popupTargetFromElement(event.currentTarget), id);
        }
        render() {
            return html `
      <component-overflow
        .renderItem="${this.renderViews()}"
        .renderMore="${this.renderMore}"
      ></component-overflow>
    `;
        }
        #onChangeView_accessor_storage;
        get onChangeView() { return this.#onChangeView_accessor_storage; }
        set onChangeView(value) { this.#onChangeView_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._addViewMenu = (event) => {
                popFilterableSimpleMenu(popupTargetFromElement(event.currentTarget), this.dataSource.viewMetas.map(v => {
                    return menu.action({
                        name: v.model.defaultName,
                        prefix: html `<uni-lit .uni=${v.renderer.icon}></uni-lit>`,
                        select: () => {
                            const id = this.viewManager.viewAdd(v.type);
                            this.viewManager.setCurrentView(id);
                        },
                    });
                }));
            };
            this._showMore = (event) => {
                const views = this.viewManager.views$.value;
                popFilterableSimpleMenu(popupTargetFromElement(event.currentTarget), [
                    menu.group({
                        items: views.map(id => {
                            const openViewOption = (event) => {
                                event.stopPropagation();
                                this.openViewOption(popupTargetFromElement(event.currentTarget), id);
                            };
                            const view = this.viewManager.viewGet(id);
                            return menu.action({
                                prefix: html `<uni-lit
                .uni=${this.getRenderer(id).icon}
              ></uni-lit>`,
                                name: view.name$.value ?? '',
                                label: () => html `${view.name$.value}`,
                                isSelected: this.viewManager.currentViewId$.value === id,
                                select: () => {
                                    this.viewManager.setCurrentView(id);
                                },
                                postfix: html `<div
                class="dv-hover dv-round-4"
                @click="${openViewOption}"
                style="display:flex;align-items:center;"
              >
                ${MoreHorizontalIcon()}
              </div>`,
                            });
                        }),
                    }),
                    menu.group({
                        items: this.dataSource.viewMetas.map(v => {
                            return menu.action({
                                name: `Create ${v.model.defaultName}`,
                                hide: () => this.readonly,
                                prefix: PlusIcon(),
                                select: () => {
                                    const id = this.viewManager.viewAdd(v.type);
                                    this.viewManager.setCurrentView(id);
                                },
                            });
                        }),
                    }),
                ]);
            };
            this.openViewOption = (target, id) => {
                if (this.readonly) {
                    return;
                }
                const views = this.viewManager.views$.value;
                const index = views.findIndex(v => v === id);
                const view = this.viewManager.viewGet(views[index]);
                if (!view) {
                    return;
                }
                popMenu(target, {
                    options: {
                        items: [
                            menu.input({
                                initialValue: view.name$.value,
                                onChange: text => {
                                    view.nameSet(text);
                                },
                            }),
                            menu.group({
                                items: [
                                    menu.action({
                                        name: 'Edit View',
                                        prefix: InfoIcon(),
                                        select: () => {
                                            this.closest('affine-data-view-renderer')
                                                ?.querySelector('data-view-header-tools-view-options')
                                                ?.openMoreAction(target);
                                        },
                                    }),
                                    menu.action({
                                        name: 'Move Left',
                                        hide: () => index === 0,
                                        prefix: MoveLeftIcon(),
                                        select: () => {
                                            const targetId = views[index - 1];
                                            this.viewManager.moveTo(id, targetId ? { before: true, id: targetId } : 'start');
                                        },
                                    }),
                                    menu.action({
                                        name: 'Move Right',
                                        prefix: MoveRightIcon(),
                                        hide: () => index === views.length - 1,
                                        select: () => {
                                            const targetId = views[index + 1];
                                            this.viewManager.moveTo(id, targetId ? { before: false, id: targetId } : 'end');
                                        },
                                    }),
                                ],
                            }),
                            menu.group({
                                items: [
                                    menu.action({
                                        name: 'Duplicate',
                                        prefix: DuplicateIcon(),
                                        select: () => {
                                            this.viewManager.viewDuplicate(id);
                                        },
                                    }),
                                    menu.action({
                                        name: 'Delete',
                                        prefix: DeleteIcon(),
                                        select: () => {
                                            view.delete();
                                        },
                                        class: { 'delete-item': true },
                                    }),
                                ],
                            }),
                        ],
                    },
                });
            };
            this.renderMore = (count) => {
                const views = this.viewManager.views$.value;
                if (count === views.length) {
                    if (this.readonly) {
                        return;
                    }
                    return html `<div
        class="database-view-button dv-icon-16 dv-hover"
        data-testid="database-add-view-button"
        @click="${this._addViewMenu}"
      >
        ${PlusIcon()}
      </div>`;
                }
                return html `
      <div class="database-view-button dv-hover" @click="${this._showMore}">
        ${views.length - count} More
      </div>
    `;
            };
            this.renderViews = () => {
                const views = this.viewManager.views$.value;
                return views.map(id => () => {
                    const classList = classMap({
                        'database-view-button': true,
                        'dv-hover': true,
                        selected: this.viewManager.currentViewId$.value === id,
                    });
                    const view = this.viewManager.viewDataGet(id);
                    return html `
        <div
          class="${classList}"
          style="margin-right: 4px;"
          @click="${(event) => this._clickView(event, id)}"
        >
          <uni-lit class="icon" .uni="${this.getRenderer(id).icon}"></uni-lit>
          <div class="name">${view?.name}</div>
        </div>
      `;
                });
            };
            this.#onChangeView_accessor_storage = __runInitializers(this, _onChangeView_initializers, void 0);
            __runInitializers(this, _onChangeView_extraInitializers);
        }
    };
})();
export { DataViewHeaderViews };
