// @blocksuite/block-std/dist/gfx/model/gfx-block-model.js
import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
import { Bound, deserializeXYWH, getBoundWithRotation, getPointsFromBoundWithRotation, linePolygonIntersects, PointLocation, polygonGetPointTangent, polygonNearestPoint, rotatePoints, } from '@blocksuite/global/utils';
import { BlockModel } from '@blocksuite/store';
import { isLockedByAncestorImpl, isLockedBySelfImpl, isLockedImpl, lockElementImpl, unlockElementImpl, } from '../../utils/tree.js';
/**
 * The graphic block model that can be rendered in the graphics mode.
 * All the graphic block model should extend this class.
 * You can use `GfxCompatibleBlockModel` to convert a BlockModel to a subclass that extends it.
 */
export class GfxBlockElementModel extends BlockModel {
    constructor() {
        super(...arguments);
        this._cacheDeserKey = null;
        this._cacheDeserXYWH = null;
        this._externalXYWH = undefined;
        this.connectable = true;
        /**
         * Defines the extension of the response area beyond the element's bounding box.
         * This tuple specifies the horizontal and vertical margins to be added to the element's [x, y, width, height].
         *
         * The first value represents the horizontal extension (added to both left and right sides),
         * and the second value represents the vertical extension (added to both top and bottom sides).
         *
         * The response area is computed as:
         * `[x - horizontal, y - vertical, width + 2 * horizontal, height + 2 * vertical]`.
         *
         * Example:
         * - Bounding box: `[0, 0, 100, 100]`, `responseExtension: [10, 20]`
         *   Resulting response area: `[-10, -20, 120, 140]`.
         * - `responseExtension: [0, 0]` keeps the response area equal to the bounding box.
         */
        this.responseExtension = [0, 0];
        this.rotate = 0;
    }
    get deserializedXYWH() {
        if (this._cacheDeserKey !== this.xywh || !this._cacheDeserXYWH) {
            this._cacheDeserKey = this.xywh;
            this._cacheDeserXYWH = deserializeXYWH(this.xywh);
        }
        return this._cacheDeserXYWH;
    }
    get elementBound() {
        return Bound.from(getBoundWithRotation(this));
    }
    get externalBound() {
        return this._externalXYWH ? Bound.deserialize(this._externalXYWH) : null;
    }
    get externalXYWH() {
        return this._externalXYWH;
    }
    set externalXYWH(xywh) {
        this._externalXYWH = xywh;
    }
    get group() {
        if (!this.surface)
            return null;
        return this.surface.getGroup(this.id) ?? null;
    }
    get groups() {
        if (!this.surface)
            return [];
        return this.surface.getGroups(this.id);
    }
    get h() {
        return this.deserializedXYWH[3];
    }
    get responseBound() {
        return this.elementBound.expand(this.responseExtension);
    }
    get surface() {
        const result = this.doc.getBlocksByFlavour('affine:surface');
        if (result.length === 0)
            return null;
        return result[0].model;
    }
    get w() {
        return this.deserializedXYWH[2];
    }
    get x() {
        return this.deserializedXYWH[0];
    }
    get y() {
        return this.deserializedXYWH[1];
    }
    containsBound(bounds) {
        const bound = Bound.deserialize(this.xywh);
        const points = getPointsFromBoundWithRotation({
            x: bound.x,
            y: bound.y,
            w: bound.w,
            h: bound.h,
            rotate: this.rotate,
        });
        return points.some(point => bounds.containsPoint(point));
    }
    getLineIntersections(start, end) {
        const bound = Bound.deserialize(this.xywh);
        return linePolygonIntersects(start, end, rotatePoints(bound.points, bound.center, this.rotate ?? 0));
    }
    getNearestPoint(point) {
        const bound = Bound.deserialize(this.xywh);
        return polygonNearestPoint(rotatePoints(bound.points, bound.center, this.rotate ?? 0), point);
    }
    getRelativePointLocation(relativePoint) {
        const bound = Bound.deserialize(this.xywh);
        const point = bound.getRelativePoint(relativePoint);
        const rotatePoint = rotatePoints([point], bound.center, this.rotate ?? 0)[0];
        const points = rotatePoints(bound.points, bound.center, this.rotate ?? 0);
        const tangent = polygonGetPointTangent(points, rotatePoint);
        return new PointLocation(rotatePoint, tangent);
    }
    includesPoint(x, y, opt, __) {
        const bound = opt.useElementBound ? this.elementBound : this.responseBound;
        return bound.isPointInBound([x, y], 0);
    }
    intersectsBound(bound) {
        return (this.containsBound(bound) ||
            bound.points.some((point, i, points) => this.getLineIntersections(point, points[(i + 1) % points.length])));
    }
    isLocked() {
        return isLockedImpl(this);
    }
    isLockedByAncestor() {
        return isLockedByAncestorImpl(this);
    }
    isLockedBySelf() {
        return isLockedBySelfImpl(this);
    }
    lock() {
        lockElementImpl(this.doc, this);
    }
    unlock() {
        unlockElementImpl(this.doc, this);
    }
}
/**
 * Convert a BlockModel to a GfxBlockElementModel.
 * @param BlockModelSuperClass The BlockModel class to be converted.
 * @returns The returned class is a subclass of the GfxBlockElementModel class and the given BlockModelSuperClass.
 */
export function GfxCompatibleBlockModel(BlockModelSuperClass) {
    if (BlockModelSuperClass === BlockModel) {
        return GfxBlockElementModel;
    }
    else {
        let currentClass = BlockModelSuperClass;
        while (Object.getPrototypeOf(currentClass.prototype) !== BlockModel.prototype &&
            Object.getPrototypeOf(currentClass.prototype) !== null) {
            currentClass = Object.getPrototypeOf(currentClass.prototype).constructor;
        }
        if (Object.getPrototypeOf(currentClass.prototype) === null) {
            throw new BlockSuiteError(ErrorCode.GfxBlockElementError, 'The SuperClass is not a subclass of BlockModel');
        }
        Object.setPrototypeOf(currentClass.prototype, GfxBlockElementModel.prototype);
    }
    return BlockModelSuperClass;
}
