// @blocksuite/data-view/dist/core/component/overflow/overflow.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { css, html } from 'lit';
import { property, query, queryAll, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { repeat } from 'lit/directives/repeat.js';
let Overflow = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _items_decorators;
    let _items_initializers = [];
    let _items_extraInitializers = [];
    let _more_decorators;
    let _more_initializers = [];
    let _more_extraInitializers = [];
    let _renderCount_decorators;
    let _renderCount_initializers = [];
    let _renderCount_extraInitializers = [];
    let _renderItem_decorators;
    let _renderItem_initializers = [];
    let _renderItem_extraInitializers = [];
    let _renderMore_decorators;
    let _renderMore_initializers = [];
    let _renderMore_extraInitializers = [];
    return class Overflow extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _items_decorators = [queryAll(':scope > .component-overflow-item')];
            _more_decorators = [query(':scope > .component-overflow-more')];
            _renderCount_decorators = [state()];
            _renderItem_decorators = [property({ attribute: false })];
            _renderMore_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _items_decorators, { kind: "accessor", name: "items", static: false, private: false, access: { has: obj => "items" in obj, get: obj => obj.items, set: (obj, value) => { obj.items = value; } }, metadata: _metadata }, _items_initializers, _items_extraInitializers);
            __esDecorate(this, null, _more_decorators, { kind: "accessor", name: "more", static: false, private: false, access: { has: obj => "more" in obj, get: obj => obj.more, set: (obj, value) => { obj.more = value; } }, metadata: _metadata }, _more_initializers, _more_extraInitializers);
            __esDecorate(this, null, _renderCount_decorators, { kind: "accessor", name: "renderCount", static: false, private: false, access: { has: obj => "renderCount" in obj, get: obj => obj.renderCount, set: (obj, value) => { obj.renderCount = value; } }, metadata: _metadata }, _renderCount_initializers, _renderCount_extraInitializers);
            __esDecorate(this, null, _renderItem_decorators, { kind: "accessor", name: "renderItem", static: false, private: false, access: { has: obj => "renderItem" in obj, get: obj => obj.renderItem, set: (obj, value) => { obj.renderItem = value; } }, metadata: _metadata }, _renderItem_initializers, _renderItem_extraInitializers);
            __esDecorate(this, null, _renderMore_decorators, { kind: "accessor", name: "renderMore", static: false, private: false, access: { has: obj => "renderMore" in obj, get: obj => obj.renderMore, set: (obj, value) => { obj.renderMore = value; } }, metadata: _metadata }, _renderMore_initializers, _renderMore_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    component-overflow {
      display: flex;
      flex-wrap: wrap;
      width: 100%;
      position: relative;
    }

    .component-overflow-item {
    }
    .component-overflow-item.hidden {
      opacity: 0;
      pointer-events: none;
      position: absolute;
    }
  `; }
        adjustStyle() {
            if (this.frameId) {
                cancelAnimationFrame(this.frameId);
            }
            this.frameId = requestAnimationFrame(() => {
                this.doAdjustStyle();
            });
        }
        connectedCallback() {
            super.connectedCallback();
            const resize = new ResizeObserver(() => {
                this.adjustStyle();
            });
            resize.observe(this);
            this.disposables.add(() => {
                resize.unobserve(this);
            });
        }
        doAdjustStyle() {
            const moreWidth = this.more.getBoundingClientRect().width;
            this.widthList[this.renderCount] = moreWidth;
            const containerWidth = this.getBoundingClientRect().width;
            let width = 0;
            for (let i = 0; i < this.items.length; i++) {
                const itemWidth = this.items[i].getBoundingClientRect().width;
                // Try to calculate the width occupied by rendering n+1 items;
                // if it exceeds the limit, render n items(in i++ round).
                const totalWidth = width + itemWidth + (this.widthList[i + 1] ?? moreWidth);
                if (totalWidth > containerWidth) {
                    this.renderCount = i;
                    return;
                }
                width += itemWidth;
            }
            this.renderCount = this.items.length;
        }
        render() {
            return html `
      ${repeat(this.renderItem, (render, index) => {
                const className = classMap({
                    'component-overflow-item': true,
                    hidden: index >= this.renderCount,
                });
                return html `<div class="${className}">${render()}</div>`;
            })}
      <div class="component-overflow-more">
        ${this.renderMore(this.renderCount)}
      </div>
    `;
        }
        updated(_changedProperties) {
            super.updated(_changedProperties);
            this.adjustStyle();
        }
        #items_accessor_storage;
        get items() { return this.#items_accessor_storage; }
        set items(value) { this.#items_accessor_storage = value; }
        #more_accessor_storage;
        get more() { return this.#more_accessor_storage; }
        set more(value) { this.#more_accessor_storage = value; }
        #renderCount_accessor_storage;
        get renderCount() { return this.#renderCount_accessor_storage; }
        set renderCount(value) { this.#renderCount_accessor_storage = value; }
        #renderItem_accessor_storage;
        get renderItem() { return this.#renderItem_accessor_storage; }
        set renderItem(value) { this.#renderItem_accessor_storage = value; }
        #renderMore_accessor_storage;
        get renderMore() { return this.#renderMore_accessor_storage; }
        set renderMore(value) { this.#renderMore_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.frameId = undefined;
            this.widthList = [];
            this.#items_accessor_storage = __runInitializers(this, _items_initializers, void 0);
            this.#more_accessor_storage = (__runInitializers(this, _items_extraInitializers), __runInitializers(this, _more_initializers, void 0));
            this.#renderCount_accessor_storage = (__runInitializers(this, _more_extraInitializers), __runInitializers(this, _renderCount_initializers, 0));
            this.#renderItem_accessor_storage = (__runInitializers(this, _renderCount_extraInitializers), __runInitializers(this, _renderItem_initializers, void 0));
            this.#renderMore_accessor_storage = (__runInitializers(this, _renderItem_extraInitializers), __runInitializers(this, _renderMore_initializers, void 0));
            __runInitializers(this, _renderMore_extraInitializers);
        }
    };
})();
export { Overflow };
