// @blocksuite/data-view/dist/view-presets/table/mobile/column-header.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { unsafeCSSVarV2 } from '@blocksuite/affine-shared/theme';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { DeleteIcon, DuplicateIcon, InsertLeftIcon, InsertRightIcon, MoveLeftIcon, MoveRightIcon, ViewIcon, } from '@blocksuite/icons/lit';
import { css } from 'lit';
import { property } from 'lit/decorators.js';
import { styleMap } from 'lit/directives/style-map.js';
import { html } from 'lit/static-html.js';
import { inputConfig, typeConfig } from '../../../core/common/property-menu.js';
import { numberFormats } from '../../../property-presets/number/utils/formats.js';
import { DEFAULT_COLUMN_TITLE_HEIGHT } from '../consts.js';
let MobileTableColumnHeader = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _column_decorators;
    let _column_initializers = [];
    let _column_extraInitializers = [];
    let _tableViewManager_decorators;
    let _tableViewManager_initializers = [];
    let _tableViewManager_extraInitializers = [];
    return class MobileTableColumnHeader extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _column_decorators = [property({ attribute: false })];
            _tableViewManager_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _column_decorators, { kind: "accessor", name: "column", static: false, private: false, access: { has: obj => "column" in obj, get: obj => obj.column, set: (obj, value) => { obj.column = value; } }, metadata: _metadata }, _column_initializers, _column_extraInitializers);
            __esDecorate(this, null, _tableViewManager_decorators, { kind: "accessor", name: "tableViewManager", static: false, private: false, access: { has: obj => "tableViewManager" in obj, get: obj => obj.tableViewManager, set: (obj, value) => { obj.tableViewManager = value; } }, metadata: _metadata }, _tableViewManager_initializers, _tableViewManager_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .mobile-table-column-header {
      display: flex;
      padding: 6px;
      gap: 6px;
      align-items: center;
    }

    .mobile-table-column-header-icon {
      font-size: 18px;
      color: ${unsafeCSSVarV2('database/textSecondary')};
      display: flex;
      align-items: center;
    }

    .mobile-table-column-header-name {
      font-weight: 500;
      font-size: 14px;
      color: ${unsafeCSSVarV2('database/textSecondary')};
    }
  `; }
        popMenu(ele) {
            const enableNumberFormatting = this.tableViewManager.featureFlags$.value.enable_number_formatting;
            popMenu(popupTargetFromElement(ele ?? this), {
                options: {
                    title: {
                        text: 'Property settings',
                    },
                    items: [
                        inputConfig(this.column),
                        typeConfig(this.column),
                        // Number format begin
                        ...(enableNumberFormatting
                            ? [
                                menu.subMenu({
                                    name: 'Number Format',
                                    hide: () => !this.column.dataUpdate ||
                                        this.column.type$.value !== 'number',
                                    options: {
                                        title: {
                                            text: 'Number Format',
                                        },
                                        items: [
                                            numberFormatConfig(this.column),
                                            ...numberFormats.map(format => {
                                                const data = this.column.data$.value;
                                                return menu.action({
                                                    isSelected: data.format === format.type,
                                                    prefix: html `<span
                            style="font-size: var(--affine-font-base); scale: 1.2;"
                            >${format.symbol}</span
                          >`,
                                                    name: format.label,
                                                    select: () => {
                                                        if (data.format === format.type)
                                                            return;
                                                        this.column.dataUpdate(() => ({
                                                            format: format.type,
                                                        }));
                                                    },
                                                });
                                            }),
                                        ],
                                    },
                                }),
                            ]
                            : []),
                        // Number format end
                        menu.group({
                            items: [
                                menu.action({
                                    name: 'Hide In View',
                                    prefix: ViewIcon(),
                                    hide: () => this.column.hide$.value ||
                                        this.column.type$.value === 'title',
                                    select: () => {
                                        this.column.hideSet(true);
                                    },
                                }),
                            ],
                        }),
                        menu.group({
                            items: [
                                menu.action({
                                    name: 'Insert Left Column',
                                    prefix: InsertLeftIcon(),
                                    select: () => {
                                        this.tableViewManager.propertyAdd({
                                            id: this.column.id,
                                            before: true,
                                        });
                                        Promise.resolve()
                                            .then(() => {
                                            const pre = this.previousElementSibling?.previousElementSibling;
                                            if (pre instanceof MobileTableColumnHeader) {
                                                pre.editTitle();
                                                pre.scrollIntoView({
                                                    inline: 'nearest',
                                                    block: 'nearest',
                                                });
                                            }
                                        })
                                            .catch(console.error);
                                    },
                                }),
                                menu.action({
                                    name: 'Insert Right Column',
                                    prefix: InsertRightIcon(),
                                    select: () => {
                                        this.tableViewManager.propertyAdd({
                                            id: this.column.id,
                                            before: false,
                                        });
                                        Promise.resolve()
                                            .then(() => {
                                            const next = this.nextElementSibling?.nextElementSibling;
                                            if (next instanceof MobileTableColumnHeader) {
                                                next.editTitle();
                                                next.scrollIntoView({
                                                    inline: 'nearest',
                                                    block: 'nearest',
                                                });
                                            }
                                        })
                                            .catch(console.error);
                                    },
                                }),
                                menu.action({
                                    name: 'Move Left',
                                    prefix: MoveLeftIcon(),
                                    hide: () => this.column.isFirst,
                                    select: () => {
                                        const preId = this.tableViewManager.propertyPreGet(this.column.id)?.id;
                                        if (!preId) {
                                            return;
                                        }
                                        this.tableViewManager.propertyMove(this.column.id, {
                                            id: preId,
                                            before: true,
                                        });
                                    },
                                }),
                                menu.action({
                                    name: 'Move Right',
                                    prefix: MoveRightIcon(),
                                    hide: () => this.column.isLast,
                                    select: () => {
                                        const nextId = this.tableViewManager.propertyNextGet(this.column.id)?.id;
                                        if (!nextId) {
                                            return;
                                        }
                                        this.tableViewManager.propertyMove(this.column.id, {
                                            id: nextId,
                                            before: false,
                                        });
                                    },
                                }),
                            ],
                        }),
                        menu.group({
                            items: [
                                menu.action({
                                    name: 'Duplicate',
                                    prefix: DuplicateIcon(),
                                    hide: () => !this.column.duplicate || this.column.type$.value === 'title',
                                    select: () => {
                                        this.column.duplicate?.();
                                    },
                                }),
                                menu.action({
                                    name: 'Delete',
                                    prefix: DeleteIcon(),
                                    hide: () => !this.column.delete || this.column.type$.value === 'title',
                                    select: () => {
                                        this.column.delete?.();
                                    },
                                    class: {
                                        'delete-item': true,
                                    },
                                }),
                            ],
                        }),
                    ],
                },
            });
        }
        render() {
            const column = this.column;
            const style = styleMap({
                height: DEFAULT_COLUMN_TITLE_HEIGHT + 'px',
            });
            return html `
      <div
        style=${style}
        class="mobile-table-column-header"
        @click="${this._clickColumn}"
      >
        <uni-lit
          class="mobile-table-column-header-icon"
          .uni="${column.icon}"
        ></uni-lit>
        <div class="mobile-table-column-header-name">${column.name$.value}</div>
      </div>
    `;
        }
        #column_accessor_storage;
        get column() { return this.#column_accessor_storage; }
        set column(value) { this.#column_accessor_storage = value; }
        #tableViewManager_accessor_storage;
        get tableViewManager() { return this.#tableViewManager_accessor_storage; }
        set tableViewManager(value) { this.#tableViewManager_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this._clickColumn = () => {
                if (this.tableViewManager.readonly$.value) {
                    return;
                }
                this.popMenu();
            };
            this.editTitle = () => {
                this._clickColumn();
            };
            this.#column_accessor_storage = __runInitializers(this, _column_initializers, void 0);
            this.#tableViewManager_accessor_storage = (__runInitializers(this, _column_extraInitializers), __runInitializers(this, _tableViewManager_initializers, void 0));
            __runInitializers(this, _tableViewManager_extraInitializers);
        }
    };
})();
export { MobileTableColumnHeader };
function numberFormatConfig(column) {
    return () => html ` <affine-database-number-format-bar
      .column="${column}"
    ></affine-database-number-format-bar>`;
}
