// @blocksuite/data-view/dist/view-presets/table/mobile/menu.js
import { menu, popFilterableSimpleMenu, } from '@blocksuite/affine-components/context-menu';
import { DeleteIcon, ExpandFullIcon } from '@blocksuite/icons/lit';
export const popMobileRowMenu = (target, rowId, dataViewEle, view) => {
    popFilterableSimpleMenu(target, [
        menu.group({
            items: [
                menu.action({
                    name: 'Expand Row',
                    prefix: ExpandFullIcon(),
                    select: () => {
                        dataViewEle.openDetailPanel({
                            view: view,
                            rowId: rowId,
                        });
                    },
                }),
            ],
        }),
        menu.group({
            name: '',
            items: [
                menu.action({
                    name: 'Delete Row',
                    class: { 'delete-item': true },
                    prefix: DeleteIcon(),
                    select: () => {
                        view.rowDelete([rowId]);
                    },
                }),
            ],
        }),
    ]);
};
