// @blocksuite/data-view/dist/widget-presets/quick-setting-bar/sort/root-panel.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { menu, popMenu, popupTargetFromElement, } from '@blocksuite/affine-components/context-menu';
import { unsafeCSSVarV2 } from '@blocksuite/affine-shared/theme';
import { ShadowlessElement } from '@blocksuite/block-std';
import { SignalWatcher, WithDisposable } from '@blocksuite/global/utils';
import { ArrowDownSmallIcon, CloseIcon, DeleteIcon, PlusIcon, } from '@blocksuite/icons/lit';
import { computed } from '@preact/signals-core';
import { css, html } from 'lit';
import { property } from 'lit/decorators.js';
import { keyed } from 'lit/directives/keyed.js';
import { repeat } from 'lit/directives/repeat.js';
import { renderUniLit } from '../../../core/index.js';
import { popCreateSort } from '../../../core/sort/add-sort.js';
import { dragHandler } from '../../../core/utils/wc-dnd/dnd-context.js';
import { defaultActivators } from '../../../core/utils/wc-dnd/sensors/index.js';
import { createSortContext, sortable, } from '../../../core/utils/wc-dnd/sort/sort-context.js';
import { verticalListSortingStrategy } from '../../../core/utils/wc-dnd/sort/strategies/index.js';
let SortRootView = (() => {
    let _classSuper = SignalWatcher(WithDisposable(ShadowlessElement));
    let _sortUtils_decorators;
    let _sortUtils_initializers = [];
    let _sortUtils_extraInitializers = [];
    return class SortRootView extends _classSuper {
        constructor() {
            super(...arguments);
            this.#sortUtils_accessor_storage = __runInitializers(this, _sortUtils_initializers, void 0);
            this.items$ = (__runInitializers(this, _sortUtils_extraInitializers), computed(() => {
                return this.sortUtils.sortList$.value.map(v => v.ref.name);
            }));
            this.sortContext = createSortContext({
                activators: defaultActivators,
                container: this,
                onDragEnd: evt => {
                    const over = evt.over;
                    if (over) {
                        const list = this.sortUtils.sortList$.value;
                        this.sortUtils.move(list.findIndex(v => v.ref.name === evt.active.id), list.findIndex(v => v.ref.name === over.id));
                    }
                },
                modifiers: [
                    ({ transform }) => {
                        return {
                            ...transform,
                            x: 0,
                        };
                    },
                ],
                items: this.items$,
                strategy: verticalListSortingStrategy,
            });
        }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _sortUtils_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _sortUtils_decorators, { kind: "accessor", name: "sortUtils", static: false, private: false, access: { has: obj => "sortUtils" in obj, get: obj => obj.sortUtils, set: (obj, value) => { obj.sortUtils = value; } }, metadata: _metadata }, _sortUtils_initializers, _sortUtils_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .sort-root-container {
      margin-bottom: 8px;
      gap: 8px;
      display: flex;
      flex-direction: column;
    }

    .sort-item {
      display: flex;
      align-items: center;
    }
  `; }
        #sortUtils_accessor_storage;
        get sortUtils() { return this.#sortUtils_accessor_storage; }
        set sortUtils(value) { this.#sortUtils_accessor_storage = value; }
        render() {
            const list = this.sortUtils.sortList$.value;
            return html `
      <div class="sort-root-container">
        ${repeat(list, (sort, index) => {
                const id = sort.ref.name;
                const variable = this.sortUtils.vars$.value.find(v => v.id === id);
                let content;
                const deleteRule = () => {
                    this.sortUtils.remove(index);
                };
                const changeRule = (rule) => {
                    this.sortUtils.change(index, rule);
                };
                if (!variable) {
                    content = html `
              <data-view-component-button
                style="color: var(--affine-error-color);border-color: color: var(--affine-error-color)"
                @click="${deleteRule}"
                .text="${html `This rule is invalid, click to delete`}"
              ></data-view-component-button>
            `;
                }
                else {
                    const descName = sort.desc ? 'Descending' : 'Ascending';
                    const clickField = (event) => {
                        popMenu(popupTargetFromElement(event.currentTarget), {
                            options: {
                                items: this.sortUtils.vars$.value.map(v => {
                                    return menu.action({
                                        name: v.name,
                                        prefix: renderUniLit(v.icon),
                                        isSelected: v.id === id,
                                        select: () => {
                                            changeRule({
                                                ...sort,
                                                ref: { type: 'ref', name: v.id },
                                            });
                                        },
                                    });
                                }),
                            },
                        });
                    };
                    const clickOrder = (event) => {
                        popMenu(popupTargetFromElement(event.currentTarget), {
                            options: {
                                items: [false, true].map(desc => {
                                    return menu.action({
                                        name: desc ? 'Descending' : 'Ascending',
                                        isSelected: desc === sort.desc,
                                        select: () => {
                                            changeRule({ ...sort, desc });
                                        },
                                    });
                                }),
                            },
                        });
                    };
                    content = html `
              <data-view-component-button
                style="margin-right: 6px;margin-left: 4px"
                @click="${clickField}"
                .icon="${renderUniLit(variable.icon)}"
                .text="${variable.name}"
                .postfix="${ArrowDownSmallIcon()}"
              ></data-view-component-button>
              <data-view-component-button
                @click="${clickOrder}"
                .text="${html ` <div style="padding: 0 4px">${descName}</div>`}"
                .postfix="${ArrowDownSmallIcon()}"
              ></data-view-component-button>
            `;
                }
                return keyed(id, html `
              <div
                ${sortable(id)}
                class='sort-item'
              >
                <div style='display: flex;align-items: center;flex:1;margin-right: 16px;'>
                  <div
                    ${dragHandler(id)}
                    style='border-radius:2px;cursor:pointer;width: 4px;height: 12px;background-color: ${unsafeCSSVarV2('button/grabber/default')}'
                  ></div>
                  ${content}
                </div>
                <div
                  @click='${deleteRule}'
                  style='padding: 2px;display: flex;align-items: center;border-radius: 2px;color:${unsafeCSSVarV2('icon/primary')}'
                  class='dv-hover dv-rounded'>${CloseIcon({ width: '16px', height: '16px' })}
                </div>
              </div>
              </div>
            `);
            })}
      </div>
    `;
        }
    };
})();
export { SortRootView };
export const popSortRoot = (target, props) => {
    const sortUtils = props.sortUtils;
    popMenu(target, {
        options: {
            title: props.title,
            items: [
                () => {
                    return html ` <sort-root-view
            .sortUtils="${sortUtils}"
          ></sort-root-view>`;
                },
                menu.action({
                    name: 'Add sort',
                    prefix: PlusIcon(),
                    select: ele => {
                        popCreateSort(popupTargetFromElement(ele), {
                            sortUtils: props.sortUtils,
                        });
                        return false;
                    },
                }),
                menu.action({
                    name: 'Delete',
                    class: { 'delete-item': true },
                    prefix: DeleteIcon(),
                    select: () => {
                        props.sortUtils.removeAll();
                    },
                }),
            ],
        },
    });
};
