// @blocksuite/affine-block-surface/dist/adapters/plain-text/plain-text.js
import { BlockPlainTextAdapterExtension, } from '@blocksuite/affine-shared/adapters';
import { getMindMapNodeMap } from '../utils/mindmap.js';
import { PlainTextElementModelAdapter } from './element-adapter/index.js';
export const surfaceBlockPlainTextAdapterMatcher = {
    flavour: 'affine:surface',
    toMatch: () => false,
    fromMatch: o => o.node.flavour === 'affine:surface',
    toBlockSnapshot: {},
    fromBlockSnapshot: {
        enter: (_, context) => {
            context.walkerContext.skipAllChildren();
        },
    },
};
export const SurfaceBlockPlainTextAdapterExtension = BlockPlainTextAdapterExtension(surfaceBlockPlainTextAdapterMatcher);
export const edgelessSurfaceBlockPlainTextAdapterMatcher = {
    flavour: 'affine:surface',
    toMatch: () => false,
    fromMatch: o => o.node.flavour === 'affine:surface',
    toBlockSnapshot: {},
    fromBlockSnapshot: {
        enter: (o, context) => {
            const { walkerContext } = context;
            const plainTextElementModelAdapter = new PlainTextElementModelAdapter();
            if ('elements' in o.node.props) {
                const elements = o.node.props.elements;
                // Get all the node maps of mindMap elements
                const mindMapArray = Object.entries(elements)
                    .filter(([_, element]) => element.type === 'mindmap')
                    .map(([_, element]) => getMindMapNodeMap(element));
                walkerContext.setGlobalContext('surface:mindMap:nodeMapArray', mindMapArray);
                Object.entries(o.node.props.elements).forEach(([_, element]) => {
                    const plainText = plainTextElementModelAdapter.fromElementModel(element, { walkerContext, elements });
                    if (plainText) {
                        context.textBuffer.content += plainText + '\n';
                    }
                });
            }
        },
    },
};
export const EdgelessSurfaceBlockPlainTextAdapterExtension = BlockPlainTextAdapterExtension(edgelessSurfaceBlockPlainTextAdapterMatcher);
