// @blocksuite/affine-block-surface/dist/adapters/utils/text.js
import { buildMindMapTree } from './mindmap.js';
export function getShapeType(elementModel) {
    let shapeType = '';
    if (elementModel.type !== 'shape') {
        return shapeType;
    }
    if ('shapeType' in elementModel &&
        typeof elementModel.shapeType === 'string') {
        shapeType = elementModel.shapeType;
    }
    return shapeType;
}
export function getShapeText(elementModel) {
    let text = '';
    if (elementModel.type !== 'shape') {
        return text;
    }
    if ('text' in elementModel &&
        typeof elementModel.text === 'object' &&
        elementModel.text) {
        let delta = [];
        if ('delta' in elementModel.text) {
            delta = elementModel.text.delta;
        }
        text = delta.map(d => d.insert).join('');
    }
    return text;
}
export function getConnectorText(elementModel) {
    let text = '';
    if (elementModel.type !== 'connector') {
        return text;
    }
    if ('text' in elementModel &&
        typeof elementModel.text === 'object' &&
        elementModel.text) {
        let delta = [];
        if ('delta' in elementModel.text) {
            delta = elementModel.text.delta;
        }
        text = delta.map(d => d.insert).join('');
    }
    return text;
}
export function getGroupTitle(elementModel) {
    let title = '';
    if (elementModel.type !== 'group') {
        return title;
    }
    if ('title' in elementModel &&
        typeof elementModel.title === 'object' &&
        elementModel.title) {
        let delta = [];
        if ('delta' in elementModel.title) {
            delta = elementModel.title.delta;
        }
        title = delta.map(d => d.insert).join('');
    }
    return title;
}
export function getTextElementText(elementModel) {
    let text = '';
    if (elementModel.type !== 'text') {
        return text;
    }
    if ('text' in elementModel &&
        typeof elementModel.text === 'object' &&
        elementModel.text) {
        let delta = [];
        if ('delta' in elementModel.text) {
            delta = elementModel.text.delta;
        }
        text = delta.map(d => d.insert).join('');
    }
    return text;
}
/**
 * traverse the mindMapTree and construct the content string
 * like:
 * - Root
 *   - Child 1
 *     - Child 1.1
 *     - Child 1.2
 *   - Child 2
 *     - Child 2.1
 *     - Child 2.2
 *   - Child 3
 *     - Child 3.1
 *     - Child 3.2
 * @param elementModel - the mindmap element model
 * @param elements - the elements map
 * @returns the mindmap tree text
 */
export function getMindMapTreeText(elementModel, elements, options = {
    prefix: ' ',
    repeat: 2,
}) {
    let mindMapContent = '';
    if (elementModel.type !== 'mindmap') {
        return mindMapContent;
    }
    const mindMapTree = buildMindMapTree(elementModel);
    if (!mindMapTree) {
        return mindMapContent;
    }
    let layer = 0;
    const traverseMindMapTree = (node, prefix, repeat) => {
        const shapeElement = elements[node.id];
        const shapeText = getShapeText(shapeElement);
        if (shapeElement) {
            mindMapContent += `${prefix.repeat(layer * repeat)}- ${shapeText}\n`;
        }
        node.children.forEach(child => {
            layer++;
            traverseMindMapTree(child, prefix, repeat);
            layer--;
        });
    };
    traverseMindMapTree(mindMapTree, options.prefix, options.repeat);
    return mindMapContent;
}
