// @blocksuite/affine-components/dist/context-menu/button.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { IS_MOBILE } from '@blocksuite/global/env';
import { CheckBoxCkeckSolidIcon, CheckBoxUnIcon, DoneIcon, } from '@blocksuite/icons/lit';
import { cssVarV2 } from '@toeverything/theme/v2';
import { css, html, unsafeCSS } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { keyed } from 'lit/directives/keyed.js';
import { MenuFocusable } from './focusable.js';
let MenuButton = (() => {
    let _classSuper = MenuFocusable;
    let _data_decorators;
    let _data_initializers = [];
    let _data_extraInitializers = [];
    return class MenuButton extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _data_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _data_decorators, { kind: "accessor", name: "data", static: false, private: false, access: { has: obj => "data" in obj, get: obj => obj.data, set: (obj, value) => { obj.data = value; } }, metadata: _metadata }, _data_initializers, _data_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .affine-menu-button {
      display: flex;
      width: 100%;
      font-size: 20px;
      cursor: pointer;
      align-items: center;
      padding: 4px;
      gap: 8px;
      border-radius: 4px;
      color: var(--affine-icon-color);
    }

    .affine-menu-button:hover,
    affine-menu-button.active .affine-menu-button {
      background-color: var(--affine-hover-color);
    }

    .affine-menu-button .affine-menu-action-text {
      flex: 1;
      font-size: 14px;
      line-height: 22px;
      color: var(--affine-text-primary-color);
    }

    .affine-menu-button.focused {
      outline: 1px solid ${unsafeCSS(cssVarV2.layer.insideBorder.primaryBorder)};
    }

    .affine-menu-button.delete-item:hover {
      background-color: var(--affine-background-error-color);
      color: var(--affine-error-color);
    }

    .affine-menu-button.delete-item:hover .affine-menu-action-text {
      color: var(--affine-error-color);
    }
  `; }
        connectedCallback() {
            super.connectedCallback();
            this.disposables.addFromEvent(this, 'mouseenter', () => {
                this.data.onHover?.(true);
                this.menu.closeSubMenu();
            });
            this.disposables.addFromEvent(this, 'mouseleave', () => {
                this.data.onHover?.(false);
            });
            this.disposables.addFromEvent(this, 'click', this.onClick);
        }
        disconnectedCallback() {
            super.disconnectedCallback();
            this.data.onHover?.(false);
        }
        onClick() {
            if (this.data.select(this) !== false) {
                this.menu.options.onComplete?.();
                this.menu.close();
            }
        }
        onPressEnter() {
            this.onClick();
        }
        render() {
            const classString = classMap({
                'affine-menu-button': true,
                focused: this.isFocused$.value,
                ...this.data.class,
            });
            return html ` <div class="${classString}">${this.data.content()}</div>`;
        }
        #data_accessor_storage = __runInitializers(this, _data_initializers, void 0);
        get data() { return this.#data_accessor_storage; }
        set data(value) { this.#data_accessor_storage = value; }
        constructor() {
            super(...arguments);
            __runInitializers(this, _data_extraInitializers);
        }
    };
})();
export { MenuButton };
let MobileMenuButton = (() => {
    let _classSuper = MenuFocusable;
    let _data_decorators;
    let _data_initializers = [];
    let _data_extraInitializers = [];
    return class MobileMenuButton extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _data_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _data_decorators, { kind: "accessor", name: "data", static: false, private: false, access: { has: obj => "data" in obj, get: obj => obj.data, set: (obj, value) => { obj.data = value; } }, metadata: _metadata }, _data_initializers, _data_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .mobile-menu-button {
      display: flex;
      width: 100%;
      cursor: pointer;
      align-items: center;
      font-size: 20px;
      padding: 11px 8px;
      gap: 8px;
      border-radius: 4px;
      color: var(--affine-icon-color);
    }

    .mobile-menu-button .affine-menu-action-text {
      flex: 1;
      color: var(--affine-text-primary-color);
      font-size: 17px;
      line-height: 22px;
    }

    .mobile-menu-button.delete-item {
      color: var(--affine-error-color);
    }

    .mobile-menu-button.delete-item .mobile-menu-action-text {
      color: var(--affine-error-color);
    }
  `; }
        connectedCallback() {
            super.connectedCallback();
            this.disposables.addFromEvent(this, 'click', this.onClick);
        }
        onClick() {
            if (this.data.select(this) !== false) {
                this.menu.options.onComplete?.();
                this.menu.close();
            }
        }
        onPressEnter() {
            this.onClick();
        }
        render() {
            const classString = classMap({
                'mobile-menu-button': true,
                focused: this.isFocused$.value,
                ...this.data.class,
            });
            return html ` <div class="${classString}">${this.data.content()}</div>`;
        }
        #data_accessor_storage = __runInitializers(this, _data_initializers, void 0);
        get data() { return this.#data_accessor_storage; }
        set data(value) { this.#data_accessor_storage = value; }
        constructor() {
            super(...arguments);
            __runInitializers(this, _data_extraInitializers);
        }
    };
})();
export { MobileMenuButton };
const renderButton = (data, menu) => {
    if (IS_MOBILE) {
        return html `<mobile-menu-button
      .data="${data}"
      .menu="${menu}"
    ></mobile-menu-button>`;
    }
    return html `<affine-menu-button
    .data="${data}"
    .menu="${menu}"
  ></affine-menu-button>`;
};
export const menuButtonItems = {
    action: (config) => menu => {
        if (config.hide?.() || !menu.search(config.name)) {
            return;
        }
        const data = {
            content: () => {
                return html `
            ${config.prefix}
            <div class="affine-menu-action-text">
              ${config.label?.() ?? config.name}
            </div>
            ${config.postfix ?? (config.isSelected ? DoneIcon() : undefined)}
          `;
            },
            onHover: config.onHover,
            select: config.select,
            class: {
                'selected-item': config.isSelected ?? false,
                ...config.class,
            },
        };
        return renderButton(data, menu);
    },
    checkbox: (config) => menu => {
        if (!menu.search(config.name)) {
            return;
        }
        const data = {
            content: () => html `
          ${config.checked.value
                ? CheckBoxCkeckSolidIcon({ style: `color:#1E96EB` })
                : CheckBoxUnIcon()}
          <div class="affine-menu-action-text">
            ${config.label?.() ?? config.name}
          </div>
          ${config.postfix}
        `,
            select: () => {
                config.select(config.checked.value);
                return false;
            },
            class: config.class ?? {},
        };
        return html `${keyed(config.name, renderButton(data, menu))}`;
    },
    toggleSwitch: (config) => menu => {
        if (!menu.search(config.name)) {
            return;
        }
        const onChange = (on) => {
            config.onChange(on);
        };
        const data = {
            content: () => html `
          <div class="affine-menu-action-text">
            ${config.label?.() ?? config.name}
          </div>
          <toggle-switch
            .on="${config.on}"
            .onChange="${onChange}"
          ></toggle-switch>
          ${config.postfix}
        `,
            select: () => {
                config.onChange(config.on);
                return false;
            },
            class: config.class ?? {},
        };
        return html `${keyed(config.name, renderButton(data, menu))}`;
    },
};
