// @blocksuite/affine-components/dist/context-menu/input.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { unsafeCSSVarV2 } from '@blocksuite/affine-shared/theme';
import { IS_MOBILE } from '@blocksuite/global/env';
import { cssVarV2 } from '@toeverything/theme/v2';
import { css, html } from 'lit';
import { property, query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { styleMap } from 'lit/directives/style-map.js';
import { MenuFocusable } from './focusable.js';
let MenuInput = (() => {
    let _classSuper = MenuFocusable;
    let _data_decorators;
    let _data_initializers = [];
    let _data_extraInitializers = [];
    let _inputRef_decorators;
    let _inputRef_initializers = [];
    let _inputRef_extraInitializers = [];
    return class MenuInput extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _data_decorators = [property({ attribute: false })];
            _inputRef_decorators = [query('input')];
            __esDecorate(this, null, _data_decorators, { kind: "accessor", name: "data", static: false, private: false, access: { has: obj => "data" in obj, get: obj => obj.data, set: (obj, value) => { obj.data = value; } }, metadata: _metadata }, _data_initializers, _data_extraInitializers);
            __esDecorate(this, null, _inputRef_decorators, { kind: "accessor", name: "inputRef", static: false, private: false, access: { has: obj => "inputRef" in obj, get: obj => obj.inputRef, set: (obj, value) => { obj.inputRef = value; } }, metadata: _metadata }, _inputRef_initializers, _inputRef_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .affine-menu-input {
      flex: 1;
      outline: none;
      border-radius: 4px;
      font-size: 14px;
      line-height: 22px;
      padding: 4px 6px;
      border: 1px solid var(--affine-border-color);
      width: 100%;
      color: ${unsafeCSSVarV2('text/primary')};
      background-color: transparent;
    }

    .affine-menu-input.focused {
      border: 1px solid ${unsafeCSSVarV2('layer/insideBorder/primaryBorder')};
    }

    .affine-menu-input:focus {
      border: 1px solid ${unsafeCSSVarV2('layer/insideBorder/primaryBorder')};
      box-shadow: 0px 0px 0px 2px rgba(28, 158, 228, 0.3);
    }
  `; }
        complete() {
            this.data.onComplete?.(this.inputRef.value);
        }
        connectedCallback() {
            super.connectedCallback();
            this.disposables.addFromEvent(this, 'click', e => {
                e.stopPropagation();
            });
            this.disposables.addFromEvent(this, 'mouseenter', () => {
                this.menu.closeSubMenu();
            });
            requestAnimationFrame(() => {
                requestAnimationFrame(() => {
                    this.inputRef.select();
                });
            });
        }
        onPressEnter() {
            this.inputRef.focus();
        }
        render() {
            const classString = classMap({
                [this.data.class ?? '']: true,
                'affine-menu-input': true,
                focused: this.isFocused$.value,
            });
            return html `<input
      @focus="${() => {
                this.menu.setFocusOnly(this);
            }}"
      @input="${this.onInput}"
      @keydown="${this.onKeydown}"
      @copy="${this.stopPropagation}"
      @paste="${this.stopPropagation}"
      @compositionend="${this.onCompositionEnd}"
      class="${classString}"
      value="${this.data.initialValue ?? ''}"
      type="text"
    />`;
        }
        #data_accessor_storage;
        get data() { return this.#data_accessor_storage; }
        set data(value) { this.#data_accessor_storage = value; }
        #inputRef_accessor_storage;
        get inputRef() { return this.#inputRef_accessor_storage; }
        set inputRef(value) { this.#inputRef_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.onCompositionEnd = () => {
                this.data.onChange?.(this.inputRef.value);
            };
            this.onInput = (e) => {
                e.stopPropagation();
                if (e.isComposing)
                    return;
                this.data.onChange?.(this.inputRef.value);
            };
            this.onKeydown = (e) => {
                e.stopPropagation();
                if (e.isComposing)
                    return;
                if (e.key === 'Escape') {
                    this.complete();
                    this.inputRef.blur();
                    this.menu.focusTo(this);
                    return;
                }
                if (e.key === 'Enter') {
                    this.complete();
                    this.menu.close();
                    return;
                }
            };
            this.stopPropagation = (e) => {
                e.stopPropagation();
            };
            this.#data_accessor_storage = __runInitializers(this, _data_initializers, void 0);
            this.#inputRef_accessor_storage = (__runInitializers(this, _data_extraInitializers), __runInitializers(this, _inputRef_initializers, void 0));
            __runInitializers(this, _inputRef_extraInitializers);
        }
    };
})();
export { MenuInput };
let MobileMenuInput = (() => {
    let _classSuper = MenuFocusable;
    let _data_decorators;
    let _data_initializers = [];
    let _data_extraInitializers = [];
    let _inputRef_decorators;
    let _inputRef_initializers = [];
    let _inputRef_extraInitializers = [];
    return class MobileMenuInput extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _data_decorators = [property({ attribute: false })];
            _inputRef_decorators = [query('input')];
            __esDecorate(this, null, _data_decorators, { kind: "accessor", name: "data", static: false, private: false, access: { has: obj => "data" in obj, get: obj => obj.data, set: (obj, value) => { obj.data = value; } }, metadata: _metadata }, _data_initializers, _data_extraInitializers);
            __esDecorate(this, null, _inputRef_decorators, { kind: "accessor", name: "inputRef", static: false, private: false, access: { has: obj => "inputRef" in obj, get: obj => obj.inputRef, set: (obj, value) => { obj.inputRef = value; } }, metadata: _metadata }, _inputRef_initializers, _inputRef_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static { this.styles = css `
    .mobile-menu-input {
      flex: 1;
      outline: none;
      font-size: 17px;
      line-height: 22px;
      border: none;
      width: 100%;
      color: ${unsafeCSSVarV2('text/primary')};
    }
  `; }
        complete() {
            this.data.onComplete?.(this.inputRef.value);
        }
        onPressEnter() {
            this.inputRef.focus();
        }
        render() {
            const classString = classMap({
                [this.data.class ?? '']: true,
                'mobile-menu-input': true,
                focused: this.isFocused$.value,
            });
            return html `<input
      @focus="${() => {
                this.menu.setFocusOnly(this);
            }}"
      @input="${this.onInput}"
      @copy="${this.stopPropagation}"
      @paste="${this.stopPropagation}"
      @compositionend="${this.onCompositionEnd}"
      class="${classString}"
      value="${this.data.initialValue ?? ''}"
      type="text"
    />`;
        }
        #data_accessor_storage;
        get data() { return this.#data_accessor_storage; }
        set data(value) { this.#data_accessor_storage = value; }
        #inputRef_accessor_storage;
        get inputRef() { return this.#inputRef_accessor_storage; }
        set inputRef(value) { this.#inputRef_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.onCompositionEnd = () => {
                this.data.onChange?.(this.inputRef.value);
            };
            this.onInput = (e) => {
                e.stopPropagation();
                if (e.isComposing)
                    return;
                this.data.onChange?.(this.inputRef.value);
            };
            this.stopPropagation = (e) => {
                e.stopPropagation();
            };
            this.#data_accessor_storage = __runInitializers(this, _data_initializers, void 0);
            this.#inputRef_accessor_storage = (__runInitializers(this, _data_extraInitializers), __runInitializers(this, _inputRef_initializers, void 0));
            __runInitializers(this, _inputRef_extraInitializers);
        }
    };
})();
export { MobileMenuInput };
const renderInput = (data, menu) => {
    if (IS_MOBILE) {
        return html ` <mobile-menu-input
      style="flex:1"
      .data="${data}"
      .menu="${menu}"
    ></mobile-menu-input>`;
    }
    return html ` <affine-menu-input
    style="flex:1"
    .data="${data}"
    .menu="${menu}"
  ></affine-menu-input>`;
};
export const menuInputItems = {
    input: (config) => menu => {
        if (menu.showSearch$.value) {
            return;
        }
        const data = {
            placeholder: config.placeholder,
            initialValue: config.initialValue,
            class: config.class,
            onComplete: config.onComplete,
            onChange: config.onChange,
        };
        const style = styleMap({
            display: 'flex',
            alignItems: 'center',
            ...(IS_MOBILE
                ? {
                    borderRadius: '12px',
                    backgroundColor: cssVarV2('layer/background/primary'),
                    padding: '12px',
                    gap: '8px',
                }
                : {
                    marginBottom: '8px',
                    gap: '4px',
                }),
            ...config.style,
        });
        return html `
        <div style="${style}">
          ${config.prefix} ${renderInput(data, menu)} ${config.postfix}
        </div>
      `;
    },
};
