// @blocksuite/affine-components/dist/context-menu/sub-menu.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { IS_MOBILE } from '@blocksuite/global/env';
import { ArrowRightSmallIcon } from '@blocksuite/icons/lit';
import { autoPlacement, autoUpdate, computePosition, offset, } from '@floating-ui/dom';
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { MenuFocusable } from './focusable.js';
import { Menu } from './menu.js';
import { popMenu, popupTargetFromElement } from './menu-renderer.js';
export const subMenuOffset = offset({
    mainAxis: 16,
    crossAxis: -8.5,
});
export const subMenuPlacements = autoPlacement({
    allowedPlacements: ['right-start', 'left-start', 'right-end', 'left-end'],
});
export const subMenuMiddleware = [subMenuOffset, subMenuPlacements];
let MenuSubMenu = (() => {
    let _classSuper = MenuFocusable;
    let _data_decorators;
    let _data_initializers = [];
    let _data_extraInitializers = [];
    return class MenuSubMenu extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _data_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _data_decorators, { kind: "accessor", name: "data", static: false, private: false, access: { has: obj => "data" in obj, get: obj => obj.data, set: (obj, value) => { obj.data = value; } }, metadata: _metadata }, _data_initializers, _data_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        connectedCallback() {
            super.connectedCallback();
            this.createTime = Date.now();
            this.disposables.addFromEvent(this, 'mouseenter', this.onMouseEnter);
            this.disposables.addFromEvent(this, 'click', e => {
                e.preventDefault();
                e.stopPropagation();
                if (this.data.select) {
                    this.data.select();
                    this.menu.close();
                }
                else {
                    this.openSubMenu();
                }
            });
        }
        onMouseEnter() {
            if (Date.now() - this.createTime > 100) {
                this.openSubMenu();
            }
        }
        onPressEnter() {
            this.onMouseEnter();
        }
        openSubMenu() {
            const focus = this.menu.currentFocused$.value;
            const menu = new Menu({
                ...this.data.options,
                onComplete: () => {
                    this.menu.close();
                },
                onClose: () => {
                    menu.menuElement.remove();
                    this.menu.focusTo(focus);
                    this.data.options.onClose?.();
                    unsub();
                },
            });
            this.menu.menuElement.parentElement?.append(menu.menuElement);
            const unsub = autoUpdate(this, menu.menuElement, () => {
                computePosition(this, menu.menuElement, {
                    middleware: subMenuMiddleware,
                })
                    .then(({ x, y }) => {
                    menu.menuElement.style.left = `${x}px`;
                    menu.menuElement.style.top = `${y}px`;
                })
                    .catch(err => console.error(err));
            });
            this.menu.openSubMenu(menu);
        }
        render() {
            const classString = classMap({
                [this.data.class ?? '']: true,
                'affine-menu-button': true,
                focused: this.isFocused$.value,
            });
            return html ` <div class="${classString}">${this.data.content()}</div>`;
        }
        #data_accessor_storage;
        get data() { return this.#data_accessor_storage; }
        set data(value) { this.#data_accessor_storage = value; }
        constructor() {
            super(...arguments);
            this.createTime = 0;
            this.#data_accessor_storage = __runInitializers(this, _data_initializers, void 0);
            __runInitializers(this, _data_extraInitializers);
        }
    };
})();
export { MenuSubMenu };
let MobileSubMenu = (() => {
    let _classSuper = MenuFocusable;
    let _data_decorators;
    let _data_initializers = [];
    let _data_extraInitializers = [];
    return class MobileSubMenu extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _data_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _data_decorators, { kind: "accessor", name: "data", static: false, private: false, access: { has: obj => "data" in obj, get: obj => obj.data, set: (obj, value) => { obj.data = value; } }, metadata: _metadata }, _data_initializers, _data_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        connectedCallback() {
            super.connectedCallback();
            this.disposables.addFromEvent(this, 'click', e => {
                e.preventDefault();
                e.stopPropagation();
                this.openSubMenu();
            });
        }
        onMouseEnter() {
            this.openSubMenu();
        }
        onPressEnter() {
            this.onMouseEnter();
        }
        openSubMenu() {
            const { menu } = popMenu(popupTargetFromElement(this), {
                options: {
                    ...this.data.options,
                    onComplete: () => {
                        this.menu.close();
                    },
                    onClose: () => {
                        menu.menuElement.remove();
                        this.data.options.onClose?.();
                    },
                },
            });
            this.menu.openSubMenu(menu);
        }
        render() {
            const classString = classMap({
                [this.data.class ?? '']: true,
                'mobile-menu-button': true,
                focused: this.isFocused$.value,
            });
            return html ` <div class="${classString}">${this.data.content()}</div>`;
        }
        #data_accessor_storage = __runInitializers(this, _data_initializers, void 0);
        get data() { return this.#data_accessor_storage; }
        set data(value) { this.#data_accessor_storage = value; }
        constructor() {
            super(...arguments);
            __runInitializers(this, _data_extraInitializers);
        }
    };
})();
export { MobileSubMenu };
export const renderSubMenu = (data, menu) => {
    if (IS_MOBILE) {
        return html ` <mobile-sub-menu
      .data="${data}"
      .menu="${menu}"
    ></mobile-sub-menu>`;
    }
    return html ` <affine-menu-sub-menu
    .data="${data}"
    .menu="${menu}"
  ></affine-menu-sub-menu>`;
};
export const subMenuItems = {
    subMenu: (config) => menu => {
        if (config.hide?.() || !menu.search(config.name)) {
            return;
        }
        const data = {
            content: () => html `${config.prefix}
            <div class="affine-menu-action-text">
              ${config.label?.() ?? config.name}
            </div>
            ${config.postfix}
            ${config.disableArrow ? nothing : ArrowRightSmallIcon()} `,
            class: config.class,
            options: config.options,
        };
        return renderSubMenu(data, menu);
    },
};
