// @blocksuite/affine-shared/dist/adapters/notion-html/delta-converter.js
import { createIdentifier, } from '@blocksuite/global/di';
import { isEqual } from '@blocksuite/global/utils';
import { DeltaASTConverter, } from '../types/adapter.js';
export const NotionHtmlASTToDeltaMatcherIdentifier = createIdentifier('NotionHtmlASTToDeltaMatcher');
export function NotionHtmlASTToDeltaExtension(matcher) {
    const identifier = NotionHtmlASTToDeltaMatcherIdentifier(matcher.name);
    return {
        setup: di => {
            di.addImpl(identifier, () => matcher);
        },
        identifier,
    };
}
export class NotionHtmlDeltaConverter extends DeltaASTConverter {
    constructor(configs, inlineDeltaMatchers, htmlASTToDeltaMatchers) {
        super();
        this.configs = configs;
        this.inlineDeltaMatchers = inlineDeltaMatchers;
        this.htmlASTToDeltaMatchers = htmlASTToDeltaMatchers;
    }
    _spreadAstToDelta(ast, options = Object.create(null)) {
        const context = {
            configs: this.configs,
            options,
            toDelta: (ast, options) => this._spreadAstToDelta(ast, options),
        };
        for (const matcher of this.htmlASTToDeltaMatchers) {
            if (matcher.match(ast)) {
                return matcher.toDelta(ast, context);
            }
        }
        const result = 'children' in ast
            ? ast.children.flatMap(child => this._spreadAstToDelta(child, options))
            : [];
        if (options.removeLastBr && result.length > 0) {
            const lastItem = result[result.length - 1];
            if (lastItem.insert === '\n') {
                result.pop();
            }
        }
        return result;
    }
    astToDelta(ast, options = Object.create(null)) {
        return this._spreadAstToDelta(ast, options).reduce((acc, cur) => {
            if (acc.length === 0) {
                return [cur];
            }
            const last = acc[acc.length - 1];
            if (typeof last.insert === 'string' &&
                typeof cur.insert === 'string' &&
                isEqual(last.attributes, cur.attributes)) {
                last.insert += cur.insert;
                return acc;
            }
            return [...acc, cur];
        }, []);
    }
    deltaToAST(_) {
        return [];
    }
}
