// @blocksuite/block-std/dist/gfx/model/surface/element-model.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var _a;
import { Slot, } from '@blocksuite/global/utils';
import { Bound, deserializeXYWH, DisposableGroup, getBoundWithRotation, getPointsFromBoundWithRotation, isEqual, linePolygonIntersects, PointLocation, polygonGetPointTangent, polygonNearestPoint, randomSeed, rotatePoints, } from '@blocksuite/global/utils';
import { DocCollection } from '@blocksuite/store';
import { createMutex } from 'lib0/mutex';
import { descendantElementsImpl, hasDescendantElementImpl, isLockedByAncestorImpl, isLockedBySelfImpl, isLockedImpl, lockElementImpl, unlockElementImpl, } from '../../../utils/tree.js';
import { gfxGroupCompatibleSymbol } from '../base.js';
import { convertProps, field, getDerivedProps, getFieldPropsSet, local, updateDerivedProps, watch, } from './decorators/index.js';
let GfxPrimitiveElementModel = (() => {
    let _display_decorators;
    let _display_initializers = [];
    let _display_extraInitializers = [];
    let _externalXYWH_decorators;
    let _externalXYWH_initializers = [];
    let _externalXYWH_extraInitializers = [];
    let _hidden_decorators;
    let _hidden_initializers = [];
    let _hidden_extraInitializers = [];
    let _index_decorators;
    let _index_initializers = [];
    let _index_extraInitializers = [];
    let _lockedBySelf_decorators;
    let _lockedBySelf_initializers = [];
    let _lockedBySelf_extraInitializers = [];
    let _opacity_decorators;
    let _opacity_initializers = [];
    let _opacity_extraInitializers = [];
    let _responseExtension_decorators;
    let _responseExtension_initializers = [];
    let _responseExtension_extraInitializers = [];
    let _seed_decorators;
    let _seed_initializers = [];
    let _seed_extraInitializers = [];
    return class GfxPrimitiveElementModel {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _display_decorators = [local()];
            _externalXYWH_decorators = [watch((_, instance) => {
                    instance['_local'].delete('externalBound');
                }), local()];
            _hidden_decorators = [field(false)];
            _index_decorators = [field()];
            _lockedBySelf_decorators = [field()];
            _opacity_decorators = [local()];
            _responseExtension_decorators = [local()];
            _seed_decorators = [field()];
            __esDecorate(this, null, _display_decorators, { kind: "accessor", name: "display", static: false, private: false, access: { has: obj => "display" in obj, get: obj => obj.display, set: (obj, value) => { obj.display = value; } }, metadata: _metadata }, _display_initializers, _display_extraInitializers);
            __esDecorate(this, null, _externalXYWH_decorators, { kind: "accessor", name: "externalXYWH", static: false, private: false, access: { has: obj => "externalXYWH" in obj, get: obj => obj.externalXYWH, set: (obj, value) => { obj.externalXYWH = value; } }, metadata: _metadata }, _externalXYWH_initializers, _externalXYWH_extraInitializers);
            __esDecorate(this, null, _hidden_decorators, { kind: "accessor", name: "hidden", static: false, private: false, access: { has: obj => "hidden" in obj, get: obj => obj.hidden, set: (obj, value) => { obj.hidden = value; } }, metadata: _metadata }, _hidden_initializers, _hidden_extraInitializers);
            __esDecorate(this, null, _index_decorators, { kind: "accessor", name: "index", static: false, private: false, access: { has: obj => "index" in obj, get: obj => obj.index, set: (obj, value) => { obj.index = value; } }, metadata: _metadata }, _index_initializers, _index_extraInitializers);
            __esDecorate(this, null, _lockedBySelf_decorators, { kind: "accessor", name: "lockedBySelf", static: false, private: false, access: { has: obj => "lockedBySelf" in obj, get: obj => obj.lockedBySelf, set: (obj, value) => { obj.lockedBySelf = value; } }, metadata: _metadata }, _lockedBySelf_initializers, _lockedBySelf_extraInitializers);
            __esDecorate(this, null, _opacity_decorators, { kind: "accessor", name: "opacity", static: false, private: false, access: { has: obj => "opacity" in obj, get: obj => obj.opacity, set: (obj, value) => { obj.opacity = value; } }, metadata: _metadata }, _opacity_initializers, _opacity_extraInitializers);
            __esDecorate(this, null, _responseExtension_decorators, { kind: "accessor", name: "responseExtension", static: false, private: false, access: { has: obj => "responseExtension" in obj, get: obj => obj.responseExtension, set: (obj, value) => { obj.responseExtension = value; } }, metadata: _metadata }, _responseExtension_initializers, _responseExtension_extraInitializers);
            __esDecorate(this, null, _seed_decorators, { kind: "accessor", name: "seed", static: false, private: false, access: { has: obj => "seed" in obj, get: obj => obj.seed, set: (obj, value) => { obj.seed = value; } }, metadata: _metadata }, _seed_initializers, _seed_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        get connectable() {
            return true;
        }
        get deserializedXYWH() {
            if (!this._lastXYWH || this.xywh !== this._lastXYWH) {
                const xywh = this.xywh;
                this._local.set('deserializedXYWH', deserializeXYWH(xywh));
                this._lastXYWH = xywh;
            }
            return this._local.get('deserializedXYWH') ?? [0, 0, 0, 0];
        }
        /**
         * The bound of the element after rotation.
         * The bound without rotation should be created by `Bound.deserialize(this.xywh)`.
         */
        get elementBound() {
            if (this.rotate) {
                return Bound.from(getBoundWithRotation(this));
            }
            return Bound.deserialize(this.xywh);
        }
        get externalBound() {
            if (!this._local.has('externalBound')) {
                const bound = this.externalXYWH
                    ? Bound.deserialize(this.externalXYWH)
                    : null;
                this._local.set('externalBound', bound);
            }
            return this._local.get('externalBound');
        }
        get group() {
            return this.surface.getGroup(this.id);
        }
        /**
         * Return the ancestor elements in order from the most recent to the earliest.
         */
        get groups() {
            return this.surface.getGroups(this.id);
        }
        get h() {
            return this.deserializedXYWH[3];
        }
        get id() {
            return this._id;
        }
        get isConnected() {
            return this.surface.hasElementById(this.id);
        }
        get responseBound() {
            return this.elementBound.expand(this.responseExtension);
        }
        get w() {
            return this.deserializedXYWH[2];
        }
        get x() {
            return this.deserializedXYWH[0];
        }
        get y() {
            return this.deserializedXYWH[1];
        }
        constructor(options) {
            this._disposable = new DisposableGroup();
            this._local = new Map();
            /**
             * Used to store a copy of data in the yMap.
             */
            this._preserved = new Map();
            this.propsUpdated = new Slot();
            this.#display_accessor_storage = __runInitializers(this, _display_initializers, true);
            this.#externalXYWH_accessor_storage = (__runInitializers(this, _display_extraInitializers), __runInitializers(this, _externalXYWH_initializers, undefined));
            this.#hidden_accessor_storage = (__runInitializers(this, _externalXYWH_extraInitializers), __runInitializers(this, _hidden_initializers, false));
            this.#index_accessor_storage = (__runInitializers(this, _hidden_extraInitializers), __runInitializers(this, _index_initializers, void 0));
            this.#lockedBySelf_accessor_storage = (__runInitializers(this, _index_extraInitializers), __runInitializers(this, _lockedBySelf_initializers, false));
            this.#opacity_accessor_storage = (__runInitializers(this, _lockedBySelf_extraInitializers), __runInitializers(this, _opacity_initializers, 1));
            this.#responseExtension_accessor_storage = (__runInitializers(this, _opacity_extraInitializers), __runInitializers(this, _responseExtension_initializers, [0, 0]));
            this.#seed_accessor_storage = (__runInitializers(this, _responseExtension_extraInitializers), __runInitializers(this, _seed_initializers, void 0));
            __runInitializers(this, _seed_extraInitializers);
            const { id, yMap, model, stashedStore, onChange } = options;
            this._id = id;
            this.yMap = yMap;
            this.surface = model;
            this._stashed = stashedStore;
            this._onChange = onChange;
            this.index = 'a0';
            this.seed = randomSeed();
        }
        static propsToY(props) {
            return props;
        }
        containsBound(bounds) {
            return getPointsFromBoundWithRotation(this).some(point => bounds.containsPoint(point));
        }
        getLineIntersections(start, end) {
            const points = getPointsFromBoundWithRotation(this);
            return linePolygonIntersects(start, end, points);
        }
        getNearestPoint(point) {
            const points = getPointsFromBoundWithRotation(this);
            return polygonNearestPoint(points, point);
        }
        getRelativePointLocation(relativePoint) {
            const bound = Bound.deserialize(this.xywh);
            const point = bound.getRelativePoint(relativePoint);
            const rotatePoint = rotatePoints([point], bound.center, this.rotate)[0];
            const points = rotatePoints(bound.points, bound.center, this.rotate);
            const tangent = polygonGetPointTangent(points, rotatePoint);
            return new PointLocation(rotatePoint, tangent);
        }
        includesPoint(x, y, opt, __) {
            const bound = opt.useElementBound ? this.elementBound : this.responseBound;
            return bound.isPointInBound([x, y]);
        }
        intersectsBound(bound) {
            return (this.containsBound(bound) ||
                bound.points.some((point, i, points) => this.getLineIntersections(point, points[(i + 1) % points.length])));
        }
        isLocked() {
            return isLockedImpl(this);
        }
        isLockedByAncestor() {
            return isLockedByAncestorImpl(this);
        }
        isLockedBySelf() {
            return isLockedBySelfImpl(this);
        }
        lock() {
            lockElementImpl(this.surface.doc, this);
        }
        onCreated() { }
        onDestroyed() {
            this._disposable.dispose();
            this.propsUpdated.dispose();
        }
        pop(prop) {
            if (!this._stashed.has(prop)) {
                return;
            }
            const value = this._stashed.get(prop);
            this._stashed.delete(prop);
            // @ts-ignore
            delete this[prop];
            if (getFieldPropsSet(this).has(prop)) {
                if (!isEqual(value, this.yMap.get(prop))) {
                    this.surface.doc.transact(() => {
                        this.yMap.set(prop, value);
                    });
                }
            }
            else {
                console.warn('pop a prop that is not field or local:', prop);
            }
        }
        serialize() {
            const result = this.yMap.toJSON();
            result.xywh = this.xywh;
            return result;
        }
        stash(prop) {
            if (this._stashed.has(prop)) {
                return;
            }
            if (!getFieldPropsSet(this).has(prop)) {
                return;
            }
            const curVal = this[prop];
            this._stashed.set(prop, curVal);
            Object.defineProperty(this, prop, {
                configurable: true,
                enumerable: true,
                get: () => this._stashed.get(prop),
                set: (original) => {
                    const value = convertProps(prop, original, this);
                    const oldValue = this._stashed.get(prop);
                    const derivedProps = getDerivedProps(prop, original, this);
                    this._stashed.set(prop, value);
                    this._onChange({
                        props: {
                            [prop]: value,
                        },
                        oldValues: {
                            [prop]: oldValue,
                        },
                        local: true,
                    });
                    updateDerivedProps(derivedProps, this);
                },
            });
        }
        unlock() {
            unlockElementImpl(this.surface.doc, this);
        }
        #display_accessor_storage;
        get display() { return this.#display_accessor_storage; }
        set display(value) { this.#display_accessor_storage = value; }
        #externalXYWH_accessor_storage;
        /**
         * In some cases, you need to draw something related to the element, but it does not belong to the element itself.
         * And it is also interactive, you can select element by clicking on it. E.g. the title of the group element.
         * In this case, we need to store this kind of external xywh in order to do hit test. This property should not be synced to the doc.
         * This property should be updated every time it gets rendered.
         */
        get externalXYWH() { return this.#externalXYWH_accessor_storage; }
        set externalXYWH(value) { this.#externalXYWH_accessor_storage = value; }
        #hidden_accessor_storage;
        get hidden() { return this.#hidden_accessor_storage; }
        set hidden(value) { this.#hidden_accessor_storage = value; }
        #index_accessor_storage;
        get index() { return this.#index_accessor_storage; }
        set index(value) { this.#index_accessor_storage = value; }
        #lockedBySelf_accessor_storage;
        get lockedBySelf() { return this.#lockedBySelf_accessor_storage; }
        set lockedBySelf(value) { this.#lockedBySelf_accessor_storage = value; }
        #opacity_accessor_storage;
        get opacity() { return this.#opacity_accessor_storage; }
        set opacity(value) { this.#opacity_accessor_storage = value; }
        #responseExtension_accessor_storage;
        get responseExtension() { return this.#responseExtension_accessor_storage; }
        set responseExtension(value) { this.#responseExtension_accessor_storage = value; }
        #seed_accessor_storage;
        get seed() { return this.#seed_accessor_storage; }
        set seed(value) { this.#seed_accessor_storage = value; }
    };
})();
export { GfxPrimitiveElementModel };
export class GfxGroupLikeElementModel extends GfxPrimitiveElementModel {
    constructor() {
        super(...arguments);
        this._childIds = [];
        this._mutex = createMutex();
        this[_a] = true;
    }
    static { _a = gfxGroupCompatibleSymbol; }
    get childElements() {
        const elements = [];
        for (const key of this.childIds) {
            const element = this.surface.getElementById(key) ||
                this.surface.doc.getBlockById(key);
            element && elements.push(element);
        }
        return elements;
    }
    /**
     * The ids of the children. Its role is to provide a unique way to access the children.
     * You should update this field through `setChildIds` when the children are added or removed.
     */
    get childIds() {
        return this._childIds;
    }
    get descendantElements() {
        return descendantElementsImpl(this);
    }
    get xywh() {
        this._mutex(() => {
            const curXYWH = this._local.get('xywh') ?? '[0,0,0,0]';
            const newXYWH = this._getXYWH().serialize();
            if (curXYWH !== newXYWH || !this._local.has('xywh')) {
                this._local.set('xywh', newXYWH);
                if (curXYWH !== newXYWH) {
                    this._onChange({
                        props: {
                            xywh: newXYWH,
                        },
                        oldValues: {
                            xywh: curXYWH,
                        },
                        local: true,
                    });
                }
            }
        });
        return this._local.get('xywh') ?? '[0,0,0,0]';
    }
    set xywh(_) { }
    _getXYWH() {
        let bound;
        this.childElements.forEach(child => {
            if (child instanceof GfxPrimitiveElementModel && child.hidden) {
                return;
            }
            bound = bound ? bound.unite(child.elementBound) : child.elementBound;
        });
        if (bound) {
            this._local.set('xywh', bound.serialize());
        }
        else {
            this._local.delete('xywh');
        }
        return bound ?? new Bound(0, 0, 0, 0);
    }
    /**
     * The actual field that stores the children of the group.
     * It should be a ymap decorated with `@field`.
     */
    hasChild(element) {
        return this.childElements.includes(element);
    }
    /**
     * Check if the group has the given descendant.
     */
    hasDescendant(element) {
        return hasDescendantElementImpl(this, element);
    }
    /**
     * Set the new value of the childIds
     * @param value the new value of the childIds
     * @param fromLocal if true, the change is happened in the local
     */
    setChildIds(value, fromLocal) {
        const oldChildIds = this.childIds;
        this._childIds = value;
        this._onChange({
            props: {
                childIds: value,
            },
            oldValues: {
                childIds: oldChildIds,
            },
            local: fromLocal,
        });
    }
}
export function syncElementFromY(model, callback) {
    const disposables = {};
    const observer = (event, transaction) => {
        const props = {};
        const oldValues = {};
        event.keysChanged.forEach(key => {
            const type = event.changes.keys.get(key);
            const oldValue = event.changes.keys.get(key)?.oldValue;
            if (!type) {
                return;
            }
            if (type.action === 'update' || type.action === 'add') {
                const value = model.yMap.get(key);
                if (value instanceof DocCollection.Y.Text) {
                    disposables[key]?.();
                    disposables[key] = watchText(key, value, callback);
                }
                model['_preserved'].set(key, value);
                props[key] = value;
                oldValues[key] = oldValue;
            }
            else {
                model['_preserved'].delete(key);
                oldValues[key] = oldValue;
            }
        });
        callback({
            props,
            oldValues,
            local: transaction.local,
        });
    };
    Array.from(model.yMap.entries()).forEach(([key, value]) => {
        if (value instanceof DocCollection.Y.Text) {
            disposables[key] = watchText(key, value, callback);
        }
        model['_preserved'].set(key, value);
    });
    model.yMap.observe(observer);
    disposables['ymap'] = () => {
        model.yMap.unobserve(observer);
    };
    return () => {
        Object.values(disposables).forEach(fn => fn());
    };
}
function watchText(key, value, callback) {
    const fn = (_, transaction) => {
        callback({
            props: {
                [key]: value,
            },
            oldValues: {},
            local: transaction.local,
        });
    };
    value.observe(fn);
    return () => {
        value.unobserve(fn);
    };
}
