// @blocksuite/block-std/dist/gfx/model/surface/local-element-model.js
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { Bound, deserializeXYWH, getPointsFromBoundWithRotation, linePolygonIntersects, PointLocation, polygonGetPointTangent, polygonNearestPoint, rotatePoints, } from '@blocksuite/global/utils';
import { mutex } from 'lib0';
export function prop() {
    return function propDecorator(_target, context) {
        const prop = context.name;
        return {
            init(val) {
                this._props.add(prop);
                this._local.set(prop, val);
            },
            get() {
                return this._local.get(prop);
            },
            set(val) {
                this._local.set(prop, val);
            },
        };
    };
}
let GfxLocalElementModel = (() => {
    let _groupId_decorators;
    let _groupId_initializers = [];
    let _groupId_extraInitializers = [];
    let _hidden_decorators;
    let _hidden_initializers = [];
    let _hidden_extraInitializers = [];
    let _index_decorators;
    let _index_initializers = [];
    let _index_extraInitializers = [];
    let _opacity_decorators;
    let _opacity_initializers = [];
    let _opacity_extraInitializers = [];
    let _responseExtension_decorators;
    let _responseExtension_initializers = [];
    let _responseExtension_extraInitializers = [];
    let _rotate_decorators;
    let _rotate_initializers = [];
    let _rotate_extraInitializers = [];
    let _seed_decorators;
    let _seed_initializers = [];
    let _seed_extraInitializers = [];
    let _xywh_decorators;
    let _xywh_initializers = [];
    let _xywh_extraInitializers = [];
    return class GfxLocalElementModel {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _groupId_decorators = [prop()];
            _hidden_decorators = [prop()];
            _index_decorators = [prop()];
            _opacity_decorators = [prop()];
            _responseExtension_decorators = [prop()];
            _rotate_decorators = [prop()];
            _seed_decorators = [prop()];
            _xywh_decorators = [prop()];
            __esDecorate(this, null, _groupId_decorators, { kind: "accessor", name: "groupId", static: false, private: false, access: { has: obj => "groupId" in obj, get: obj => obj.groupId, set: (obj, value) => { obj.groupId = value; } }, metadata: _metadata }, _groupId_initializers, _groupId_extraInitializers);
            __esDecorate(this, null, _hidden_decorators, { kind: "accessor", name: "hidden", static: false, private: false, access: { has: obj => "hidden" in obj, get: obj => obj.hidden, set: (obj, value) => { obj.hidden = value; } }, metadata: _metadata }, _hidden_initializers, _hidden_extraInitializers);
            __esDecorate(this, null, _index_decorators, { kind: "accessor", name: "index", static: false, private: false, access: { has: obj => "index" in obj, get: obj => obj.index, set: (obj, value) => { obj.index = value; } }, metadata: _metadata }, _index_initializers, _index_extraInitializers);
            __esDecorate(this, null, _opacity_decorators, { kind: "accessor", name: "opacity", static: false, private: false, access: { has: obj => "opacity" in obj, get: obj => obj.opacity, set: (obj, value) => { obj.opacity = value; } }, metadata: _metadata }, _opacity_initializers, _opacity_extraInitializers);
            __esDecorate(this, null, _responseExtension_decorators, { kind: "accessor", name: "responseExtension", static: false, private: false, access: { has: obj => "responseExtension" in obj, get: obj => obj.responseExtension, set: (obj, value) => { obj.responseExtension = value; } }, metadata: _metadata }, _responseExtension_initializers, _responseExtension_extraInitializers);
            __esDecorate(this, null, _rotate_decorators, { kind: "accessor", name: "rotate", static: false, private: false, access: { has: obj => "rotate" in obj, get: obj => obj.rotate, set: (obj, value) => { obj.rotate = value; } }, metadata: _metadata }, _rotate_initializers, _rotate_extraInitializers);
            __esDecorate(this, null, _seed_decorators, { kind: "accessor", name: "seed", static: false, private: false, access: { has: obj => "seed" in obj, get: obj => obj.seed, set: (obj, value) => { obj.seed = value; } }, metadata: _metadata }, _seed_initializers, _seed_extraInitializers);
            __esDecorate(this, null, _xywh_decorators, { kind: "accessor", name: "xywh", static: false, private: false, access: { has: obj => "xywh" in obj, get: obj => obj.xywh, set: (obj, value) => { obj.xywh = value; } }, metadata: _metadata }, _xywh_initializers, _xywh_extraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        get deserializedXYWH() {
            if (!this._local.has('deserializedXYWH')) {
                const xywh = this.xywh;
                const deserialized = deserializeXYWH(xywh);
                this._local.set('deserializedXYWH', deserialized);
            }
            return this._local.get('deserializedXYWH');
        }
        get elementBound() {
            return new Bound(this.x, this.y, this.w, this.h);
        }
        get group() {
            return (this.groupId ? this._surface.getElementById(this.groupId) : null);
        }
        get groups() {
            if (this.group) {
                const groups = this._surface.getGroups(this.group.id);
                groups.unshift(this.group);
                return groups;
            }
            return [];
        }
        get h() {
            return this.deserializedXYWH[3];
        }
        get responseBound() {
            return this.elementBound.expand(this.responseExtension);
        }
        get surface() {
            return this._surface;
        }
        get w() {
            return this.deserializedXYWH[2];
        }
        get x() {
            return this.deserializedXYWH[0];
        }
        get y() {
            return this.deserializedXYWH[1];
        }
        constructor(surfaceModel) {
            this._mutex = mutex.createMutex();
            this._local = new Map();
            /**
             * Used to store all the name of the properties that have been decorated
             * with the `@prop`
             */
            this._props = new Set();
            /**
             * used to store the properties' cache key
             * when the properties required heavy computation
             */
            this.cache = new Map();
            this.id = '';
            this.#groupId_accessor_storage = __runInitializers(this, _groupId_initializers, '');
            this.#hidden_accessor_storage = (__runInitializers(this, _groupId_extraInitializers), __runInitializers(this, _hidden_initializers, false));
            this.#index_accessor_storage = (__runInitializers(this, _hidden_extraInitializers), __runInitializers(this, _index_initializers, 'a0'));
            this.#opacity_accessor_storage = (__runInitializers(this, _index_extraInitializers), __runInitializers(this, _opacity_initializers, 1));
            this.#responseExtension_accessor_storage = (__runInitializers(this, _opacity_extraInitializers), __runInitializers(this, _responseExtension_initializers, [0, 0]));
            this.#rotate_accessor_storage = (__runInitializers(this, _responseExtension_extraInitializers), __runInitializers(this, _rotate_initializers, 0));
            this.#seed_accessor_storage = (__runInitializers(this, _rotate_extraInitializers), __runInitializers(this, _seed_initializers, Math.random()));
            this.#xywh_accessor_storage = (__runInitializers(this, _seed_extraInitializers), __runInitializers(this, _xywh_initializers, '[0,0,0,0]'));
            __runInitializers(this, _xywh_extraInitializers);
            this._surface = surfaceModel;
            const p = new Proxy(this, {
                set: (target, prop, value) => {
                    if (prop === 'xywh') {
                        this._local.delete('deserializedXYWH');
                    }
                    // @ts-ignore
                    const oldValue = target[prop];
                    if (oldValue === value) {
                        return true;
                    }
                    // @ts-ignore
                    target[prop] = value;
                    if (!this._props.has(prop)) {
                        return true;
                    }
                    if (surfaceModel.localElementModels.has(p)) {
                        this._mutex(() => {
                            surfaceModel.localElementUpdated.emit({
                                model: p,
                                props: {
                                    [prop]: value,
                                },
                                oldValues: {
                                    [prop]: oldValue,
                                },
                            });
                        });
                    }
                    return true;
                },
            });
            return p;
        }
        containsBound(bounds) {
            return getPointsFromBoundWithRotation(this).some(point => bounds.containsPoint(point));
        }
        getLineIntersections(start, end) {
            const points = getPointsFromBoundWithRotation(this);
            return linePolygonIntersects(start, end, points);
        }
        getNearestPoint(point) {
            const points = getPointsFromBoundWithRotation(this);
            return polygonNearestPoint(points, point);
        }
        getRelativePointLocation(relativePoint) {
            const bound = Bound.deserialize(this.xywh);
            const point = bound.getRelativePoint(relativePoint);
            const rotatePoint = rotatePoints([point], bound.center, this.rotate)[0];
            const points = rotatePoints(bound.points, bound.center, this.rotate);
            const tangent = polygonGetPointTangent(points, rotatePoint);
            return new PointLocation(rotatePoint, tangent);
        }
        includesPoint(x, y, opt, __) {
            const bound = opt.useElementBound ? this.elementBound : this.responseBound;
            return bound.isPointInBound([x, y]);
        }
        intersectsBound(bound) {
            return (this.containsBound(bound) ||
                bound.points.some((point, i, points) => this.getLineIntersections(point, points[(i + 1) % points.length])));
        }
        isLocked() {
            return false;
        }
        isLockedByAncestor() {
            return false;
        }
        isLockedBySelf() {
            return false;
        }
        lock() {
            return;
        }
        unlock() {
            return;
        }
        #groupId_accessor_storage;
        get groupId() { return this.#groupId_accessor_storage; }
        set groupId(value) { this.#groupId_accessor_storage = value; }
        #hidden_accessor_storage;
        get hidden() { return this.#hidden_accessor_storage; }
        set hidden(value) { this.#hidden_accessor_storage = value; }
        #index_accessor_storage;
        get index() { return this.#index_accessor_storage; }
        set index(value) { this.#index_accessor_storage = value; }
        #opacity_accessor_storage;
        get opacity() { return this.#opacity_accessor_storage; }
        set opacity(value) { this.#opacity_accessor_storage = value; }
        #responseExtension_accessor_storage;
        get responseExtension() { return this.#responseExtension_accessor_storage; }
        set responseExtension(value) { this.#responseExtension_accessor_storage = value; }
        #rotate_accessor_storage;
        get rotate() { return this.#rotate_accessor_storage; }
        set rotate(value) { this.#rotate_accessor_storage = value; }
        #seed_accessor_storage;
        get seed() { return this.#seed_accessor_storage; }
        set seed(value) { this.#seed_accessor_storage = value; }
        #xywh_accessor_storage;
        get xywh() { return this.#xywh_accessor_storage; }
        set xywh(value) { this.#xywh_accessor_storage = value; }
    };
})();
export { GfxLocalElementModel };
