// @blocksuite/affine-shared/dist/services/drag-handle-config.js
import { Extension, StdIdentifier, } from '@blocksuite/block-std';
import { createIdentifier } from '@blocksuite/global/di';
import { Job, Slice } from '@blocksuite/store';
export const DragHandleConfigIdentifier = createIdentifier('AffineDragHandleIdentifier');
export function DragHandleConfigExtension(option) {
    return {
        setup: di => {
            const key = typeof option.flavour === 'string'
                ? option.flavour
                : option.flavour.source;
            di.addImpl(DragHandleConfigIdentifier(key), () => option);
        },
    };
}
export const DndApiExtensionIdentifier = createIdentifier('AffineDndApiIdentifier');
export class DNDAPIExtension extends Extension {
    constructor(std) {
        super();
        this.std = std;
        this.mimeType = 'application/x-blocksuite-dnd';
    }
    static setup(di) {
        di.add(this, [StdIdentifier]);
        di.addImpl(DndApiExtensionIdentifier, provider => provider.get(this));
    }
    decodeSnapshot(data) {
        return JSON.parse(decodeURIComponent(data));
    }
    encodeSnapshot(json) {
        const snapshot = JSON.stringify(json);
        return encodeURIComponent(snapshot);
    }
    fromEntity(options) {
        const { docId, flavour = 'affine:embed-linked-doc', blockId } = options;
        const slice = Slice.fromModels(this.std.doc, []);
        const job = new Job({ collection: this.std.collection });
        const snapshot = job.sliceToSnapshot(slice);
        if (!snapshot) {
            console.error('Failed to convert slice to snapshot');
            return null;
        }
        const props = {
            ...(blockId ? { blockId } : {}),
            pageId: docId,
        };
        return {
            ...snapshot,
            content: [
                {
                    id: this.std.collection.idGenerator(),
                    type: 'block',
                    flavour,
                    props,
                    children: [],
                },
            ],
        };
    }
}
